<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><title>Release Notes (October 2, 2004) - The NCBI C++ Toolkit Book [Internet] - NCBI</title><meta name="description" content="A comprehensive manual on the NCBI C++ toolkit, including its design and development framework, a C++ library reference, software examples and demos, FAQs and release notes." /><meta name="robots" content="index,nofollow,noarchive" /><meta name="page_id" content="release_notes_10_2_2004" /><meta name="page_type" content="chapter" /><meta name="ncbipopper" content="triggerPosition:'bottom left',destPosition:'top left',openEvent:'click',closeEvent:'click'" /><script type="text/javascript" src="http://www.ncbi.nlm.nih.gov/core/jig/1.13/js/jig.min.js "></script><link type="text/css" rel="stylesheet" href="css/ptbooks.css" /><link type="text/css" rel="stylesheet" href="css/toolkit.css" /></head><body><div class="grid no_max_width"><div class="col twelve_col nomargin shadow"><div class="content"><div class="col eight_col"><div class="nvgtn"><div class="pagination"><a class="pagelink active prev" title="Previous page" href="release_notes_11_22_2004.html">&lt; Prev</a><a class="pagelink active next" title="Next page" href="release_notes_7_8_2004.html">Next &gt;</a></div><div class="pdf-link"><a class="print-link" title="PDF version of this page" href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/public_releases/pdf/release_notes_10_2_2004.pdf">PDF</a></div></div><div class="meta_content"><h1><span class="title">Release Notes (October 2, 2004)</span></h1></div><div class="body-content jig-ncbiinpagenav" data-jigconfig="allHeadingLevels: ['h2'],smoothScroll: false"><ul><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.download_loc_10_">Download Location</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.source_archive_1">Source Archive Contents</a>
					<ul><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.source_code_10_2">Source Code Archives</a></p></li></ul>
				</p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.new_development_">New Development</a>
					<ul><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.corelib_10_2_200">CORELIB -- Portability and Application Framework</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.connect_10_2_200">CONNECT -- Data streaming, Networking, and Dispatching</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.html_10_2_2004">HTML -- HTML Generation Library</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.util_comp_10_2_2">UTIL/COMPRESSION -- Data Compression (GZIP, BZ2)</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.serial_10_2_2004">SERIAL -- Data Serialization (ASN.1, XML)</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.datatool_10_2_20">DATATOOL -- Code Generator and Data Converter Tool</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.cgi_10_2_2004">CGI  --  CGI and Fast-CGI Application Framework</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.berkeley_dbapi_1">Berkeley DB API (bdb) -- Much Enriched C++ API Based On BerkeleyDB</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.dbapi_10_2_2004">DBAPI -- Generic SQL Database Connectivity</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.algoalign_10_2_2">ALGO/ALIGN -- Generic Alignment Algorithms</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.algo_phy_10_2_20">ALGO/PHY_TREE -- Phylogenetic and bio tree algorithms</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.algoalign_gen_10">ALGO/ALIGN -- Generic Alignment Algorithms</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.blast_10_2_2004">BLAST</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.bioobj_10_2_2004">BIO-OBJECTS -- Bio-Object Specific Utility Functions (Not Involving OM++)</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.om_10_2_2004">OM++ -- Object Manager -- For Retrieving and Processing Bio-Objects</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.omlr_10_2_2004">OM++ LOADERS/READERS -- Data Retrieval Libraries for OM++</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.id2_10_2_2004">ID2</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.biotools_10_2_20">BIO-TOOLS</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.bfmsvc_10_2_2004">BUILD FRAMEWORK (MSVC++ 6.0)</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.bfdotnet_10_2_20">BUILD FRAMEWORK (MSVC++ .NET)</a></p></li></ul>
				</p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.apps_10_2_2004">APPLICATIONS</a>
					<ul><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.apps_objmgr_demo">OBJMGR DEMO program (objmgr_demo)</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.apps_objmgr_test">OBJMGR TEST programs (test_objmgr_data &amp; test_objmgr_data_mt)</a></p></li></ul>
				</p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.documentation_10">Documentation</a>
					<ul><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.doc_loc_10_2_200">Document Location</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.doc_changes_10_2">Document Content</a></p></li></ul>
				</p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.build_macos_10_2">Building on the MacOS</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.platforms_10_2_2">Platforms (OS's, compilers used inside NCBI)</a>
					<ul><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.platform_unix_10">Unix</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.platform_windows">MS Windows</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.platform_mac_10_">Mac OS X</a></p></li></ul>
				</p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.caveats_10_2_200">Caveats and Hints</a>
					<ul><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.macos_10x_10_2_2">MacOS 10.X / CodeWarrior 9.1</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.macos_10_X_10_2_">MacOS 10.2/GCC 3.3</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.gcc_2_95_3_10_2_">GCC 2.95</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.gcc_3_0_4_10_2_2">GCC 3.0.4</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.gcc_3_3_10_2_200">GCC 3.3</a></p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.gcc_3_4_10_2_200">GCC 3.4</a></p></li></ul>
				</p></li><li><p><a href="release_notes_10_2_2004.html#release_notes_10_2_2004.last_update_10_2">Last Updated</a></p></li></ul><div class="sec" id="release_notes_10_2_2004.content"><h2><span class="title">Release Notes (October 2, 2004)</span></h2><div class="sec" id="release_notes_10_2_2004.download_loc_10_"><h3><span class="title">Download Location</span></h3><p>
				<a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2004/Oct_02_2004/">ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2004/Oct_02_2004/</a>
			</p></div><div class="sec" id="release_notes_10_2_2004.source_archive_1"><h3><span class="title">Source Archive Contents</span></h3><div class="sec" id="release_notes_10_2_2004.source_code_10_2"><h4><span class="title">Source Code Archives</span></h4><ul><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_unix--Oct_02_2004.tar.gz</span>-- for UNIX'es (see the list of UNIX flavors below) and MacOSX/GCC</p></li><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_unix--Oct_02_2004.gtar.gz</span>-- for UNIX'es (see the list of UNIX flavors below) and MacOSX/GCC</p></li><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_win--Oct_02_2004.exe</span>-- for MS-Windows / MSVC++ 7.1 (self-extracting)</p></li><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_win--Oct_02_2004.zip</span>-- for MS-Windows / MSVC++ 7.1</p></li><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_mac_cw--Oct_02_2004.tgz</span>-- for MacOS 10.3.4 / CodeWarrior DevStudio for MacOS 9.2</p></li><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_mac_xcode--Oct_02_2004.tgz</span>-- for MacOS 10.3.4 / xCode 1.[1-5]</p></li><li><p>
							<span class="nctnt ncbi-path">ncbi_cxx_mac_gcc--Oct_02_2004.tar.gz</span>-- for MacOS 10.2, 10.3.4 / GCC 3.3.2</p></li></ul><p>
					<span class="nctnt highlight">The sources correspond to the NCBI production tree sources from patch "CATCHUP_004", which in turn corresponds to the development tree sources from around the very end of August, 2004.</span>
				</p></div></div><div class="sec" id="release_notes_10_2_2004.new_development_"><h3><span class="title">New Development</span></h3><p>
				<span class="nctnt highlight">Some of the new development that may introduce potentially backward-incompatible changes are marked with an asterisk(*).</span>
			</p><div class="sec" id="release_notes_10_2_2004.corelib_10_2_200"><h4><span class="title">CORELIB -- Portability and Application Framework</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-func">GetVirtualMemoryPageSize()</span>-- new function, to get the granularity with which virtual memory is allocated.</p></dd><dt>2</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CNcbiApplication</span>-- added<span class="nctnt ncbi-func">GetProgramExecutablePath()</span>method.</p></dd><dt>3</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CDllResolver</span>-- extend class to make it also look in "standard" DLL paths. Added<span class="nctnt ncbi-func">AddExtraDllPath()</span>method. Added accessory parameter to<span class="nctnt ncbi-func">FindCandidates()</span>.</p></dd><dt>4</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CDllResolver</span>: added support for driver name (for Plugin Mananger)</p></dd><dt>5</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-func">PluginManager_ConvertRegToTree()</span>- to convert application config file into a parameters tree.</p></dd><dt>6</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CSimpleClassFactoryImpl</span>-- added methods to extract information from parameters tree.</p></dd><dt>7</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CMemoryFile</span>API -- added possibility to map parts of a file.<ul><li><p>
										<span class="nctnt ncbi-class">CMemoryFileSegment</span>- new auxiliary class for mapping a memory file region.</p></li><li><p>
										<span class="nctnt ncbi-class">CMemoryFileMap</span> - new class to support a partial file memory mapping.</p></li><li><p>
										<span class="nctnt ncbi-class">CMemoryFile</span> - now is the same as<span class="nctnt ncbi-class">CMemoryFileMap</span>but have only one big mapped segment with offset 0 and length equal to the length of file.</p></li></ul>
						</p></dd><dt>8</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-func">StringTo*()</span>  -- added additional parameter to specify the action to be performed on conversion error: to throw an exception, or just to return zero.</p></dd><dt>9</dt><dd><p class="no_top_margin">Fixed multithreading on BSD.</p></dd><dt>10</dt><dd><p class="no_top_margin">Added<span class="nctnt ncbi-func">CStopWatch::Restart()</span>to reuse the same timer sequentially.</p></dd><dt>11</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CTree&lt;&gt;</span>-- added new methods:<span class="nctnt ncbi-func">AddNode()</span>,<span class="nctnt ncbi-func">FindNode()</span>,<span class="nctnt ncbi-func">FindSubNode()</span>.</p></dd></dl></div><div class="sec" id="release_notes_10_2_2004.connect_10_2_200"><h4><span class="title">CONNECT -- Data streaming, Networking, and Dispatching</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-lib">LSOCK</span>and<span class="nctnt ncbi-func">CListeningSocket::CListeningSocket()</span>to accept flags (<span class="nctnt ncbi-type">TLSCE_Flags</span>), which extended from flags for simple logging to flags that also control binding of a socket being created:  whether to bind locally (127.0.0.1) or promiscuously (any address) -- as before. Restricted binding helps limit incoming connections to loopback only, thus allowing to create local-only servers (which listen on and reply to localhost only).</p></dd><dt>2</dt><dd><p class="no_top_margin">Threaded Server Framework -- added a mechanism to request graceful termination.</p></dd><dt>3</dt><dd><p class="no_top_margin">Threaded Server Framework -- added support for running user-supplied code if a specified amount of time has occurred since the last client connected.</p></dd></dl></div><div class="sec" id="release_notes_10_2_2004.html_10_2_2004"><h4><span class="title">HTML -- HTML Generation Library</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added class<span class="nctnt ncbi-class">CHTML_password</span>(HTML tag &lt;input type=password&gt;)</p></dd><dt>2</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CHTML_table</span>-- replaced old unimplemented method<span class="nctnt ncbi-func">ColumnWidth()</span>with new one<span class="nctnt ncbi-func">SetColumnWidth()</span>.</p></dd></dl></div><div class="sec" id="release_notes_10_2_2004.util_comp_10_2_2"><h4><span class="title">UTIL/COMPRESSION -- Data Compression (GZIP, BZ2)</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CBZip2Compression</span>, <span class="nctnt ncbi-class">CZipCompression</span>- fixed error in<span class="nctnt ncbi-func">CompressFile()</span>function with using incorrect file stream open mode.</p></dd><dt>2</dt><dd><p class="no_top_margin">Implemented<span class="nctnt ncbi-func">CNlmZipBtRdr::Pushback()</span>for uncompressed data.</p></dd></dl></div><div class="sec" id="release_notes_10_2_2004.serial_10_2_2004"><h4><span class="title">SERIAL -- Data Serialization (ASN.1, XML)</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Completed first step in creation of infrastructure for developing .NET-compatible XML Web services. Created sample XML Web server and client that work on all platforms supported by the Toolkit.</p></dd><dt>2</dt><dd><p class="no_top_margin">Added possibility to scan input stream finding and processing objects of a specific type only, while skipping all other data: see<span class="nctnt ncbi-func">Serial_FilterObjects</span>function template.</p></dd><dt>3</dt><dd><p class="no_top_margin">Added (to "type iterators") the possibility to retrieve context information: type info, item info, object pointer: see<span class="nctnt ncbi-func">CTreeIteratorTmpl&lt;&gt;::GetContext</span>,<span class="nctnt ncbi-func">CTreeIteratorTmpl&lt;&gt;::GetContextData</span>.</p></dd><dt>4</dt><dd><p class="no_top_margin">Modified serial library to support large files (greater than 4GB) on some 32 bits platforms - when compiler supports it.</p></dd><dt>5</dt><dd><p class="no_top_margin">RPC clients (for ID1, Entrez2, etc.) now support adjusting some additional parameters, including in particular timeouts.</p></dd><dt>6</dt><dd><p class="no_top_margin">Added operators to read/write<span class="nctnt ncbi-class">CObjectInfo</span>/<span class="nctnt ncbi-class">CConstObjectInfo</span>.</p></dd></dl></div><div class="sec" id="release_notes_10_2_2004.datatool_10_2_20"><h4><span class="title">DATATOOL -- Code Generator and Data Converter Tool</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Modified code generator to set and return primitive type data by value instead of by reference.</p></dd><dt>2</dt><dd><p class="no_top_margin">Corrected XML schema generation for named integers.</p></dd><dt>3</dt><dd><p class="no_top_margin">Added possibility to specify namespace name when generating XML schema.</p></dd><dt>4</dt><dd><p class="no_top_margin">Added possibility to specify extra headers to include into generated code; such headers are to be defined in DEF file.</p></dd></dl></div><div class="sec" id="release_notes_10_2_2004.cgi_10_2_2004"><h4><span class="title">CGI -- CGI and Fast-CGI Application Framework</span></h4><p>The default implementation of<span class="nctnt ncbi-func">CCgiApplication::OnException()</span>handler to issue a "400" HTTP error if it's the HTTP request itself that seems to be syntactically incorrect. "500" is still issued in all other cases.</p></div><div class="sec" id="release_notes_10_2_2004.berkeley_dbapi_1"><h4><span class="title">Berkeley DB API (bdb) -- Much Enriched C++ API Based On BerkeleyDB</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">New utility BDB dumper, to see BDB database content.</p></dd><dt>2</dt><dd><p class="no_top_margin">Bug fix in BLOB reading writing (cross platform compatibility).</p></dd><dt>3</dt><dd><p class="no_top_margin">Added log cleaning function (<span class="nctnt ncbi-func">CBDB_Env::CleanLog()</span>).</p></dd><dt>4</dt><dd><p class="no_top_margin">Added support for database recovery procedures when opening the Database.</p></dd><dt>5</dt><dd><p class="no_top_margin">Added new data field type "uint1" (unsigned char).</p></dd><dt>6</dt><dd><p class="no_top_margin">BDB_find_field improved to search in non-text fields.</p></dd><dt>7</dt><dd><p class="no_top_margin">BDB Cache(ICache):<ol><li><p>support ASYNCronous write mode</p></li><li><p>improved page size control for better performance tuning</p></li><li><p>more efficient BDB environment join</p></li><li><p>cache verification method added (<span class="nctnt ncbi-func">Verify()</span>)</p></li><li><p>added locking and transaction timeouts</p></li><li><p>added method Remove for key, version, subkey</p></li><li><p>added read-only cache</p></li></ol>
						</p></dd></dl></div><div class="sec" id="release_notes_10_2_2004.dbapi_10_2_2004"><h4><span class="title">DBAPI -- Generic SQL Database Connectivity</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added<span class="nctnt ncbi-class">IReader</span>/<span class="nctnt ncbi-class">IWriter</span>support for BLOB I/O.</p></dd><dt>2</dt><dd><p class="no_top_margin">Added dedicated row counter for rowset results.</p></dd></dl></div><div class="sec" id="release_notes_10_2_2004.algoalign_10_2_2"><h4><span class="title">ALGO/ALIGN -- Generic Alignment Algorithms</span></h4><p>New general purpose tree algorithms:</p><ul><li><p>TreeCompare - compare two trees using comparison functor</p></li><li><p>TreeForEachParent - visit every parent of the tree node</p></li><li><p>TreeTraceToRoot - find root of the tree</p></li><li><p>TreeReRoot - tree re-rooting(rotation)</p></li><li><p>TreeFindCommonParent - find common parent of two nodes</p></li><li><p>TreeListToSet - convert list of nodes to bitset</p></li><li><p>TreeMakeParentsSet - traverse all parents, create set of ids</p></li><li><p>TreeMakeSubNodesSet - create set of subnode ids</p></li><li><p>CTreeNonRedundantSet - compute non-redundant set of nodes</p></li><li><p>CTreeMinimalSet - compute minimal set of nodes</p></li><li><p>Functions to perform logical operations (AND, OR) on nodes</p></li></ul></div><div class="sec" id="release_notes_10_2_2004.algo_phy_10_2_20"><h4><span class="title">ALGO/PHY_TREE -- Phylogenetic and bio tree algorithms</span></h4><p>
					<span class="nctnt ncbi-class">CBioTree</span>- New node design, improved usability.</p></div><div class="sec" id="release_notes_10_2_2004.algoalign_gen_10"><h4><span class="title">ALGO/ALIGN -- Generic Alignment Algorithms</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Supported low-case and zero-based nucleotide sequence characters.</p></dd><dt>2</dt><dd><p class="no_top_margin">Added a capability of parralel computing of parts of dynprog matrix in<span class="nctnt ncbi-class">CNWAligner</span>.</p></dd><dt>3</dt><dd><p class="no_top_margin">Synchronized compartment IDs for transcript alignments calculated in different query strands.</p></dd><dt>4</dt><dd><p class="no_top_margin">Provided a mechanism for elimination of weak/far terminal exons. At the preprocessing step, an empiric was introduced that limits external genomic search space based on the size of non-covered transcript ends. At the post-processing step, a heuristic was added to check for weak terminals.</p></dd></dl></div><div class="sec" id="release_notes_10_2_2004.blast_10_2_2004"><h4><span class="title">BLAST</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Consolidation of the traceback extension types.</p></dd><dt>2</dt><dd><p class="no_top_margin">Implemented stacks initial word container for all blastn extension methods.</p></dd><dt>3</dt><dd><p class="no_top_margin">Implemented uneven gap linking of HSPs for blastx.</p></dd><dt>4</dt><dd><p class="no_top_margin">Refactorings to PSSM engine.</p></dd><dt>5</dt><dd><p class="no_top_margin">Modularized code to link HSPs.</p></dd><dt>6</dt><dd><p class="no_top_margin">MT-support added to<span class="nctnt ncbi-class">CDbBlast</span>.</p></dd><dt>7</dt><dd><p class="no_top_margin">Added implementations of the BlastHspStream interface to collect HSPs according to the traditional BLAST algorithm and to collect all HSPs for on-the-fly processing.</p></dd><dt>8</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CSeqDb</span>: Added support for ISAM indices to allow lookup of sequences by GI, PIG, or String identifier; Allow memory bounds to be set; Added "chunk"  interface to iterate over set of included OIDs.</p></dd><dt>9</dt><dd><p class="no_top_margin">The coordinates remapping for BLAST results when comparing subsequences or sequences from different strands was fixed.</p></dd><dt>10</dt><dd><p class="no_top_margin">Fixed compiler warnings and memory leaks, updated documentation, and removed obsolete defines.</p></dd></dl></div><div class="sec" id="release_notes_10_2_2004.bioobj_10_2_2004"><h4><span class="title">BIO-OBJECTS -- Bio-Object Specific Utility Functions (Not Involving OM++)</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added support for converting generic serial objects to and from User-object trees.</p></dd><dt>2</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CSeq_id</span>-- support several new prefixes in<span class="nctnt ncbi-func">IdentifyAccession()</span>method.</p></dd><dt>3</dt><dd><p class="no_top_margin">Check strand in<span class="nctnt ncbi-func">CSeq_loc::GetStart/End()</span>.</p></dd></dl></div><div class="sec" id="release_notes_10_2_2004.om_10_2_2004"><h4><span class="title">OM++ -- Object Manager -- For Retrieving and Processing Bio-Objects</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">(*)<span class="nctnt ncbi-class">CObjectManager</span>made singleton.</p></dd><dt>2</dt><dd><p class="no_top_margin">(*)<span class="nctnt ncbi-class">CObjectManager</span>-- added methods to manipulate Data Loaders.</p></dd><dt>3</dt><dd><p class="no_top_margin">objmgr_sample.cpp updated to fit current APIs.</p></dd><dt>4</dt><dd><p class="no_top_margin">(*) Seq-id mapper moved from OBJMGR to SEQ library.</p></dd><dt>5</dt><dd><p class="no_top_margin">(*) Removed duplicate methods from<span class="nctnt ncbi-class">CSeqMap</span>.</p></dd><dt>6</dt><dd><p class="no_top_margin">Modified annot iterators to better resolve synonyms. Do not cache synonyms in the scope's history.</p></dd><dt>7</dt><dd><p class="no_top_margin">Detect circular locations in Annot Iterators.</p></dd><dt>8</dt><dd><p class="no_top_margin">Added annotation mapping through annot.locs.</p></dd><dt>9</dt><dd><p class="no_top_margin">Added<span class="nctnt ncbi-func">CScope::GetIds()</span>to fetch synonyms without fetching the whole bioseq.</p></dd><dt>10</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CSeqVector</span>-- added ncbi2na ambiguity randomizer.</p></dd><dt>11</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CBioseq_set_Handle</span>-- added<span class="nctnt ncbi-func">GetComplexityLevel()</span>and<span class="nctnt ncbi-func">GetExactComplexityLevel()</span>.</p></dd><dt>12</dt><dd><p class="no_top_margin">Added operators<span class="nctnt ncbi-func">bool()</span>and<span class="nctnt ncbi-func">!()</span>to all Annot Iterators.</p></dd><dt>13</dt><dd><p class="no_top_margin">Revamped object manager: Changed TSE locking scheme;  and, TSE cache is now maintained by<span class="nctnt ncbi-class">CDataSource</span>.</p></dd><dt>14</dt><dd><p class="no_top_margin">Fixed performance degradation of<span class="nctnt ncbi-func">GetTSESetWithAnnots()</span>, used time was quadratic on amount of loaded blobs.</p></dd><dt>15</dt><dd><p class="no_top_margin">Fixed processing of trace assemblies: segment shift in Seq-annot.locs processing, and duplication of discontinuous alignments.</p></dd><dt>16</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CAnnotName</span>and<span class="nctnt ncbi-class">CAnnotTypeSelector</span>moved to separate headers.</p></dd><dt>17</dt><dd><p class="no_top_margin">Several fixes to make object manager thread-safe.</p></dd><dt>18</dt><dd><p class="no_top_margin">Understand "weight" param in qual field and string dbSNP value when parsing SNP features into table. Allow SNP Seq-entry in addition to SNP Seq-annot.</p></dd><dt>19</dt><dd><p class="no_top_margin">Added delayed loading of external annotations from satellite 26.</p></dd><dt>20</dt><dd><p class="no_top_margin">Fixed loading of missing split chunks when<span class="nctnt ncbi-func">GetCompleteXxx()</span>method is called.</p></dd><dt>21</dt><dd><p class="no_top_margin">Added splitting of whole Bioseqs. Postpone indexing of retrieved split blobs to reduce memory usage by big nucprot sets.</p></dd></dl></div><div class="sec" id="release_notes_10_2_2004.omlr_10_2_2004"><h4><span class="title">OM++ LOADERS/READERS -- Data Retrieval Libraries for OM++</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">(*)Data Loaders do not have public constructors anymore,<span class="nctnt ncbi-func">RegisterInObjectManager()</span>should be used instead.</p></dd><dt>2</dt><dd><p class="no_top_margin">(*)Added Data Loader factories using<span class="nctnt ncbi-class">CPluginManager</span>, fixed driver names.</p></dd><dt>3</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CSeqref</span>is replaced by<span class="nctnt ncbi-class">CBlobId</span>.</p></dd><dt>4</dt><dd><p class="no_top_margin">Added loading of external annotations from 26 satellite. Use new split features to represent external annotation blobs (sat 26).</p></dd><dt>5</dt><dd><p class="no_top_margin">Allow SNP Seq-entry in addition to SNP Seq-annot. Load SNPs from satellite 26.</p></dd></dl></div><div class="sec" id="release_notes_10_2_2004.id2_10_2_2004"><h4><span class="title">ID2</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added GENNBANK_READER_ID2_xxx macros to Makefile.mk.</p></dd><dt>2</dt><dd><p class="no_top_margin">Added exports for ID2 libraries. ID1 &amp; ID2 moved to ncbi_seqext.dll on Windows.</p></dd></dl></div><div class="sec" id="release_notes_10_2_2004.biotools_10_2_20"><h4><span class="title">BIO-TOOLS</span></h4><p>Added preliminary support for importing and exporting GFF version 3 data, complete with gapped alignments.</p></div><div class="sec" id="release_notes_10_2_2004.bfmsvc_10_2_2004"><h4><span class="title">BUILD FRAMEWORK (MSVC++ 6.0)</span></h4><p>Not supported anymore.</p></div><div class="sec" id="release_notes_10_2_2004.bfdotnet_10_2_20"><h4><span class="title">BUILD FRAMEWORK (MSVC++ .NET)</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Allow for conditional macro definition.</p></dd><dt>2</dt><dd><p class="no_top_margin">Implemented optional dependency on a third-party library.</p></dd><dt>3</dt><dd><p class="no_top_margin">Do include EXPENDABLE projects to the build.</p></dd><dt>4</dt><dd><p class="no_top_margin">Added possibility to set up a precompiled header for a subdirectory which is not direct child of the root folder.</p></dd><dt>5</dt><dd><p class="no_top_margin">Corrected new_project and import_project Windows scripts to add  more diagnostics, and to add the ability to import projects from the non-standard build tree.</p></dd></dl></div></div><div class="sec" id="release_notes_10_2_2004.apps_10_2_2004"><h3><span class="title">APPLICATIONS</span></h3><div class="sec" id="release_notes_10_2_2004.apps_objmgr_demo"><h4><span class="title">OBJMGR DEMO program (objmgr_demo)</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Allow limited processing to have benefits from split blobs. Added new options to control tasks to be done:<ul><li><p>count_types</p></li><li><p>reset_scope</p></li><li><p>limit_tse</p></li><li><p>used_memory_check</p></li><li><p>print_cds</p></li><li><p>range_from</p></li><li><p>range_to</p></li></ul>
						</p></dd><dt>2</dt><dd><p class="no_top_margin">Use<span class="nctnt ncbi-class">CBlob_id</span>instead of obsolete<span class="nctnt ncbi-class">CSeqref</span>. Removed obsolete code for old blob cache interface.</p></dd><dt>3</dt><dd><p class="no_top_margin">Seq-id cache is put in the same directory as blob cache. Tuned cache parameters on Windows.</p></dd><dt>4</dt><dd><p class="no_top_margin">Optional compilation with Berkley DB.</p></dd></dl></div><div class="sec" id="release_notes_10_2_2004.apps_objmgr_test"><h4><span class="title">OBJMGR TEST programs (test_objmgr_data &amp; test_objmgr_data_mt)</span></h4><p>Added new options -no_seq_map, -no_named, -verbose, -adaptive -idlist.</p></div></div><div class="sec" id="release_notes_10_2_2004.documentation_10"><h3><span class="title">Documentation</span></h3><div class="sec" id="release_notes_10_2_2004.doc_loc_10_2_200"><h4><span class="title">Document Location</span></h4><p>The documentation is available online at<!--ro--><a href="toc.html">http://www.ncbi.nlm.nih.gov/books/bv.fcgi?call=bv.View..ShowTOC&amp;rid=toolkit.TOC&amp;depth=2</a>as a book titled "The NCBI C++ Toolkit". This is an online searchable book.</p><p>The C++ Toolkit book also provides PDF version of the chapters; although these are not up to date in this release. The PDF version can be accessed by a link that appears on each page.</p><p>The older HTML documentation has been deprecated and is no longer being updated, and  "The NCBI C++ Toolkit" online book at the previously listed URLs is the official documentation.</p></div><div class="sec" id="release_notes_10_2_2004.doc_changes_10_2"><h4><span class="title">Document Content</span></h4><p>Documentation has been grouped into chapters and sections that provide a more logical coherence and flow. New sections and paragraphs continue to be added to update and clarify the older documentation or provide new documentation. The chapter titled "Introduction to the C++ Toolkit"  gives an overview of the C++ Toolkit. This chapter contains links to other chapters containing more details on a specific topic and is a good starting point for the new comer.</p><p>The DOXYGEN source browser is used to complement the traditional docs with structured DOXYGEN-generated "Library Reference" ones based on the in-source comments. You can access the DOXYGEN source browser for the NCBI code from:</p><p>
					<a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/">http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/</a>
				</p><p>The above link is also available under the "Browsers" that appears on each page.</p><p>You can also access the CVS code repository via a Web interface. These links also appear in the sidebar box on each page.</p><p>A C/C++ Symbol Search query appears on each page of the online Toolkit documentation. You can use this to perform a symbol search on the public or in-house versions of LXR, Doxygen and Library. The Library search runs a CGI script that lists the library name where the symbol (such as function) is defined in.</p><p>The current release notes as well as past release notes are now in an appendix in the C++ Toolkit Book and a link to the current release notes appears on each page of the online Toolkit document.</p></div></div><div class="sec" id="release_notes_10_2_2004.build_macos_10_2"><h3><span class="title">Building on the MacOS</span></h3><p>We now build all the libraries and most of the applications including the Genome Workbench (<span class="nctnt ncbi-app">gbench</span>).  We do not build any of the many test or demo apps (except<span class="nctnt ncbi-app">testvalidator</span>).  We also build the FLTK library's GUI editor,<span class="nctnt ncbi-app">fluid</span>.</p><p>All apps are built as application bundles except<span class="nctnt ncbi-app">gbench_plugin_scan</span>and<span class="nctnt ncbi-app">datatool</span>which are built as command line apps.  Any of the applications can be built as command line apps by tweaking the build scripts or the Codewarrior projects.</p><p>Build procedure is still the same: use an AppleScript editor to open and run the script files makeLibs.met and makeApps.met.  You must use a script editor capable of opening a script file larger than 32K, such as Apple's Script Editor v2.0, or Smile.  Script Editor v1.9 will not work since makeLibs.met just got too big.  The command line tool osascript also works.</p><p>Projects include targets to compile with BSD/Apple headers and libraries and with MSL headers and libraries, but plugins (for gbench) built with MSL do not link properly, and so, because of lack of interest to keep up with changes, some source code does not currently  compile with MSL.  Hopefully this will become easier to work with and get fixed with Codewarrior v.9.</p><p>Targets to be compiled can be controlled by including an empty file or folder in the compilers:mac_prj folder with the name 'Build' followed by the keywords of the targets you want built.  The keywords are: MSL, BSD, Debug and Final.  For example, to build only the BSD Debug targets use: "Build BSD Debug", to build both BSD debug and release (final) versions: "Build BSD".  The default is to build everything.</p><p>If you install the C++ toolkit under a different name than "ncbi_cxx" or in a different location than your home directory, you can edit the  script's properties, pRootFolderName and pRootFolderPath, to override these defaults.  Note: these paths, and those mentioned below, must be entered in mac format (e.g. disk:Users:username:) not Unix format (e.g.<span class="nctnt ncbi-path">/Users/username/</span>).  The disk name (and its following colon) may be omitted.</p><p>Certain third party libraries (see<a href="release_notes_10_2_2004.html#release_notes_10_2_2004.T1">Table 14</a>) are required to build some parts of the C++ toolkit. The scripts will try and find them if they are in your home directory, or you can specify where they were installed using properties at the beginning of the script.</p><div class="table" id="release_notes_10_2_2004.T1"><h4><span class="label">Table 14</span><span class="title">Third Party Libraries</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Library</th><th align="left" valign="top" rowspan="1" colspan="1">Property</th><th align="left" valign="top" rowspan="1" colspan="1">Example</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-lib">FLTK 1.1.5rc2 (w/ NCBI patches)</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-var">pFLTKRootFolder</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">"usr:local:src:fltk-1.1.5rc2:"</span>
							</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-lib">BerkeleyDB 4.2.52</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-var">pBdbRootFolder</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">"Users:myhome:mylibs:db-4.2.52"</span>
							</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-lib">SQLite 2.8.15</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-var">gSqliteFolder</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">

</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-lib">dlcompat</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-var">pDLRootFolder</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">"usr:"</span>
							</td></tr></tbody></table></div></div><p>You do not have to build the FLTK or BDB libraries separately.  This is done by the scripts and Codewarrior along with the toolkit libraries.  Just unpack the source bundles in your home directory or where ever you have specified in the appropriate properties.  The root folders for FLTK and BDB do not have to have any particular names.  If there is more than one version the scripts will grab whichever is last alphabetically (e.g. fltk-1.1.4r2 will get used instead of fltk-1.1.3).</p><p>The scripts normally halt on any Codewarrior compilation errors.  If you want them to continue and save errors, set the next script property, pSaveContinueOnErrors, to true.  Compilation errors for a project will be saved in a file in the same folder as the project being built, with a name in the following format: projectName-targetNumber.errs (e.g. xncbi-2.errs).</p><p>The Genome Workbench's configuration file(s) is stored in the users Library:Application Support:gbench folder.</p><p>CFM builds are not supported.  OS 8 or 9 are not supported.  We know 10.2 works.  We think 10.1 still works, and 10.3 might work.</p></div><div class="sec" id="release_notes_10_2_2004.platforms_10_2_2"><h3><span class="title">Platforms (OS's, compilers used inside NCBI)</span></h3><p>This release was successfully tested on the following platforms -- but may also work on other platforms. Since the previous release, some platforms were dropped from this list, just because we do not use them here anymore, and some were added (these new platforms are highlighted using bold font). Also, it can happen that some projects would not work (or even compile) in the absence of 3rd-party packages, or with older or newer versions of such packages -- in these cases, just skipping such projects (e.g. using flag "-k" for<span class="nctnt ncbi-app">make</span>on UNIX), can get you through.</p><div class="sec" id="release_notes_10_2_2004.platform_unix_10"><h4><span class="title">Unix</span></h4><div class="table" id="release_notes_10_2_2004.T2"><h4><span class="label">Table 15</span><span class="title">Unix OS's and Supported Compilers</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System</th><th align="left" valign="top" rowspan="1" colspan="1">Architecture</th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Linux-2.4.23 (w/ LIBC 2.2.5)</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">INTEL</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 2.95.3, 3.0.4, 3.4.0</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Linux-2.4.26 (w/ LIBC 2.3.2)</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">INTEL</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 3.4.0</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Linux-2.6.7 (w/ LIBC 2.3.3)</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">INTEL/<b>64</b></span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 3.4.0</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Linux-2.4.23 (w/ LIBC 2.2.5)</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">INTEL</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">ICC <b>8.0</b></span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Solaris-8</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">SPARC</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">WorkShop 6 update 2 C++ 5.3 Patch 111685-13  (64-bit, 32-bit)</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Solaris-8</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">SPARC</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 3.0.4</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Solaris-<b>9</b></span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">INTEL</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">WorkShop 6 update 2 C++ 5.3 Patch 111685-13</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Solaris-<b>9</b></span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">INTEL</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC <b>3.3.3</b></span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">IRIX64-6.5</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">SGI-Mips</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">MIPSpro 7.3.1.2m (64-bit, 32-bit)</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">FreeBSD-4.5</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">INTEL</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 3.0.4</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">FreeBSD-4.10</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">INTEL</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC <b>3.4.2</b></span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Tru64 (OSF1) V5.1</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">ALPHA</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 3.3.2</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">Tru64 (OSF1) V5.1</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">ALPHA</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">Compaq C++ V6.5-014</span>
								</td></tr></tbody></table></div></div></div><div class="sec" id="release_notes_10_2_2004.platform_windows"><h4><span class="title">MS Windows</span></h4><div class="table" id="release_notes_10_2_2004.T3"><h4><span class="label">Table 16</span><span class="title">MS Windows and Supported Compilers</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System</th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">MS Windows</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">MSVC++ 6.0 Service Pack 5</span>.  To be DEPRECATED -- this is the last release where it is supported.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">MS Windows</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">MSVC++<b>7.1</b></span>. See documentation on<a href="ch_config.html#ch_config.MS_Windows">MS Visual C++.NET</a>.</td></tr></tbody></table></div></div></div><div class="sec" id="release_notes_10_2_2004.platform_mac_10_"><h4><span class="title">Mac OS X</span></h4><div class="table" id="release_notes_10_2_2004.T4"><h4><span class="label">Table 17</span><span class="title">Mac OS, and Supported Compilers</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System</th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">MacOS 10.2, 10.3.4</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">GCC 3.3</span>
								</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-monospace">MacOS 10.3.4</span>
								</td><td align="left" valign="top" rowspan="1" colspan="1">
									<span class="nctnt ncbi-app">CodeWarrior 9.2</span>
								</td></tr></tbody></table></div></div></div></div><div class="sec" id="release_notes_10_2_2004.caveats_10_2_200"><h3><span class="title">Caveats and Hints</span></h3><div class="sec" id="release_notes_10_2_2004.macos_10x_10_2_2"><h4><span class="title">MacOS 10.X / CodeWarrior 9.1</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Not all of the test or demo applications are built.</p></dd><dt>2</dt><dd><p class="no_top_margin">The source code for the latest release of FLTK (1.1.x), BerkeleyDB (4.x) and SQLite (2.x) should be present. See the  installation instructions for details.</p></dd></dl></div><div class="sec" id="release_notes_10_2_2004.macos_10_X_10_2_"><h4><span class="title">MacOS 10.2/GCC 3.3</span></h4><p>At least the GCC 3.3 update for Dec. 2002 Developers Tools required from Apple.</p></div><div class="sec" id="release_notes_10_2_2004.gcc_2_95_3_10_2_"><h4><span class="title">GCC 2.95</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Poor MT-safety record.</p></dd><dt>2</dt><dd><p class="no_top_margin">Relatively incomplete/incorrect (comparing to modern compilers) STL implementation.</p></dd><dt>3</dt><dd><p class="no_top_margin">It is going to be deprecated in NCBI rather soon -- as soon as we have any significant trouble with its maintenance.</p></dd></dl></div><div class="sec" id="release_notes_10_2_2004.gcc_3_0_4_10_2_2"><h4><span class="title">GCC 3.0.4</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Destructor of constructed class member is not called when exception is thrown from a method called from class constructor body (fixed in GCC 3.3).</p></dd><dt>2</dt><dd><p class="no_top_margin">STL stream uses locale in thread unsafe way which may result to segmentation fault when run in multithread mode (fixed in GCC 3.3).</p></dd><dt>3</dt><dd><p class="no_top_margin">Long-file support for C++ streams is disabled/broken (first broken in 3.0, fixed in 3.4).</p></dd></dl></div><div class="sec" id="release_notes_10_2_2004.gcc_3_3_10_2_200"><h4><span class="title">GCC 3.3</span></h4><p>Other than the feature described below, GCC 3.3.2 has been very good for us; it has a lot of very ugly bugs finally fixed.</p><p>Painfully slow linking in debug mode on Linux with GCC-3.3 compiler:<dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Starting with binutils 2.12 linker tries to merge constant/debug strings marked for merging in object files. But it seems it does this job very inefficiently - I've seen messages about it in internet.</p><p>GCC starting with version 3.2 marks section of string constants ready for merging, and also has an option to disable this flag in object files (-fno-merge-constants). Adding this flag to compilation stage allows to avoid slow linking.</p><p>GCC 3.3 also sets merge flag for debug sections and unfortunately there is no option to disable this flag. As a result, linking of debug executables significatly slower than with gcc 3.0.4.</p><p>The slowdown rate depends on size of debug strings section and it's non-linear, so bigger projects will suffer more of this bug.</p><p>We had to patch GCC 3.3 in-house with the fix described at<a href="http://lists.boost.org/MailArchives/boost/msg53004.php">http://lists.boost.org/MailArchives/boost/msg53004.php</a>.</p></dd><dt>2</dt><dd><p class="no_top_margin">Long-file support still broken (first broken in 3.1).</p></dd></dl>
				</p></div><div class="sec" id="release_notes_10_2_2004.gcc_3_4_10_2_200"><h4><span class="title">GCC 3.4</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">The "Painfully slow linking..." (see GCC3.3 [1] above) is still an issue. -- Again, had to patch it in-house to speed it up, a la GCC 3.3.</p></dd><dt>2</dt><dd><p class="no_top_margin">At least on Linux, ifstream::readsome() does not always work for large files, as it calls an ioctl  that doesn't work properly for large files.</p></dd><dt>3</dt><dd><p class="no_top_margin">At least on Linux, GCC 3.4.[1,2] optimizer (very rarely) generates incorrect code when comparing enumerated values in else-ifs. (Fixed in 3.4.2)</p></dd></dl></div></div><div class="sec" id="release_notes_10_2_2004.last_update_10_2"><h3><span class="title">Last Updated</span></h3><p>This section last updated on October 22, 2004</p></div></div></div></div><div id="tkb_internal" class="col four_col last"><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" id="source-branding"><div class="iconblock clearfix ten_col"><a class="img_link icnblk_img" title="Table of Contents Page" target="mainwindow" href="toc.html"><img class="source-thumb" src="img/th-toolkit-lrg.png" alt="Cover of The NCBI C++ Toolkit Book" height="100px" width="80px" /></a><div class="icnblk_cntnt"><div class="_bk_pgbnr_cit"><div class="_bk_pgbnr_cit_ln">The NCBI C++ Toolkit Book [Internet].</div><div class="_bk_pgbnr_cit_ln">Vakatov D, editor.</div><div class="_bk_pgbnr_cit_ln">Bethesda (MD): <a href="http://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (US)</a>; 2004-.</div></div></div></div></div><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" class="half_rhythm"><ul class="inline_list"><li><a href="toc.html">Table of Contents Page</a></li><li> | </li><li><a class="jig-ncbidialog" data-jigconfig="width:400 , modal:true" href="#citation-window">Cite this Page</a><div class="jig-ncbidialog-box" style="display:none" title="Cite this Page" id="citation-window"><div class="bk_tt">
        Vakatov D, editor. The NCBI C++ Toolkit Book [Internet]. Bethesda (MD): National Center for Biotechnology Information (US); 2004-.  , Release Notes (October 2, 2004)
                    . .

                            <div class="bk_tt"><span class="bk_cite_avail">Available from: http://www.ncbi.nlm.nih.gov/toolkit/doc/book/release_notes_10_2_2004</span></div></div><span class="bk_cite_avail"></span></div></li></ul></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Search</h3></div></div><div class="portlet_content"><form name="frmSymbolSearch" onsubmit="return SymbolSearch('toolkit')" action="/" id="frmSymbolSearch"><table><tbody><tr align="center" valign="middle"><td colspan="2" nowrap="nowrap"><input id="search_box" name="__symbol" style="vertical-align:                                        middle; line-height:normal " onkeypress="SymbolSearchKeyPress('toolkit',event)" type="text" />  					<label for="search_box" class="offscreen_noflow">Search term</label><button type="submit" value="Search" style="vertical-align:                                        middle; line-height:normal;" onclick="javasrcipt:SymbolSearch('toolkit');">Search</button></td></tr><tr class="internal-only"><th style="text-align:left">Public</th><th style="text-align:left">Internal</th></tr><tr><td><input id="pToolkitAll" name="__symboloc" type="radio" /><label for="pToolkitAll" title="Search in the C/C++ source code and in this Book">All</label></td><td class="internal-only"><input id="iToolkitAll" name="__symboloc" type="radio" checked="checked" /><label for="iToolkitAll" title="Search in the C/C++ source code,  this Book, and Wiki">All</label></td></tr><tr><td><input id="pLXR" name="__symboloc" type="radio" /><label for="pLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td><td class="internal-only"><input id="iLXR" name="__symboloc" type="radio" /><label for="iLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td></tr><tr><td><input id="pLib" name="__symboloc" type="radio" /><label for="pLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td><td class="internal-only"><input id="iLib" name="__symboloc" type="radio" /><label for="iLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td></tr></tbody></table></form></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Source Browsers</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><span class="bkp_prtlt_txt"> LXR: </span><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a><span class="bkp_prtlt_txt"> / </span><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/lxr/ident" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Doxygen: </span><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a><span class="bkp_prtlt_txt"> / </span><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/doxyhtml" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>SVN Source Repository</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><span class="bkp_prtlt_txt"> Documentation: </span><a href="ch_getcode_svn.html#ch_getcode_svn.code_retrieval">in-house</a><span class="bkp_prtlt_txt"> / </span><a href="ch_getcode_svn.html#ch_getcode_svn.external">public</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (dev): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/c++" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a><span class="bkp_prtlt_txt"> / </span><a href="http://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/c++/" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (extra): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/internal/c++" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (prod): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/production/components" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Resources</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/lib_search/libsearch.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Search Tool</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/depgraphs/dglistdeps.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Dependencies Tool</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/wiki-private/CxxToolkit" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Wiki</a></li><li class="internal-only"><a href="https://jira.ncbi.nlm.nih.gov/browse/CXX" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">JIRA</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/STAT/test_stat/test_stat_ext.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="View results of testsuite runs on the automatic C++ Toolkit builds">Testsuite</span></a></li><li><a href="release_notes.html">Release Notes</a></li><li><a href="ch_getcode_svn.html#ch_getcode_svn.ftp_download">Download the Toolkit</a></li><li><a href="http://www.ncbi.nlm.nih.gov/toolkit/doc/book/pdf/TOC.pdf">PDF of the book</a></li><li><a href="ch_style.html">Coding Style</a></li><li><a href="ch_faq.html#ch_faq.mailing_lists">Mailing Lists</a></li><li><a href="mailto:CPP-CORE@NCBI.NLM.NIH.GOV" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="Contact C++ Toolkit group">Help and Support</span></a></li></ul></div></div></div><div id="tkb_public" class="col four_col last"><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" id="source-branding"><div class="iconblock clearfix ten_col"><a class="img_link icnblk_img" title="Table of Contents Page" target="mainwindow" href="toc.html"><img class="source-thumb" src="img/th-toolkit-lrg.png" alt="Cover of The NCBI C++ Toolkit Book" height="100px" width="80px" /></a><div class="icnblk_cntnt"><div class="_bk_pgbnr_cit"><div class="_bk_pgbnr_cit_ln">The NCBI C++ Toolkit Book [Internet].</div><div class="_bk_pgbnr_cit_ln">Vakatov D, editor.</div><div class="_bk_pgbnr_cit_ln">Bethesda (MD): <a href="http://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (US)</a>; 2004-.</div></div></div></div></div><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" class="half_rhythm"><ul class="inline_list"><li><a href="toc.html">Table of Contents Page</a></li><li> | </li><li><a class="jig-ncbidialog" data-jigconfig="width:400 , modal:true" href="#citation-window">Cite this Page</a><div class="jig-ncbidialog-box" style="display:none" title="Cite this Page" id="citation-window"><div class="bk_tt">
        Vakatov D, editor. The NCBI C++ Toolkit Book [Internet]. Bethesda (MD): National Center for Biotechnology Information (US); 2004-.  , Release Notes (October 2, 2004)
                    . .

                            <div class="bk_tt"><span class="bk_cite_avail">Available from: http://www.ncbi.nlm.nih.gov/toolkit/doc/book/release_notes_10_2_2004</span></div></div><span class="bk_cite_avail"></span></div></li></ul></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Search</h3></div></div><div class="portlet_content"><form name="frmSymbolSearch" onsubmit="return SymbolSearch('toolkit')" action="/" id="frmSymbolSearch"><table><tbody><tr align="center" valign="middle"><td colspan="2" nowrap="nowrap"><input id="search_box" name="__symbol" style="vertical-align:                                        middle; line-height:normal " onkeypress="SymbolSearchKeyPress('toolkit',event)" type="text" />  					<label for="search_box" class="offscreen_noflow">Search term</label><button type="submit" value="Search" style="vertical-align:                                        middle; line-height:normal;" onclick="javasrcipt:SymbolSearch('toolkit');">Search</button></td></tr><tr><td colspan="2"><input id="pToolkitAll" name="__symboloc" type="radio" checked="checked" /><label for="pToolkitAll" title="Search in the C/C++ source code and in this Book">All</label></td></tr><tr><td colspan="2"><input id="pLXR" name="__symboloc" type="radio" /><label for="pLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td></tr><tr><td colspan="2"><input id="pLib" name="__symboloc" type="radio" /><label for="pLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td></tr></tbody></table></form></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Source Browsers</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">LXR</a></li><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">Doxygen</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>SVN Source Repository</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="ch_getcode_svn.html#ch_getcode_svn.external">Documentation</a></li><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/c++/" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Web (dev)</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Resources</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lib_search/libsearch.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Search Tool</a></li><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/depgraphs/dglistdeps.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Dependencies Tool</a></li><li><a href="release_notes.html">Release Notes</a></li><li><a href="ch_getcode_svn.html#ch_getcode_svn.ftp_download">Download</a></li><li><a href="http://www.ncbi.nlm.nih.gov/toolkit/doc/book/pdf/TOC.pdf">PDF of the book</a></li><li><a href="ch_style.html">Coding Style</a></li><li><a href="ch_faq.html#ch_faq.mailing_lists">Mailing Lists</a></li><li><a href="mailto:CPP-GROUP@NCBI.NLM.NIH.GOV" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="Contact C++ Toolkit group">Help and Support</span></a></li></ul></div></div></div></div></div></div></body></html>
