<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><title>Release Notes (April 30, 2006) - The NCBI C++ Toolkit Book [Internet] - NCBI</title><meta name="description" content="A comprehensive manual on the NCBI C++ toolkit, including its design and development framework, a C++ library reference, software examples and demos, FAQs and release notes." /><meta name="robots" content="index,nofollow,noarchive" /><meta name="page_id" content="release_notes_04_30_2006" /><meta name="page_type" content="chapter" /><meta name="ncbipopper" content="triggerPosition:'bottom left',destPosition:'top left',openEvent:'click',closeEvent:'click'" /><script type="text/javascript" src="http://www.ncbi.nlm.nih.gov/core/jig/1.13/js/jig.min.js "></script><link type="text/css" rel="stylesheet" href="css/ptbooks.css" /><link type="text/css" rel="stylesheet" href="css/toolkit.css" /></head><body><div class="grid no_max_width"><div class="col twelve_col nomargin shadow"><div class="content"><div class="col eight_col"><div class="nvgtn"><div class="pagination"><a class="pagelink active prev" title="Previous page" href="release_notes_08_14_2006.html">&lt; Prev</a><a class="pagelink active next" title="Next page" href="release_notes_12_31_2005.html">Next &gt;</a></div><div class="pdf-link"><a class="print-link" title="PDF version of this page" href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/public_releases/pdf/release_notes_04_30_2006.pdf">PDF</a></div></div><div class="meta_content"><h1><span class="title">Release Notes (April 30, 2006)</span></h1><p class="small">Created: April 30, 2006; Last Update: May 2, 2006.</p></div><div class="body-content jig-ncbiinpagenav" data-jigconfig="allHeadingLevels: ['h2'],smoothScroll: false"><ul><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.Download_Location">Download Location</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.Source_Archive_Conte">Source Archive Contents</a></p><ul><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.Source_Code_Archives">Source Code Archives</a></p></li></ul></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.New_Development">New Development</a></p><ul><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.PLATFORMS_AND_CONFIG">PLATFORMS AND CONFIGURATIONS</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.CORELIB__Portability">CORELIB — Portability and Application Framework</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.CONNECT__Data_stream">CONNECT — Data streaming, Networking, and Dispatching</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.UTIL__Miscellaneous_">UTIL — Miscellaneous Low-Level APIs</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.SERIAL__Data_Seriali">SERIAL — Data Serialization (ASN.1, XML)</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.CGI__CGI_and_FastCGI">CGI — CGI and Fast-CGI Application Framework</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.HTML__HTML_Generatio">HTML — HTML Generation Library</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.BerkeleyDB_API_bdb__">BDB — Yet Another C++ API Based On BerkeleyDB</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.Local_data_storage">Local data storage</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.DBAPI__SQL_Database_">DBAPI — Generic SQL Database Connectivity</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.ALGO_ALIGN_SPLIGN__S">ALGO/ALIGN/SPLIGN — Spliced Alignment Algorithms</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.ALNMGR__Biosequence_">ALNMGR — Bio-sequence Alignment Manager</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.BLAST">BLAST</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.BIOOBJECTS__BioObjec">BIO-OBJECTS — Bio-Object Specific Utility Functions (Not Involving OM++)</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.OM__Object_Manager__">OM++ — Object Manager — For Retrieving and Processing Bio-Objects</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.OM_LOADERS_READERS__">OM++ LOADERS/READERS — Data Retrieval Libraries for OM++</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.OM_DEMO_program_objm">OM++ DEMO program (objmgr_demo)</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.OM_DEMO_program_objm">BUILD FRAMEWORK (UNIX)</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.PTB__Project_Tree_Bu">PTB — Project Tree Builder for MSVC++ .NET</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.APPLICATIONS">APPLICATIONS</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.GRID_DISTRIBUTED_COM">GRID (DISTRIBUTED COMPUTING) FRAMEWORK</a></p></li></ul></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.Documentation">Documentation</a></p><ul><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.Document_Location">Document Location</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.Document_Content">Document Content</a></p></li></ul></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.Building_on_the_MacO">Building on the MacOS</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.Platforms_OSs__compi">Platforms (OS's, compilers used inside NCBI)</a></p><ul><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.T1">Unix</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.T2">MS Windows</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.T3">Mac OS X</a></p></li></ul></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.Caveats_and_Hints">Caveats and Hints</a></p><ul><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.GCC_295">GCC 2.95</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.GCC_304">GCC 3.0.4</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.GCC_33">GCC 3.3</a></p></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.GCC_34x__40x">GCC 3.4.x, 4.0.x</a></p></li></ul></li><li><p><a href="release_notes_04_30_2006.html#release_notes_04_30_2006.Last_Updated">Last Updated</a></p></li></ul><div class="sec" id="release_notes_04_30_2006.Download_Location"><h2><span class="title">Download Location</span></h2><p>
			<a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2006/Apr_30_2006">ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2006/Apr_30_2006/</a>
		</p></div><div class="sec" id="release_notes_04_30_2006.Source_Archive_Conte"><h2><span class="title">Source Archive Contents</span></h2><div class="sec" id="release_notes_04_30_2006.Source_Code_Archives"><h3><span class="title">Source Code Archives</span></h3><ul><li><p>
						<a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2006/Apr_30_2006/ncbi_cxx--Apr_30_2006.tar.gz"><span class="nctnt ncbi-path">ncbi_cxx--Apr_30_2006.tar.gz</span></a> — for UNIX'es (see the list of UNIX flavors below) and MacOSX/GCC</p></li><li><p>
						<a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2006/Apr_30_2006/ncbi_cxx--Apr_30_2006.gtar.gz"><span class="nctnt ncbi-path">ncbi_cxx--Apr_30_2006.gtar.gz</span></a>  for UNIX'es (see the list of UNIX flavors below) and MacOSX/GCC</p></li><li><p>
						<a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2006/Apr_30_2006/ncbi_cxx--Apr_30_2006.exe"><span class="nctnt ncbi-path">ncbi_cxx--Apr_30_2006.exe</span></a>  — for MS-Windows (32- and 64-bit)  / MSVC++ (7.1, 8.0) — self-extracting</p></li><li><p>
						<a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2006/Apr_30_2006/ncbi_cxx--Apr_30_2006.zip"><span class="nctnt ncbi-path">ncbi_cxx--Apr_30_2006.zip</span></a> — for MS-Windows (32- and 64-bit) / MSVC++ (7.1, 8.0) </p></li><li><p>
						<a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2006/Apr_30_2006/ncbi_cxx_mac_xcode--Apr_30_2006.gtar.gz"><span class="nctnt ncbi-path">ncbi_cxx_mac_xcode--Apr_30_2006.gtar.gz</span></a> — for MacOSX 10.4 / Xcode 1.5-2.2.1</p></li></ul><p>
				<span class="nctnt ncbi-monospace">The sources correspond to the NCBI </span>
				production tree
				<span class="nctnt ncbi-monospace"> sources from patch "CATCHUP_APR_2006", which in turn roughly corresponds to the </span>
				development tree
				<span class="nctnt ncbi-monospace"> sources from the very beginning of April, 2006.</span>
			</p><p>
				<span class="nctnt ncbi-monospace">There are also two sub-directories, containing easily buildable source distributives of the NCBI C Toolkit (for MS Windows and UNIX) and selected 3rd-party packages (for MS Windows only). These are the versions that the NCBI C++ Toolkit should build with. For build instructions, see README files there:</span>
			</p><ul><li><p>
						<a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2006/Apr_30_2006/NCBI_C_Toolkit/README"><span class="nctnt ncbi-path">NCBI_C_Toolkit </span></a>
					</p></li><li><p>
						<a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2006/Apr_30_2006/ThirdParty/README"><span class="nctnt ncbi-path">ThirdParty</span></a>
					</p></li></ul></div></div><div class="sec" id="release_notes_04_30_2006.New_Development"><h2><span class="title">New Development</span></h2><div class="sec" id="release_notes_04_30_2006.PLATFORMS_AND_CONFIG"><h3><span class="title">PLATFORMS AND CONFIGURATIONS</span></h3><dl class="labeled-list"><dt>A</dt><dd><p class="no_top_margin">Newly (re)supported:</p><p><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
							MS Visual Studio 2005 (both 32- and 64-bit).</p></dd><dt>2</dt><dd><p class="no_top_margin">
							Mac OS X 10.4 (Tiger).</p></dd><dt>3</dt><dd><p class="no_top_margin">
							GCC/Cygwin.
  <b>  NOTE:</b> As of now, it supports static builds only, DLLs are not buildable.</p></dd><dt>4</dt><dd><p class="no_top_margin">
							GCC with GNU ld on Solaris 10 (as long as the patch from
    <a href="http://sourceware.org/bugzilla/show_bug.cgi?id=1031">http://sourceware.org/bugzilla/show_bug.cgi?id=1031</a> has been applied)</p></dd></dl></p></dd><dt>B</dt><dd><p class="no_top_margin">Discontinued:</p><p><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">CodeWarrior on MacOSX</p></dd></dl></p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.CORELIB__Portability"><h3><span class="title">CORELIB — Portability and Application Framework</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CDirEntry::SplitPathEx()</span> — new function, to split a path string into its basic components (not OS-specific version).</p></dd><dt>2</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">CDir::SetCwd()</span> — new function, to change the current working directory.</p></dd><dt>3</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">CFileUtil::GetFileSystemInfo()</span> — new function.</p></dd><dt>4</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">CMemoryFile</span> — new method.added constructor parameters for automatic creating/extention of the mapped file. Added new <span class="nctnt ncbi-func">Extend()</span> method.</p></dd><dt>5</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">NStr::TStringToNumFlags</span> — replaced by the integer-based counterparts. Replace <span class="nctnt ncbi-class">NStr::fStringToNumDefault</span> by 0 if you have compile errors.</p></dd><dt>6</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">NStr::StringToDouble()</span> — added <span class="nctnt ncbi-monospace">TStringToNumFlags</span> flags support.</p></dd><dt>7</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">CStopWatch::Stop()</span> — new function, to suspend a "timer".</p></dd><dt>8</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">SetCpuTimeLimit()</span> — added new parameter - maximum process termination time (default 5 seconds)</p></dd><dt>9</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">CProcess::Kill()</span> — added linger_timeout parameter to more control of process termination.</p></dd><dt>10</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">Sleep*()</span> — try to utilize unslept part of the time if interrupted by a signal.</p></dd><dt>11</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">CExec::RunSilent()</span> — new function to run console applications on MS Windows without terminal window. On UNIX, it is equivalent to <span class="nctnt ncbi-func">SpawnL()</span>.</p></dd><dt>12</dt><dd><p class="no_top_margin">IBlobStorage:
<ul><li><p>changed its name from bormer INetScheduleStorage</p></li><li><p>moved to the corelib (from connect/services)</p></li><li><p>got equipped with a class factory CBlobStorageFactory</p></li></ul>
					</p></dd><dt>13</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CSafeStatic[Ptr|Ref]</span> — added life span parameter for safe static objects. By default protect only initialization of an object, but do not control its lifetime.</p></dd><dt>14</dt><dd><p class="no_top_margin">Added groups and error handler for application arguments.</p></dd><dt>15</dt><dd><p class="no_top_margin">
							IReader/IWriter abstraction as well as stream implementation on top of IReader/IWriter-based stream buffer moved from UTIL to here.</p></dd><dt>16</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">CPluginManager</span> — new methods, to tune the DLL search paths: <span class="nctnt ncbi-func">ResetDllSearchPath()</span>, <span class="nctnt ncbi-func">SetDllStdSearchPath()</span> and <span class="nctnt ncbi-func">GetDllStdSearchPath()</span>.</p></dd><dt>17</dt><dd><p class="no_top_margin">
							Macros <span class="nctnt ncbi-macro">NCBI_RESTRICT</span> and <span class="nctnt ncbi-macro">NCBI_FORCEINLINE</span> — now available on all platforms.</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.CONNECT__Data_stream"><h3><span class="title">CONNECT — Data streaming, Networking, and Dispatching</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">CPipe::Poll()</span> — new function to check for the immediate availability of data in the pipe.</p></dd><dt>2</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">SOCK_HostPortToString()</span> and <span class="nctnt ncbi-func">SOCK_StringToHostPort()</span> — new functions equivalent to the now obsolete <span class="nctnt ncbi-func">HostPortToString()</span> and <span class="nctnt ncbi-func">StringToHostPort()</span>.</p></dd><dt>3</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">CONNUTIL_GetUsername()</span> and <span class="nctnt ncbi-func">ConnNetInfo_SetupStandardArgs()</span> — new.</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.UTIL__Miscellaneous_"><h3><span class="title">UTIL — Miscellaneous Low-Level APIs</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CCache</span> — new class template, implements a generic cache of objects..</p></dd><dt>2</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CRegexp</span> — added new methods <span class="nctnt ncbi-func">Escape()</span>, <span class="nctnt ncbi-func">IsMatch()</span>.</p></dd><dt>3</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CBlockingQueue&lt;&gt;</span>, <span class="nctnt ncbi-class">CPoolOfThreads&lt;&gt;</span> — reworked to support increased user control (via handles) and optionally run callbacks upon status changes.</p></dd><dt>4</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CBlockingQueue&lt;&gt;::Put</span>, <span class="nctnt ncbi-class">CPoolOfThreads&lt;&gt;::[x_]Accept[Urgent]Request</span> — Accept an optional timeout in which space may become available, defaulting to zero for consistency with existing behavior.</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.SERIAL__Data_Seriali"><h3><span class="title">SERIAL — Data Serialization (ASN.1, XML)</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added support of <span class="nctnt ncbi-type">BIT STRING</span> type data in serialization streams.</p></dd><dt>2</dt><dd><p class="no_top_margin">Added serial stream manipulators with parameters:
<span class="nctnt ncbi-monospace">MSerial_Format</span>, <span class="nctnt ncbi-monospace">MSerial_VerifyData</span>, <span class="nctnt ncbi-monospace">MSerialXml_DefaultStringEncoding</span>.</p></dd><dt>3</dt><dd><p class="no_top_margin">Added input stream object iterators:<br />
						<span class="nctnt ncbi-class">CIStreamObjectIterator</span> and <span class="nctnt ncbi-class">CIStreamStdIterator</span>
					</p></dd><dt>4</dt><dd><p class="no_top_margin">Added an option to write 64 bits integer data in ASNTOOL-compatible way.</p></dd><dt>5</dt><dd><p class="no_top_margin">
						ASN.1-based-object converter (for integrating C Toolkit code) — accommodate the C Toolkit's nonstandard treatment of Int8 fields, and allow text mode as an option in case other binary-representation issues somehow crop up.</p></dd><dt>6</dt><dd><p class="no_top_margin">RPC clients (for ID1, Entrez2, etc.) automatically retry in a wider range of circumstances.</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.CGI__CGI_and_FastCGI"><h3><span class="title">CGI — CGI and Fast-CGI Application Framework</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CCgiSession</span> — new class, to pass user data between CGI calls.
The default storage for the data is uses NetCache, but that can be easily changed by implementing ICgiSessionStorage interface in any other way.</p></dd><dt>2</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CCgiException</span> —  added HTTP status code and message.<br />
						<span class="nctnt ncbi-class">CCgiApplication</span> —  generates HTTP status header if it is set in CCgiException.</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.HTML__HTML_Generatio"><h3><span class="title">HTML — HTML Generation Library</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CHTMLPage</span> —  added possibility to cache template files and template libraries.</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.BerkeleyDB_API_bdb__"><h3><span class="title">BDB — Yet Another C++ API Based On BerkeleyDB</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-monospace">CBDB_Env</span> —  added method to control reverse splitting of pages (<span class="nctnt ncbi-func">RevSplitOff()</span>).</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.Local_data_storage"><h3><span class="title">Local data storage</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added read-only open mode.</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.DBAPI__SQL_Database_"><h3><span class="title">DBAPI — SQL Database Connectivity</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-monospace">IConnValidator</span> — new interface, to detect the status of database connection.</p></dd><dt>2</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-monospace">CTrivialConnValidator</span> and <span class="nctnt ncbi-monospace">CConnValidatorCoR</span> implementations of <span class="nctnt ncbi-monospace">IConnValidator</span>.</p></dd><dt>3</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">I_DeriverContext::ConnectValidated() </span> — new method, to check for the overall validity of the database connection after establishing the connection.</p></dd><dt>4</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">Close()</span> — method added to all context/command-aware classes.</p></dd><dt>5</dt><dd><p class="no_top_margin">Treat database messages with <span class="nctnt ncbi-app">severity == 10 &amp;&amp; msgnumber == 0</span> as having informational severity. Replaced fatal severity of SQL Server error messages with error severity.</p></dd><dt>6</dt><dd><p class="no_top_margin">
						Added new sample application <span class="nctnt ncbi-app">dbapi_conn_policy</span>.</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.ALGO_ALIGN_SPLIGN__S"><h3><span class="title">ALGO/ALIGN/SPLIGN — Spliced Alignment Algorithms</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CSplign</span> -<br />
Min singleton identity parameter re-introduced to control the identity cut-off for compartments unique per subject per strand.
Parts of the code have been optimized including data loading and core dynamic programming</p></dd><dt>2</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CHitFilter</span> -<br />
Maximal Greedy Reconciliation algorithm have been introduced to enforce non-redundancy of pairwise alignments across multiple sequences.</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.ALNMGR__Biosequence_"><h3><span class="title">ALNMGR — Bio-sequence Alignment Manager</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-monospace">CAlnMix::fRemoveLeadTrailGaps</span> — new flag, to remove end gaps.
					</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.BLAST"><h3><span class="title">BLAST</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<b>*</b>
Variable word size feature has been removed.</p></dd><dt>2</dt><dd><p class="no_top_margin">
						<b>*</b>
Removed (no longer needed) per-search-type constructors for <span class="nctnt ncbi-class">CRemoteBlast</span>.</p></dd><dt>3</dt><dd><p class="no_top_margin">Optimizations to lookup table width choice.</p></dd><dt>4</dt><dd><p class="no_top_margin">Introduced new varieties of composition based statistics.</p></dd><dt>5</dt><dd><p class="no_top_margin">Enable the use of composition based statistics for tblastn.</p></dd><dt>6</dt><dd><p class="no_top_margin">Added interruptible API to CBl2Seq.</p></dd><dt>7</dt><dd><p class="no_top_margin">
						<b>*</b>
Add support for user-specified query masked locations in <span class="nctnt ncbi-class">CRemoteBlast</span>.</p></dd><dt>8</dt><dd><p class="no_top_margin">Optimizations to RPS-BLAST word finder.</p></dd><dt>9</dt><dd><p class="no_top_margin">
						Set number of subjects per database chunk adaptively.</p></dd><dt>10</dt><dd><p class="no_top_margin">Changed how the B, Z, X and U residues are scored within the composition adjustment library. Most significantly, the method of computing scores involving X has changed. The score of aligning X to a character is the expected score of aligning that character to any true amino acid, subject to the restriction that the score be at most -1. The score of aligning X to itself is the expected score of aligning any two true amino acids, again subject to the restriction that the score be at most -1. U is treated as equivalent to X. The score of aligning a true amino acid to the rare B or Z characters is computed by a log odds formulation that uses the sum of the target or background frequencies of the two characters each ambiguity character represents as the target or background frequency of the ambiguity character.</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.BIOOBJECTS__BioObjec"><h3><span class="title">BIO-OBJECTS — Bio-Object Specific Utility Functions (Not Involving OM++)</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<b>*</b>
						<span class="nctnt ncbi-class">CSeq_id</span> — <span class="nctnt ncbi-func">ParseFastaIds()</span> treats untagged numeric IDs as local IDs rather than GIs, which must be tagged gi|... in this context.</p></dd><dt>2</dt><dd><p class="no_top_margin">
						<b>*</b>
						<span class="nctnt ncbi-class">CSeq_id</span> — use (and accept) the FASTA ID tag pgp|... for "pre-grant patents" (applications).</p></dd><dt>3</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CSeq_id</span> — disregard extra trailing vertical bars in FASTA-style IDs.</p></dd><dt>4</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CSeq_id</span> — <span class="nctnt ncbi-func">IdentifyAccession()</span> supports several new prefixes (DX through EB).</p></dd><dt>5</dt><dd><p class="no_top_margin">
						Support the new amino acid codes J/Xle and O/Pyl.</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.OM__Object_Manager__"><h3><span class="title">OM++ — Object Manager — For Retrieving and Processing Bio-Objects</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CSeqLocMapper</span> — if intervals get truncated while mapping, it is now idicated by setting fuzz to lim tl/tr.</p></dd><dt>2</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CSeqLocMapper</span> — added flag to check strands before mapping (allows to filter out locations which are not on the source strand).</p></dd><dt>3</dt><dd><p class="no_top_margin">
						Added methods to collect only the annotations of selected types rather than all annotation objects.</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.OM_LOADERS_READERS__"><h3><span class="title">OM++ LOADERS/READERS — Data Retrieval Libraries for OM++</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Modified Phrap reader to allow aligned segment to be shorter than base segment, skip WR, CT and RT tags in some cases. Generate warnings when skipping tags.</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.OM_DEMO_program_objm"><h3><span class="title">OM++ DEMO program (objmgr_demo)</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added options to use LDS and BLAST data loaders.</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.BUILD_FRAMEWORK_UNIX"><h3><span class="title">BUILD FRAMEWORK (UNIX)</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<b>*</b>
The minimum supported Berkeley DB version is now 4.3. (4.4 is also fully supported.)</p></dd><dt>2</dt><dd><p class="no_top_margin">Configure now supports —with-distcc, which behaves similarly to —with-ccache apart from being off by default.</p></dd><dt>3</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-path">compilers/GCC.sh</span> now supports looking for a specific GCC version, which the user can supply as an optional first argument.</p></dd><dt>4</dt><dd><p class="no_top_margin">Configure now supports a —with-gbench option that ensures correct compilation options (—with-mt —with-dll) and errors out if any of the third-party libraries Genome Workbench requires are unavailable.</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.PTB__Project_Tree_Bu"><h3><span class="title">PTB — Project Tree Builder for MSVC++ .NET</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added DATASPEC-ALL project, which includes all Toolkit projects with automatically generated sources.</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.APPLICATIONS"><h3><span class="title">APPLICATIONS</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-app">NetCache</span>:</p><ul><li><p>Server-side: added session management, such as automatic shutdown when the last client is logged off.</p></li><li><p>Client-side: added an implementation of ICache interface with NetCache.</p></li></ul></dd><dt>2</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-app">NetSchedule</span>:</p><ul><li><p>Optimization of use of BDB in order to reduce number of collisions and deadlocks when running transactions </p></li><li><p>Reduced memory consumption</p></li><li><p>Changed queue truncation algorithm not to overflow in-memory transaction log</p></li><li><p>Fixed scheduling bug related to job affinity</p></li><li><p>Added client registration command</p></li></ul></dd></dl></div><div class="sec" id="release_notes_04_30_2006.GRID_DISTRIBUTED_COM"><h3><span class="title">GRID (DISTRIBUTED COMPUTING) FRAMEWORK</span></h3><div class="sec" id="release_notes_04_30_2006.CONNECT_SERVICES"><h4><span class="title">CONNECT/SERVICES — Components for the Network Grid Framework</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						CNetScheduleNSSStorage_NetCache — renamed to CBlobStorage_NetCache and moved to a separate library.</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.Grid_Worker_Node_Imp"><h4><span class="title">Grid Worker Node Implementation Framework</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Idle task facility — to allow worker node to perform some task when it is not executing any "real" jobs.</p></dd><dt>2</dt><dd><p class="no_top_margin">"remote_app" worker node — to launch external applications or scripts via network using NetSchedule/NetCache services.</p></dd><dt>3</dt><dd><p class="no_top_margin">"max_failed_jobs" worker node configuration parameter — to specify the maximum number of failed jobs after which the worker node will shutdown itself.</p></dd><dt>4</dt><dd><p class="no_top_margin">"use_embedded_storage" configuration parameter for the worker node (and its client too) — to specify if the worker node should try to use a NetSchedule's internal storage instead of NetCache blobs for passing its input/output data. This reduces the network load and increases the data exchange.</p></dd><dt>5</dt><dd><p class="no_top_margin">"reuse_job_object" worker node configuration parameter — to have only one instance of the IWorkerNodeJob interface per job's thread (instead of creating a new object for each job).</p></dd><dt>6</dt><dd><p class="no_top_margin">"use_permanent_connection" worker node configuration parameter — to keep a permanent connection to NetSchedule server (rather than to establish a new connection for each job exchange).</p></dd></dl></div></div></div><div class="sec" id="release_notes_04_30_2006.Documentation"><h2><span class="title">Documentation</span></h2><div class="sec" id="release_notes_04_30_2006.Document_Location"><h3><span class="title">Document Location</span></h3><p>The documentation is available online at <!--ro--><a href="toc.html">http://www.ncbi.nlm.nih.gov/books/bv.fcgi?rid=toolkit.TOC&amp;depth=2</a> as a book titled "The NCBI C++ Toolkit". This is an online searchable book.</p><p>The C++ Toolkit book also provides PDF version of the chapters; although these are not up to date in this release. The PDF version can be accessed by a link that appears on each page.</p><p>The older HTML documentation has been deprecated and is no longer being updated, and "The NCBI C++ Toolkit" online book at the previously listed URLs is the official documentation.</p></div><div class="sec" id="release_notes_04_30_2006.Document_Content"><h3><span class="title">Document Content</span></h3><p>Documentation has been grouped into chapters and sections that provide a more logical coherence and flow. New sections and paragraphs continue to be added to update and clarify the older documentation or provide new documentation. The chapter titled "Introduction to the C++ Toolkit" gives an overview of the C++ Toolkit. This chapter contains links to other chapters containing more details on a specific topic and is a good starting point for the new comer.</p><p>The DOXYGEN source browser is used to complement the traditional docs with structured DOXYGEN-generated "Library Reference" ones based on the in-source comments. You can access the DOXYGEN source browser for the NCBI code from:</p><p>
				<a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml">http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/</a>
			</p><p>The above link is also available under the "Browsers" that appears on each page.</p><p>You can also access the CVS code repository via a Web interface. These links also appear in the sidebar box on each page.</p><p>A C/C++ Symbol Search query appears on each page of the online Toolkit documentation. You can use this to perform a symbol search on the public or in-house versions of LXR, Doxygen and Library. The Library search runs a CGI script that lists the library name where the symbol (such as function) is defined in.</p></div></div><div class="sec" id="release_notes_04_30_2006.Building_on_the_MacO"><h2><span class="title">Building on the MacOS</span></h2><p>We now build all the libraries and most of the applications including the Genome Workbench (<span class="nctnt ncbi-app">gbench</span>), some test and a few demo applications. We also build the FLTK library's GUI editor, <span class="nctnt ncbi-app">fluid</span>.</p><p>All apps are built as application bundles except <span class="nctnt ncbi-app">gbench_plugin_scan</span> and <span class="nctnt ncbi-app">datatool</span> which are built as command line apps.</p><div class="sec" id="release_notes_04_30_2006.GCC"><h3><span class="title">GCC</span></h3><p>Uses a regular <span class="nctnt ncbi-monospace">Unix</span> build pattern: run <span class="nctnt ncbi-app">configure</span>, then <span class="nctnt ncbi-app">make</span>.</p></div><div class="sec" id="release_notes_04_30_2006.Xcode"><h3><span class="title">Xcode</span></h3><p>When building the toolkit with <span class="nctnt ncbi-app">Xcode</span>, the latest version of <span class="nctnt ncbi-app">Xcode</span> (at least 1.5) is required for the trouble-free build. Build procedure is as follows: open, build and run a project file in <span class="nctnt ncbi-path">compilers/xCode</span>. This is a GUI tool to generate a new NCBI C++ Toolkit <span class="nctnt ncbi-app">Xcode</span> project. You'll have an option to specify third-party installation directories and choose which packages (libs, applications and tests) to include into the final project. The option to automatically download and install all the third-party libraries is also available.</p><p>
				<span class="nctnt ncbi-app">Xcode</span> build fully supports all the latest Apple innovations: distributed builds, optional CPU specific optimization, pre-compiled headers, fix &amp; continue, code cleanup and Zero Link (with few exceptions).</p><p>
				<span class="nctnt ncbi-app">Xcode</span> build has a Shell Script build phase for each Target dependent on generated ASN files. These shell scripts use <span class="nctnt ncbi-app">datatool</span> to regenerate source files each time the ASN specification files do change.</p><p>
				<span class="nctnt ncbi-app">Xcode</span> builds all libraries as a Mach-O dynamically linked shared ones (<span class="nctnt ncbi-path">.dylib</span>) and all Genome Workbench plugins as Mach-O bundles (also <span class="nctnt ncbi-path">.dylib</span> extension). Note, that <span class="nctnt ncbi-app">Xcode</span> will place Genome Workbench plugins inside Genome Workbench application bundle (<span class="nctnt ncbi-path">Genome Workbench.app/Contents/MacOS/plugins</span>).</p></div></div><div class="sec" id="release_notes_04_30_2006.Platforms_OSs__compi"><h2><span class="title">Platforms (OS's, compilers used inside NCBI)</span></h2><p>This release was successfully tested on at least the following platforms — but may also work on other platforms. Since the previous release, some platforms were dropped from this list, just because we do not use them here anymore, and some were added (these new platforms are highlighted using bold font). Also, it can happen that some projects would not work (or even compile) in the absence of 3rd-party packages, or with older or newer versions of such packages — in these cases, just skipping such projects (e.g. using flag "-k" for <span class="nctnt ncbi-app">make</span> on UNIX), can get you through.</p><div class="table" id="release_notes_04_30_2006.T1"><h4><span class="title">Table 2. Unix OS's and Supported Compilers</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Architecture </th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">Linux-2.<b>6.x</b> (w/ LIBC 2.3.2) </span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">x86-32 </span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">
								GCC 3.4.0<br />ICC 8.0 (build 20040520Z, package ID l_cc_pc_8.0.066_pe067.1)
 <br />(GCC 3.0.4, 2.95.3, <b>4.1.0</b> - nominal support)</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">Linux-2.6.x (w/ LIBC 2.3.3) </span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">x86-64 </span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">
								GCC 4.0.1<br />ICC 9.0 (build 20051201)</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">Solaris-8</span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">SPARC </span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">
								C++ 5.3 (WorkShop 6u2) patch 111685-23 (64-, 32-bit)<br />(GCC 3.4.3 - nominal support)</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">Solaris-10 </span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">SPARC </span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">
								Sun C++ 5.5 (Studio8) patch 113817-15<br />
							GCC 4.0.1 (32-bit mode only)
						</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">Solaris-9 </span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">x86-32 </span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">C++ 5.3 (WorkShop 6u2) patch 111686-13<br />GCC 3.4.3
						</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">IRIX64-6.5 </span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">SGI-Mips </span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">MIPSpro 7.3.1.3m (64-bit, 32-bit)</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">FreeBSD-4.10 </span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">x86-32 </span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">GCC 3.4.2</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">Digital Tru64 Unix 5.1 (aka OSF1) </span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">ALPHA </span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">GCC 3.3.2</td></tr></tbody></table></div></div><div class="table" id="release_notes_04_30_2006.T2"><h4><span class="title">Table 3. MS Windows and Supported Compilers</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">MS Windows-32 </span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">
								MS Visual Studio .NET 2003 (C++ 7.1). See documentation for building the Toolkit with MS Visual C++ .NET<br />
							NOTE: We also ship an easily buildable archive of <a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2006/Apr_30_2006/ThirdParty/README">3rd-party</a> packages (including <a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2006/Apr_30_2006/NCBI_C_Toolkit/README">NCBI C Toolkit</a>) for this platform.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">MS Windows-32 </span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">
								<b>MS Visual Studio .NET 2005 (C++ 8.0)</b><br />
							</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">MS Windows-64</span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">
								<b>MS Visual Studio .NET 2005 (C++ 8.0)</b><br />
							</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace"><b>Cygwin 1.5.18</b> on MS Windows-32 </span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">
								<b>GCC 3.4.4</b><br />
							</td></tr></tbody></table></div></div><div class="table" id="release_notes_04_30_2006.T3"><h4><span class="title">Table 4. Mac OS X, and Supported Compilers</span></h4><div class="table-scroll"><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">Darwin on MacOS X 10.4 ("Tiger")</span>
						</td><td align="left" valign="top" rowspan="1" colspan="1"> GCC  <b>4.0.1</b></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
							<span class="nctnt ncbi-monospace">Darwin on MacOS X 10.4 ("Tiger")</span>
						</td><td align="left" valign="top" rowspan="1" colspan="1">Xcode 1.5 - 2.2.1
						</td></tr></tbody></table></div></div></div><div class="sec" id="release_notes_04_30_2006.Caveats_and_Hints"><h2><span class="title">Caveats and Hints</span></h2><div class="sec" id="release_notes_04_30_2006.GCC_295"><h3><span class="title">GCC 2.95</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Poor MT-safety record.</p></dd><dt>2</dt><dd><p class="no_top_margin">Relatively incomplete/incorrect (comparing to modern compilers) STL implementation.</p></dd><dt>3</dt><dd><p class="no_top_margin">It is going to be deprecated in NCBI as soon as we have any significant trouble with its maintenance.</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.GCC_304"><h3><span class="title">GCC 3.0.4</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Destructor of constructed class member is not called when exception is thrown from a method called from class constructor body (fixed in GCC 3.3).</p></dd><dt>2</dt><dd><p class="no_top_margin">STL stream uses locale in thread unsafe way which may result to segmentation fault when run in multithread mode (fixed in GCC 3.3).</p></dd><dt>3</dt><dd><p class="no_top_margin">Long-file support for C++ streams is disabled/broken (first broken in 3.0, fixed in 3.4).</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.GCC_33"><h3><span class="title">GCC 3.3</span></h3><p>Other than the feature described below, GCC 3.3.2 had been very good to us; it had a lot of very ugly bugs finally fixed.</p><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Painfully slow linking in debug mode on Linux with GCC-3.3 compiler. — Starting with binutils 2.12 linker tries to merge constant/debug strings marked for merging in object files. But it seems it does this job very inefficiently - I've seen messages about it in internet.  GCC starting with version 3.2 marks section of string constants ready for merging, and also has an option to disable this flag in object files (-fno-merge-constants). Adding this flag to compilation stage allows to avoid slow linking. GCC 3.3 also sets merge flag for debug sections and unfortunately there is no option to disable this flag. As a result, linking of debug executables significatly slower than with gcc 3.0.4. The slowdown rate depends on size of debug strings section and it's non-linear, so bigger projects will suffer more of this bug (N^2). Binutils 2.15 fixes this. The link time still 2 times slower than without symbol merge, but the resultant executable is about two times smaller in size, and no compiler patching is necessary. We are still testing it in-house. We had to patch GCC 3.3 in-house with the fix described at http://lists.boost.org/MailArchives/boost/msg53004.php.</p></dd><dt>2</dt><dd><p class="no_top_margin">Long-file support still broken.</p></dd></dl></div><div class="sec" id="release_notes_04_30_2006.GCC_34x__40x"><h3><span class="title">GCC 3.4.x, 4.0.x</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">The "Painfully slow linking..." (see GCC3.3, [1] above) was an issue, and we had to patch it in-house to speed up, a la GCC 3.3 — until we finally upgraded to binutils 2.15.</p></dd><dt>2</dt><dd><p class="no_top_margin">At least on Linux, ifstream::readsome() does not always work for large files, as it calls an ioctl that doesn't work properly for large files (we didn't test whether 4.0.x fixed this).</p></dd><dt>3</dt><dd><p class="no_top_margin">At least on Linux, GCC 3.4.[0,1] optimizer (very rarely) generates incorrect code when comparing enumerated values in else-ifs. (Fixed in 3.4.2)</p></dd><dt>4</dt><dd><p class="no_top_margin">GCC 3.4.3, 3.4.4 and 4.0 have a bug in the C++ stream library that affects some parts of our code, notably CGI framework. (Fixed in 4.0.1).</p></dd></dl></div></div><div class="sec" id="release_notes_04_30_2006.Last_Updated"><h2><span class="title">Last Updated</span></h2><p>This section last updated on May 5, 2006.</p></div></div></div><div id="tkb_internal" class="col four_col last"><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" id="source-branding"><div class="iconblock clearfix ten_col"><a class="img_link icnblk_img" title="Table of Contents Page" target="mainwindow" href="toc.html"><img class="source-thumb" src="img/th-toolkit-lrg.png" alt="Cover of The NCBI C++ Toolkit Book" height="100px" width="80px" /></a><div class="icnblk_cntnt"><div class="_bk_pgbnr_cit"><div class="_bk_pgbnr_cit_ln">The NCBI C++ Toolkit Book [Internet].</div><div class="_bk_pgbnr_cit_ln">Vakatov D, editor.</div><div class="_bk_pgbnr_cit_ln">Bethesda (MD): <a href="http://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (US)</a>; 2004-.</div></div></div></div></div><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" class="half_rhythm"><ul class="inline_list"><li><a href="toc.html">Table of Contents Page</a></li><li> | </li><li><a class="jig-ncbidialog" data-jigconfig="width:400 , modal:true" href="#citation-window">Cite this Page</a><div class="jig-ncbidialog-box" style="display:none" title="Cite this Page" id="citation-window"><div class="bk_tt">
        Vakatov D, editor. The NCBI C++ Toolkit Book [Internet]. Bethesda (MD): National Center for Biotechnology Information (US); 2004-.  , Release Notes (April 30, 2006)
                    .
        [Updated: May 2, 2006]
    .

                            <div class="bk_tt"><span class="bk_cite_avail">Available from: http://www.ncbi.nlm.nih.gov/toolkit/doc/book/release_notes_04_30_2006</span></div></div><span class="bk_cite_avail"></span></div></li></ul></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>In this page</h3></div></div><div class="portlet_content"><ul><li><a href="#release_notes_04_30_2006.Download_Location">Download Location</a></li><li><a href="#release_notes_04_30_2006.Source_Archive_Conte">Source Archive Contents</a></li><li><a href="#release_notes_04_30_2006.New_Development">New Development</a></li><li><a href="#release_notes_04_30_2006.Documentation">Documentation</a></li><li><a href="#release_notes_04_30_2006.Building_on_the_MacO">Building on the MacOS</a></li><li><a href="#release_notes_04_30_2006.Platforms_OSs__compi">Platforms (OS's, compilers used inside NCBI)</a></li><li><a href="#release_notes_04_30_2006.Caveats_and_Hints">Caveats and Hints</a></li><li><a href="#release_notes_04_30_2006.Last_Updated">Last Updated</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Search</h3></div></div><div class="portlet_content"><form name="frmSymbolSearch" onsubmit="return SymbolSearch('toolkit')" action="/" id="frmSymbolSearch"><table><tbody><tr align="center" valign="middle"><td colspan="2" nowrap="nowrap"><input id="search_box" name="__symbol" style="vertical-align:                                        middle; line-height:normal " onkeypress="SymbolSearchKeyPress('toolkit',event)" type="text" />  					<label for="search_box" class="offscreen_noflow">Search term</label><button type="submit" value="Search" style="vertical-align:                                        middle; line-height:normal;" onclick="javasrcipt:SymbolSearch('toolkit');">Search</button></td></tr><tr class="internal-only"><th style="text-align:left">Public</th><th style="text-align:left">Internal</th></tr><tr><td><input id="pToolkitAll" name="__symboloc" type="radio" /><label for="pToolkitAll" title="Search in the C/C++ source code and in this Book">All</label></td><td class="internal-only"><input id="iToolkitAll" name="__symboloc" type="radio" checked="checked" /><label for="iToolkitAll" title="Search in the C/C++ source code,  this Book, and Wiki">All</label></td></tr><tr><td><input id="pLXR" name="__symboloc" type="radio" /><label for="pLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td><td class="internal-only"><input id="iLXR" name="__symboloc" type="radio" /><label for="iLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td></tr><tr><td><input id="pLib" name="__symboloc" type="radio" /><label for="pLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td><td class="internal-only"><input id="iLib" name="__symboloc" type="radio" /><label for="iLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td></tr></tbody></table></form></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Source Browsers</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><span class="bkp_prtlt_txt"> LXR: </span><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a><span class="bkp_prtlt_txt"> / </span><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/lxr/ident" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Doxygen: </span><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a><span class="bkp_prtlt_txt"> / </span><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/doxyhtml" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>SVN Source Repository</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><span class="bkp_prtlt_txt"> Documentation: </span><a href="ch_getcode_svn.html#ch_getcode_svn.code_retrieval">in-house</a><span class="bkp_prtlt_txt"> / </span><a href="ch_getcode_svn.html#ch_getcode_svn.external">public</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (dev): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/c++" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a><span class="bkp_prtlt_txt"> / </span><a href="http://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/c++/" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">public</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (extra): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/internal/c++" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li><li class="internal-only"><span class="bkp_prtlt_txt"> Web (prod): </span><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/production/components" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">in-house</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Resources</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/lib_search/libsearch.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Search Tool</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/depgraphs/dglistdeps.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Dependencies Tool</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/wiki-private/CxxToolkit" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Wiki</a></li><li class="internal-only"><a href="https://jira.ncbi.nlm.nih.gov/browse/CXX" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">JIRA</a></li><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/STAT/test_stat/test_stat_ext.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="View results of testsuite runs on the automatic C++ Toolkit builds">Testsuite</span></a></li><li><a href="release_notes.html">Release Notes</a></li><li><a href="ch_getcode_svn.html#ch_getcode_svn.ftp_download">Download the Toolkit</a></li><li><a href="http://www.ncbi.nlm.nih.gov/toolkit/doc/book/pdf/TOC.pdf">PDF of the book</a></li><li><a href="ch_style.html">Coding Style</a></li><li><a href="ch_faq.html#ch_faq.mailing_lists">Mailing Lists</a></li><li><a href="mailto:CPP-CORE@NCBI.NLM.NIH.GOV" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="Contact C++ Toolkit group">Help and Support</span></a></li></ul></div></div></div><div id="tkb_public" class="col four_col last"><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" id="source-branding"><div class="iconblock clearfix ten_col"><a class="img_link icnblk_img" title="Table of Contents Page" target="mainwindow" href="toc.html"><img class="source-thumb" src="img/th-toolkit-lrg.png" alt="Cover of The NCBI C++ Toolkit Book" height="100px" width="80px" /></a><div class="icnblk_cntnt"><div class="_bk_pgbnr_cit"><div class="_bk_pgbnr_cit_ln">The NCBI C++ Toolkit Book [Internet].</div><div class="_bk_pgbnr_cit_ln">Vakatov D, editor.</div><div class="_bk_pgbnr_cit_ln">Bethesda (MD): <a href="http://www.ncbi.nlm.nih.gov/">National Center for Biotechnology Information (US)</a>; 2004-.</div></div></div></div></div><div xmlns:str="http://exslt.org/strings" xmlns:c="http://exslt.org/common" class="half_rhythm"><ul class="inline_list"><li><a href="toc.html">Table of Contents Page</a></li><li> | </li><li><a class="jig-ncbidialog" data-jigconfig="width:400 , modal:true" href="#citation-window">Cite this Page</a><div class="jig-ncbidialog-box" style="display:none" title="Cite this Page" id="citation-window"><div class="bk_tt">
        Vakatov D, editor. The NCBI C++ Toolkit Book [Internet]. Bethesda (MD): National Center for Biotechnology Information (US); 2004-.  , Release Notes (April 30, 2006)
                    .
        [Updated: May 2, 2006]
    .

                            <div class="bk_tt"><span class="bk_cite_avail">Available from: http://www.ncbi.nlm.nih.gov/toolkit/doc/book/release_notes_04_30_2006</span></div></div><span class="bk_cite_avail"></span></div></li></ul></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>In this page</h3></div></div><div class="portlet_content"><ul><li><a href="#release_notes_04_30_2006.Download_Location">Download Location</a></li><li><a href="#release_notes_04_30_2006.Source_Archive_Conte">Source Archive Contents</a></li><li><a href="#release_notes_04_30_2006.New_Development">New Development</a></li><li><a href="#release_notes_04_30_2006.Documentation">Documentation</a></li><li><a href="#release_notes_04_30_2006.Building_on_the_MacO">Building on the MacOS</a></li><li><a href="#release_notes_04_30_2006.Platforms_OSs__compi">Platforms (OS's, compilers used inside NCBI)</a></li><li><a href="#release_notes_04_30_2006.Caveats_and_Hints">Caveats and Hints</a></li><li><a href="#release_notes_04_30_2006.Last_Updated">Last Updated</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Search</h3></div></div><div class="portlet_content"><form name="frmSymbolSearch" onsubmit="return SymbolSearch('toolkit')" action="/" id="frmSymbolSearch"><table><tbody><tr align="center" valign="middle"><td colspan="2" nowrap="nowrap"><input id="search_box" name="__symbol" style="vertical-align:                                        middle; line-height:normal " onkeypress="SymbolSearchKeyPress('toolkit',event)" type="text" />  					<label for="search_box" class="offscreen_noflow">Search term</label><button type="submit" value="Search" style="vertical-align:                                        middle; line-height:normal;" onclick="javasrcipt:SymbolSearch('toolkit');">Search</button></td></tr><tr><td colspan="2"><input id="pToolkitAll" name="__symboloc" type="radio" checked="checked" /><label for="pToolkitAll" title="Search in the C/C++ source code and in this Book">All</label></td></tr><tr><td colspan="2"><input id="pLXR" name="__symboloc" type="radio" /><label for="pLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td></tr><tr><td colspan="2"><input id="pLib" name="__symboloc" type="radio" /><label for="pLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td></tr></tbody></table></form></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Source Browsers</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">LXR</a></li><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">Doxygen</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>SVN Source Repository</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="ch_getcode_svn.html#ch_getcode_svn.external">Documentation</a></li><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/viewvc/v1/trunk/c++/" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Web (dev)</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Resources</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lib_search/libsearch.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Search Tool</a></li><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/depgraphs/dglistdeps.cgi" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri">Library Dependencies Tool</a></li><li><a href="release_notes.html">Release Notes</a></li><li><a href="ch_getcode_svn.html#ch_getcode_svn.ftp_download">Download</a></li><li><a href="http://www.ncbi.nlm.nih.gov/toolkit/doc/book/pdf/TOC.pdf">PDF of the book</a></li><li><a href="ch_style.html">Coding Style</a></li><li><a href="ch_faq.html#ch_faq.mailing_lists">Mailing Lists</a></li><li><a href="mailto:CPP-GROUP@NCBI.NLM.NIH.GOV" ref="pagearea=source-application&amp;targetsite=external&amp;targetcat=link&amp;targettype=uri"><span title="Contact C++ Toolkit group">Help and Support</span></a></li></ul></div></div></div></div></div></div></body></html>
