<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><title>Examples and Demos - The NCBI C++ Toolkit Book [Internet] - NCBI</title><meta name="description" content="A comprehensive manual on the NCBI C++ toolkit, including its design and development framework, a C++ library reference, software examples and demos, FAQs and release notes." /><meta name="robots" content="index,nofollow,noarchive" /><meta name="page_id" content="ch_demo" /><meta name="page_type" content="chapter" /><meta name="ncbipopper" content="triggerPosition:'bottom left',destPosition:'top left',openEvent:'click',closeEvent:'click'" /><script type="text/javascript" src="js/jig/1.7/js/jig.min.js"></script><script type="text/javascript" src="js/toolkitSymbolSearch.js"></script><link type="text/css" rel="stylesheet" href="css/ptbooks.css" /><link type="text/css" rel="stylesheet" href="css/toolkit.css" /></head><body><div class="grid no_max_width"><div class="col twelve_col nomargin shadow"><div class="content"><div class="col eight_col"><div class="nvgtn"><a href="#__nav_cntnt" class="jig-ncbipopper">Contents â–¾</a><div style="display:none;width:700px;height:500px;overflow:auto;" id="__nav_cntnt"><a class="right" href="toc.html">Table of Contents Page</a><ul class="simple-list"><li class="half_rhythm"><a href="toolkit.fm.html">Book Information</a></li><li class="half_rhythm"><a href="part1.html">Part 1. Overview</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_intro.html">1. Introduction to the C++ Toolkit</a></li><li class="half_rhythm"><a href="ch_start.html">2. Getting Started</a></li></ul></li><li class="half_rhythm"><a href="part2.html">Part 2. Development Framework</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_getcode_svn.html">3. Retrieve the Source Code (FTP and Subversion)</a></li><li class="half_rhythm"><a href="ch_config.html">4. Configure, Build, and Use the Toolkit</a></li><li class="half_rhythm"><a href="ch_build.html">5. Working with Makefiles</a></li><li class="half_rhythm"><a href="ch_proj.html">6. Project Creation and Management</a></li><li class="half_rhythm"><a href="ch_style.html">7. Programming Policies and Guidelines</a></li></ul></li><li class="half_rhythm"><a href="part3.html">Part 3. C++ Toolkit Library Reference</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_core.html">8. Portability, Core Functionality and Application Framework</a></li><li class="half_rhythm"><a href="ch_conn.html">9. Networking and IPC</a></li><li class="half_rhythm"><a href="ch_dbapi.html">10. Database Access - SQL, Berkley DB</a></li><li class="half_rhythm"><a href="ch_cgi.html">11. CGI and Fast-CGI </a></li><li class="half_rhythm"><a href="ch_html.html">12. HTML</a></li><li class="half_rhythm"><a href="ch_ser.html">13. Data Serialization (ASN.1, XML)</a></li><li class="half_rhythm"><a href="ch_datamod.html">14. Biological Sequence Data Model</a></li><li class="half_rhythm"><a href="ch_objmgr.html">15. Biological Object Manager</a></li><li class="half_rhythm"><a href="ch_blast.html">16. BLAST API</a></li><li class="half_rhythm"><a href="ch_dataaccess.html">17. Access to NCBI data</a></li><li class="half_rhythm"><a href="ch_algoalign.html">18. Biological Sequence Alignment</a></li><li class="half_rhythm"><a href="ch_gui.html">19. GUI and Graphics</a></li><li class="half_rhythm"><a href="ch_boost.html">20. Using the Boost Unit Test Framework</a></li></ul></li><li class="half_rhythm"><a href="part4.html">Part 4. Wrappers for 3rd-Party Packages</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_xmlwrapp.html">21. XmlWrapp (XML parsing and handling, XSLT, XPath)</a></li></ul></li><li class="half_rhythm"><a href="part5.html">Part 5. Software</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_debug.html">22. Debugging, Exceptions, and Error Handling</a></li><li class="half_rhythm"><a href="ch_grid.html">23. Distributed Computing</a></li><li class="half_rhythm"><a href="ch_app.html">24. Applications</a></li><li class="half_rhythm"><a href="ch_demo.html">25. Examples and Demos</a></li><li class="half_rhythm"><a href="ch_res.html">26. C Toolkit Resources for C++ Toolkit Users</a></li></ul></li><li class="half_rhythm"><a href="part6.html">Part 6. Help and Support</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_browse.html">27. NCBI C++ Toolkit Source Browser</a></li><li class="half_rhythm"><a href="ch_devtools.html">28. Software Development Tools</a></li><li class="half_rhythm"><a href="ch_xmlauthor.html">29. XML Authoring using Word</a></li><li class="half_rhythm"><a href="ch_faq.html">30. FAQs, Useful Documentation Links, and Mailing Lists</a></li></ul></li><li class="half_rhythm"><a href="part7.html">Part 7. Library and Applications Configuration</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_libconfig.html">31. Library Configuration</a></li></ul></li><li class="half_rhythm"><a href="part8.html">Release Notes</a><ul class="simple-list"><li class="half_rhythm"><a href="release_notes.html">Release Notes (Version 9, May 2012)</a></li><li class="half_rhythm"><a href="release_notes_7-05_2011.html">Release Notes (Version 7, May 2011)</a></li><li class="half_rhythm"><a href="release_notes_06_29_2010.html">Release Notes (June, 2010)</a></li><li class="half_rhythm"><a href="release_notes_05_15_2009.html">Release Notes (May, 2009)</a></li><li class="half_rhythm"><a href="release_notes_12_24_2008.html">Release Notes (December, 2008)</a></li><li class="half_rhythm"><a href="release_notes_12_31_2008.html">Release Notes (March, 2008)</a></li><li class="half_rhythm"><a href="release_notes_08_27_2007.html">Release Notes (August, 2007)</a></li><li class="half_rhythm"><a href="release_notes_03_12_2007.html">Release Notes (March, 2007)</a></li><li class="half_rhythm"><a href="release_notes_08_14_2006.html">Release Notes (August, 2006)</a></li><li class="half_rhythm"><a href="release_notes_04_30_2006.html">Release Notes (April 30, 2006)</a></li><li class="half_rhythm"><a href="release_notes_12_31_2005.html">Release Notes (December 31, 2005)</a></li><li class="half_rhythm"><a href="release_notes_10_03_2005.html">Release Notes (August, 2005)</a></li><li class="half_rhythm"><a href="release_notes_05_05_2005.html">Release Notes (April, 2005)</a></li><li class="half_rhythm"><a href="release_notes_03_09_2005.html">Release Notes (February, 2005)</a></li><li class="half_rhythm"><a href="release_notes_11_22_2004.html">Release Notes (November 22, 2004)</a></li><li class="half_rhythm"><a href="release_notes_10_2_2004.html">Release Notes (October 2, 2004)</a></li><li class="half_rhythm"><a href="release_notes_7_8_2004.html">Release Notes (July 8, 2004)</a></li><li class="half_rhythm"><a href="release_notes_april_16_2004.html">Release Notes (April 16, 2004)</a></li><li class="half_rhythm"><a href="release_notes_12_08_2003.html">Release Notes (December 8, 2003)</a></li><li class="half_rhythm"><a href="release_notes_08_01_2003.html">Release Notes (August 1, 2003)</a></li></ul></li><li class="half_rhythm"><a href="app1.appendix1.html">Appendix - Books and Styles</a></li></ul></div><div class="pagination"><a class="pagelink active prev" title="Previous page" href="ch_app.html">&lt; Prev</a><a class="pagelink active next" title="Next page" href="ch_res.html">Next &gt;</a></div></div><div class="meta_content"><h1><span class="label">25</span><span class="title">Examples and Demos</span></h1><p class="small">Last Update: January 19, 2011.</p></div><div class="body_content"><div class="sec"><h2><span class="title">Overview</span></h2><p>The overview for this chapter consists of the following topics:</p><ul><li><p>Introduction</p></li><li><p>Chapter Outline</p></li></ul><div class="sec"><h3><span class="title">Introduction</span></h3><p>See <a href="ch_start.html">Getting Started</a> for basic information on using the NCBI C++ Toolkit.</p></div><div class="sec"><h3><span class="title">Chapter Outline</span></h3><ul><li><p>Examples</p><ul><li><p><a href="ch_proj.html#ch_proj.new_proj_struct">Sample Applications Available with the new_project script</a></p><ul><li><p><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/source/src/sample/app/basic/">A basic example</a> using the <a href="ch_core.html">xncbi</a> core library</p></li><li><p><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/source/src/sample/app/cgi/">An example CGI application</a> using the <a href="ch_cgi.html">xcgi and xfcgi</a> libraries.</p></li><li><p><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/source/src/sample/app/objects/">An example for serializable ASN.1 objects and the <span class="nctnt ncbi-monospace">Object Manager</span></a> using the <a href="ch_ser.html">xobjects</a> library.</p></li></ul></li><li><p><a href="ch_demo.html#ch_demo.id1_fetch.html">id1_fetch</a> ID1 and Entrez2 client</p></li></ul></li><li><p><a href="ch_demo.html#ch_demo.examples4">Examples from the Programming Manual</a></p><ul><li><p><a href="ch_demo.html#ch_demo.examples_1_3">applic.cpp</a></p></li><li><p><a href="ch_demo.html#ch_demo.examples_1_4">smart.cpp</a></p></li><li><p><a href="ch_ser.html#ch_ser.ctypeiter_cpp.html">ctypeiter.cpp</a></p></li><li><p><a href="ch_cgi.html#ch_cgi.cgi_cpp.html">justcgi.cpp</a></p></li><li><p><a href="ch_ser.html#ch_ser.xml2asn_cpp.html">xml2asn.cpp</a></p></li><li><p><a href="ch_ser.html#ch_ser.traverse_cpp.html">traverseBS.cpp</a></p></li><li><p><a href="ch_demo.html#ch_demo.prog_man_CARdemo_index.html">Web-CGI demo</a></p></li></ul></li><li><p>Test and Demo Programs in the C++ Tree</p><ul><li><p><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/source/src/app/asn2asn/asn2asn.cpp">asn2asn.cpp</a></p></li><li><p><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/source/src/cgi/test/cgitest.cpp">cgitest.cpp</a></p></li><li><p><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/source/src/cgi/demo/cgidemo.cpp">cgidemo.cpp</a></p></li><li><p><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/source/src/corelib/test/coretest.cpp">coretest.cpp</a></p></li></ul></li></ul></div></div><div class="sec" id="ch_demo.id1_fetch.html"><h2><span class="title">ID1_FETCH - the ID1 and Entrez2 client</span></h2><ul><li><p><a href="ch_demo.html#ch_demo.Synopsis">Synopsis</a></p></li><li><p><a href="ch_demo.html#ch_demo.id1_fetch.html_ref_Invocation">Invocation</a></p><ul><li><p><a href="ch_demo.html#ch_demo.id1_fetch.html_ref_fmt">Output Data Formats</a></p></li><li><p><a href="ch_demo.html#ch_demo.id1_fetch.html_ref_lt">Lookup Types</a></p></li><li><p><a href="ch_demo.html#ch_demo.id1_fetch.html_ref_maxplex">Output Complexity Levels</a></p></li><li><p><a href="ch_demo.html#ch_demo.id1_fetch.html_ref_flat">Flattened Sequence ID Format</a></p></li><li><p><a href="ch_demo.html#ch_demo.id1_fetch.html_ref_fasta">FASTA Sequence ID Format</a></p></li></ul></li><li><p><a href="ch_demo.html#ch_demo.id1_fetch.html_ref_Examples">Examples of Usage</a></p></li></ul><p><b>Location</b>: <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/source/src/app/id1_fetch/id1_fetch.cpp">c++/src/app/id1_fetch/id1_fetch.cpp</a> (compiled executable is <span class="nctnt ncbi-path">$NCBI/c++/Release/bin/id1_fetch</span> on NCBI systems)</p><div class="sec" id="ch_demo.Synopsis"><h3><span class="title">Synopsis:</span></h3><ul><li><p>Accept a list of sequences, specified either directly by ID or indirectly by an <a href="http://www.ncbi.nlm.nih.gov/Entrez">Entrez</a> query.</p></li><li><p>Produce more information about the sequences, either as data from the ID server or as <a href="http://www.ncbi.nlm.nih.gov/Entrez">Entrez</a> document summaries.</p></li></ul><p>This program corresponds to <span class="nctnt ncbi-app">idfetch</span> from the C Toolkit.</p></div><div class="sec" id="ch_demo.id1_fetch.html_ref_Invocation"><h3><span class="title">Invocation</span></h3><p>See <a href="ch_demo.html#ch_demo.T1">Table 1</a>.</p><div class="table" id="ch_demo.T1"><div class="caption"><p>Table 1. Invocation flags</p></div><table><tbody><tr><th align="left" valign="top" rowspan="1" colspan="1">Argument</th><th align="left" valign="top" rowspan="1" colspan="1">Value</th><th align="left" valign="top" rowspan="1" colspan="1">Effect</th></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-cmd">-h</span></td><td align="left" valign="top" rowspan="1" colspan="1"></td><td align="left" valign="top" rowspan="1" colspan="1">Print usage message and exit.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-cmd">-gi N</span></td><td align="left" valign="top" rowspan="1" colspan="1">integer</td><td align="left" valign="top" rowspan="1" colspan="1">GenInfo ID of sequence to look up.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-cmd">-fmt fmt</span></td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_demo.html#ch_demo.id1_fetch.html_ref_fmt">format type</a></td><td align="left" valign="top" rowspan="1" colspan="1">Output data format; default is <span class="nctnt ncbi-monospace">asn</span> (ASN.1 text).</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-cmd">-out file</span></td><td align="left" valign="top" rowspan="1" colspan="1">filename</td><td align="left" valign="top" rowspan="1" colspan="1">Write output to specified file rather than stdout.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-cmd">-log file</span></td><td align="left" valign="top" rowspan="1" colspan="1">filename</td><td align="left" valign="top" rowspan="1" colspan="1">Write errors and messages to specified file rather than stderr.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-cmd">-db str</span></td><td align="left" valign="top" rowspan="1" colspan="1">string</td><td align="left" valign="top" rowspan="1" colspan="1">Use specified database. <b>Mandatory for </b><a href="http://www.ncbi.nlm.nih.gov/Entrez">Entrez</a><b> queries</b>, where it is normally either Nucleotide or Protein. Also specifies satellite database for sequence-entry lookups.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-cmd">-ent N</span></td><td align="left" valign="top" rowspan="1" colspan="1">integer</td><td align="left" valign="top" rowspan="1" colspan="1">Dump specified subentity. Only relevant for sequence-entry lookups.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-cmd">-lt type</span></td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_demo.html#ch_demo.id1_fetch.html_ref_lt">lookup type</a></td><td align="left" valign="top" rowspan="1" colspan="1">Type of lookup; default is <span class="nctnt ncbi-monospace">entry</span> (sequence entry).</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-cmd">-in file</span></td><td align="left" valign="top" rowspan="1" colspan="1">filename</td><td align="left" valign="top" rowspan="1" colspan="1">Read sequence IDs from file rather than command line. May contain raw GI IDs, <a href="ch_demo.html#ch_demo.id1_fetch.html_ref_flat">flattened IDs</a>, and <a href="ch_demo.html#ch_demo.id1_fetch.html_ref_fasta">FASTA-format IDs</a>.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-cmd">-maxplex m</span></td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_demo.html#ch_demo.id1_fetch.html_ref_maxplex">complexity</a></td><td align="left" valign="top" rowspan="1" colspan="1">Maximum output complexity level; default is <span class="nctnt ncbi-monospace">entry</span> (entire entry).</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-cmd">-flat id</span></td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_demo.html#ch_demo.id1_fetch.html_ref_flat">flat ID</a></td><td align="left" valign="top" rowspan="1" colspan="1">Flattened ID of sequence to look up.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-cmd">-fasta id</span></td><td align="left" valign="top" rowspan="1" colspan="1"><a href="ch_demo.html#ch_demo.id1_fetch.html_ref_fasta">FASTA ID</a></td><td align="left" valign="top" rowspan="1" colspan="1"><a href="http://blast.ncbi.nlm.nih.gov/blast/fasta.shtml">FASTA</a>-style ID of sequence to look up.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-cmd">-query str</span></td><td align="left" valign="top" rowspan="1" colspan="1">string</td><td align="left" valign="top" rowspan="1" colspan="1">Generate ID list from specified <a href="http://www.ncbi.nlm.nih.gov/Entrez">Entrez</a> query.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-cmd">-qf file</span></td><td align="left" valign="top" rowspan="1" colspan="1">file</td><td align="left" valign="top" rowspan="1" colspan="1">Generate ID list from <a href="http://www.ncbi.nlm.nih.gov/Entrez">Entrez</a> query in specified file.</td></tr></tbody></table></div><p><span class="nctnt highlight">Note</span>: You must specify exactly one of the options indicating what to look up: <span class="nctnt ncbi-cmd">-gi, -in, -flat, -fasta, -query, -qf</span>.</p><div class="sec" id="ch_demo.id1_fetch.html_ref_fmt"><h4><span class="title">Output Data Formats</span></h4><p>The possible values of the <span class="nctnt ncbi-cmd">-fmt</span> argument are shown in <a href="ch_demo.html#ch_demo.T2">Table 2</a>.</p><div class="table" id="ch_demo.T2"><div class="caption"><p>Table 2. Output data formats</p></div><table><tbody><tr><th align="left" valign="top" rowspan="1" colspan="1">Value</th><th align="left" valign="top" rowspan="1" colspan="1">Format</th><th align="left" valign="top" rowspan="1" colspan="1">Comments</th></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">asn</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="http://www.ncbi.nlm.nih.gov/Sitemap/Summary/asn1.html">ASN.1</a> text (default)</td><td align="left" valign="top" rowspan="1" colspan="1"></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">asnb</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="http://www.ncbi.nlm.nih.gov/Sitemap/Summary/asn1.html">ASN.1</a> binary</td><td align="left" valign="top" rowspan="1" colspan="1"></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">docsum</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="http://www.ncbi.nlm.nih.gov/Entrez">Entrez</a> document summary</td><td align="left" valign="top" rowspan="1" colspan="1">Lookup type is irrelevant.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">fasta</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="http://blast.ncbi.nlm.nih.gov/blast/fasta.shtml">FASTA</a></td><td align="left" valign="top" rowspan="1" colspan="1">Produces state as simple text; produces history in tabular form.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">genbank</td><td align="left" valign="top" rowspan="1" colspan="1"><a href="http://www.ncbi.nlm.nih.gov/Genbank/index.html">GenBank</a><a href="http://www.ncbi.nlm.nih.gov/Sitemap/samplerecord.html">flat-file format</a></td><td align="left" valign="top" rowspan="1" colspan="1">Lookup type must be entry (default).</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">genpept</td><td align="left" valign="top" rowspan="1" colspan="1">GenPept flat-file format</td><td align="left" valign="top" rowspan="1" colspan="1">Lookup type must be entry (default).</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">quality</td><td align="left" valign="top" rowspan="1" colspan="1">Quality scores</td><td align="left" valign="top" rowspan="1" colspan="1">Lookup type must be entry (default); data not always available.</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">xml</td><td align="left" valign="top" rowspan="1" colspan="1">XML</td><td align="left" valign="top" rowspan="1" colspan="1">Isomorphic to ASN.1 output.</td></tr></tbody></table></div></div><div class="sec" id="ch_demo.id1_fetch.html_ref_lt"><h4><span class="title">Lookup Types</span></h4><p>The possible values of the <span class="nctnt ncbi-cmd">-lt</span> argument are shown in <a href="ch_demo.html#ch_demo.T3">Table 3</a>.</p><div class="table" id="ch_demo.T3"><div class="caption"><p>Table 3. Lookup types</p></div><table><tbody><tr><th align="left" valign="top" rowspan="1" colspan="1">Value</th><th align="left" valign="top" rowspan="1" colspan="1">Description</th></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">entry</td><td align="left" valign="top" rowspan="1" colspan="1">The actual sequence entry (default)</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">history</td><td align="left" valign="top" rowspan="1" colspan="1">Summary of changes to the sequence data</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">ids</td><td align="left" valign="top" rowspan="1" colspan="1">All of the sequence's IDs</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">none</td><td align="left" valign="top" rowspan="1" colspan="1">Just the GI ID</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">revisions</td><td align="left" valign="top" rowspan="1" colspan="1">Summary of changes to the sequence data or annotations</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">state</td><td align="left" valign="top" rowspan="1" colspan="1">The sequence's status</td></tr></tbody></table></div></div><div class="sec" id="ch_demo.id1_fetch.html_ref_maxplex"><h4><span class="title">Output Complexity Levels</span></h4><p>The possible values of the <span class="nctnt ncbi-cmd">-maxplex</span> argument are shown in <a href="ch_demo.html#ch_demo.T4">Table 4</a>.</p><div class="table" id="ch_demo.T4"><div class="caption"><p>Table 4. Maximum output complexity level values</p></div><table><tbody><tr><th align="left" valign="top" rowspan="1" colspan="1">Value</th><th align="left" valign="top" rowspan="1" colspan="1">Description</th></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">bioseq</td><td align="left" valign="top" rowspan="1" colspan="1">Just the bioseq of interest</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">bioseq-set</td><td align="left" valign="top" rowspan="1" colspan="1">Minimal bioseq-set</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">entry</td><td align="left" valign="top" rowspan="1" colspan="1">Entire entry (default)</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">nuc-prot</td><td align="left" valign="top" rowspan="1" colspan="1">Minimal nuc-prot</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">pub-set</td><td align="left" valign="top" rowspan="1" colspan="1">Minimal pub-set</td></tr></tbody></table></div></div><div class="sec" id="ch_demo.id1_fetch.html_ref_flat"><h4><span class="title">Flattened Sequence ID Format</span></h4><p>A flattened sequence ID has one of the following three formats, where square brackets [...] surround optional elements:</p><ul><li><p>type([name or locus][,[accession][,[release][,version]]])</p></li><li><p>type=accession[.version]</p></li><li><p>type:number</p></li></ul><p>The first format requires quotes in most shells.</p><p>The type is a number, indicating who assigned the ID, as follows:</p><ul><li><p>Local use</p></li><li><p>GenInfo backbone sequence ID</p></li><li><p>GenInfo backbone molecule type</p></li><li><p>GenInfo import ID</p></li><li><p><a href="http://www.ncbi.nlm.nih.gov/Genbank/index.html">GenBank</a></p></li><li><p><a href="http://www.embl-heidelberg.de">The European Molecular Biology Laboratory (EMBL)</a></p></li><li><p><a href="http://pir.georgetown.edu">The Protein Information Resource (PIR)</a></p></li><li><p><a href="http://www.ebi.ac.uk/swissprot">SWISS-PROT</a></p></li><li><p>A patent</p></li><li><p><a href="http://www.ncbi.nlm.nih.gov/projects/RefSeq">RefSeq</a></p></li><li><p>General database reference</p></li><li><p>GenInfo integrated database (GI)</p></li><li><p><a href="http://www.ddbj.nig.ac.jp">The DNA Data Bank of Japan (DDBJ)</a></p></li><li><p><a href="http://www.prf.or.jp">The Protein Research Foundation (PRF)</a></p></li><li><p><a href="http://www.rcsb.org/pdb">The Protein DataBase (PDB)</a></p></li><li><p>Third-party annotation to <a href="http://www.ncbi.nlm.nih.gov/Genbank/index.html">GenBank</a></p></li><li><p>Third-party annotation to <a href="http://www.embl-heidelberg.de">EMBL</a></p></li><li><p>Third-party annotation to <a href="http://www.ddbj.nig.ac.jp">DDBJ</a></p></li><li><p>TrEMBL</p></li></ul></div><div class="sec" id="ch_demo.id1_fetch.html_ref_fasta"><h4><span class="title">FASTA Sequence ID Format</span></h4><p>This format consists of a two- or three-letter tag indicating the ID's type, followed by one or more data fields, which are separated from the tag and each other by vertical bars (|). The vertical bar is a special character in most command-line shells, so command-line arguments containing ID's usually must be quoted. <a href="ch_demo.html#ch_demo.T5">Table 5</a> shows the specific possibilities.</p><div class="table" id="ch_demo.T5"><div class="caption"><p>Table 5. FASTA sequence ID format values</p></div><table><tbody><tr><th align="left" valign="top" rowspan="1" colspan="1">Type</th><th align="left" valign="top" rowspan="1" colspan="1">Format(s)Â <a href="ch_demo.html#ch_demo.TF.1"><sup>1</sup></a></th><th align="left" valign="top" rowspan="1" colspan="1">Example(s)</th></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">local</td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">lcl|integer</span><br /><span class="nctnt ncbi-monospace">lcl|string</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">lcl|123</span><br /><span class="nctnt ncbi-monospace">lcl|hmm271</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">GenInfo backbone seqid</td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">bbs|integer</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">bbs|123</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">GenInfo backbone moltype</td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">bbm|integer</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">bbm|123</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">GenInfo import ID</td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">gim|integer</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">gim|123</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><a href="http://www.ncbi.nlm.nih.gov/Genbank/index.html">GenBank</a></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">gb|accession|locus</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">gb|M73307|AGMA13GT</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><a href="http://www.embl-heidelberg.de">EMBL</a></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">emb|accession|locus</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">emb|CAM43271.1|</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><a href="http://pir.georgetown.edu">PIR</a></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">pir|accession|name</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">pir||G36364</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><a href="http://www.ebi.ac.uk/swissprot">SWISS-PROT</a></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">sp|accession|name</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">sp|P01013|OVAX_CHICK</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">patent</td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">pat|country|patent|sequence</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">pat|US|RE33188|1</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">pre-grant patent</td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">pgp|country|application-number|seq-number</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">pgp|EP|0238993|7</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><a href="http://www.ncbi.nlm.nih.gov/projects/RefSeq">RefSeq</a>Â <a href="ch_demo.html#ch_demo.TF.2"><sup>2</sup></a></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">ref|accession|name</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">ref|NM_010450.1|</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">general database reference</td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">gnl|database|integer</span><br /><span class="nctnt ncbi-monospace">gnl|database|string</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">gnl|taxon|9606</span><br /><span class="nctnt ncbi-monospace">gnl|PID|e1632</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">GenInfo integrated database</td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">gi|integer</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">gi|21434723</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><a href="http://www.ddbj.nig.ac.jp">DDBJ</a></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">dbj|accession|locus</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">dbj|BAC85684.1|</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><a href="http://www.prf.or.jp">PRF</a></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">prf|accession|name</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">prf||0806162C</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1"><a href="http://www.rcsb.org/pdb">PDB</a></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">pdb|entry|chain</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">pdb|1I4L|D</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">third-party <a href="http://www.ncbi.nlm.nih.gov/Genbank/index.html">GenBank</a></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">tpg|accession|name</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">tpg|BK003456|</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">third-party <a href="http://www.embl-heidelberg.de">EMBL</a></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">tpe|accession|name</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">tpe|BN000123|</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">third-party <a href="http://www.ddbj.nig.ac.jp">DDBJ</a></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">tpd|accession|name</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">tpd|FAA00017|</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">TrEMBL</td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">tr|accession|name</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">tr|Q90RT2|Q90RT2_9HIV1</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">genome pipelineÂ <a href="ch_demo.html#ch_demo.TF.3"><sup>3</sup></a></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">gpp|accession|name</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">gpp|GPC_123456789|</span></td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">named annotation trackÂ <a href="ch_demo.html#ch_demo.TF.3"><sup>3</sup></a></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">nat|accession|name</span></td><td align="left" valign="top" rowspan="1" colspan="1"><span class="nctnt ncbi-monospace">nat|AT_123456789.1|</span></td></tr></tbody></table><div class="table-wrap-foot"><div class="small" id="ch_demo.TF.1"><p><sup>1</sup> Spaces should not be present in ID's. It's okay to leave off the final vertical bar for most text ID types (such as gb) when the locus is absent; apart from that, vertical bars must be present even if an adjacent field is omitted.</p></div><div class="small" id="ch_demo.TF.2"><p><sup>2</sup> Some RefSeq accessions have additional letters following the underscore.  See the <a href="http://www.ncbi.nlm.nih.gov/RefSeq/key.html#accession">RefSeq accession format reference</a> for details.</p></div><div class="small" id="ch_demo.TF.3"><p><sup>3</sup> For NCBI internal use.</p></div></div></div></div></div><div class="sec" id="ch_demo.id1_fetch.html_ref_Examples"><h3><span class="title">Example Usage</span></h3><pre><span class="nctnt-pre ncbi-code">id1_fetch -query '5-delta4 isomerase' -lt none -db Nucleotide<br /><br />34<br /><br />id1_fetch -fmt genbank -gi 34<br /><br />LOCUS       BT3BHSD      1632 bp    mRNA            MAM       12-SEP-1993<br />DEFINITION  Bovine mRNA for 3 beta hydroxy-5-ene steroid dehydrogenase/delta<br />            5-delta4 isomerase (EC 1.1.1.145, EC 5.3.3.1).<br />ACCESSION   X17614<br />VERSION     X17614.1  GI:34<br />KEYWORDS    3 beta-hydroxy-delta5-steroid dehydrogenase;<br />            steroid delta-isomerase.<br />...<br />FEATURES             Location/Qualifiers<br />...<br />     CDS             105..1226<br />                     /codon_start=1<br />                     /transl_table=1<br />                     /function="3 beta-HSD (AA 1-373)"<br />                     /protein_id="CAA35615.1"<br />                     /db_xref="GI:35"<br />                     /translation="MAGWSCLVTGGGGFLGQRIICLLVEEKDLQEIRVLDKVFRPEVR<br />                     EEFSKLQSKIKLTLLEGDILDEQCLKGACQGTSVVIHTASVIDVRNAVPRETIMNVNV<br />                     KGTQLLLEACVQASVPVFIHTSTIEVAGPNSYREIIQDGREEEHHESAWSSPYPYSKK<br />                     LAEKAVLGANGWALKNGGTLYTCALRPMYIYGEGSPFLSAYMHGALNNNGILTNHCKF<br />                     SRVNPVYVGNVAWAHILALRALRDPKKVPNIQGQFYYISDDTPHQSYDDLNYTLSKEW<br />                     GFCLDSRMSLPISLQYWLAFLLEIVSFLLSPIYKYNPCFNRHLVTLSNSVFTFSYKKA<br />                     QRDLGYEPLYTWEEAKQKTKEWIGSLVKQHKETLKTKIH"<br />                     /db_xref="SWISS-PROT:P14893"<br />...<br />     1441 ggacagacaa ggtgatttgc tgcagctgct ggcaccaaaa tctcagtggc agattctgag<br />     1501 ttatttgggc ttcttgtaac ttcgagtttt gcctcttagt cccactttct ttgttaaatg<br />     1561 tggaagcatt tcttttaaaa gttcatattc cttcatgtag ctcaataaaa atgatcaaca<br />     1621 ttttcatgac tc<br />//<br /><br />id1_fetch -fmt genpept -gi 35<br /><br />LOCUS       CAA35615      373 aa                    MAM       12-SEP-1993<br />DEFINITION  Bovine mRNA for 3 beta hydroxy-5-ene steroid dehydrogenase/delta<br />            5-delta4 isomerase (EC 1.1.1.145, EC 5.3.3.1), and translated<br />            products.<br />ACCESSION   CAA35615<br />VERSION     CAA35615.1  GI:35<br />PID         g35<br />SOURCE      cow.<br />  ORGANISM  Bos taurus<br />            Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi;<br />            Mammalia; Eutheria; Cetartiodactyla; Ruminantia; Pecora; Bovoidea;<br />            Bovidae; Bovinae; Bos.<br />...<br />ORIGIN<br />        1 magwsclvtg gggflgqrii cllveekdlq eirvldkvfr pevreefskl qskikltlle<br />       61 gdildeqclk gacqgtsvvi htasvidvrn avpretimnv nvkgtqllle acvqasvpvf<br />      121 ihtstievag pnsyreiiqd greeehhesa wsspypyskk laekavlgan gwalknggtl<br />      181 ytcalrpmyi ygegspflsa ymhgalnnng iltnhckfsr vnpvyvgnva wahilalral<br />      241 rdpkkvpniq gqfyyisddt phqsyddlny tlskewgfcl dsrmslpisl qywlafllei<br />      301 vsfllspiyk ynpcfnrhlv tlsnsvftfs ykkaqrdlgy eplytweeak qktkewigsl<br />      361 vkqhketlkt kih<br />//<br /><br />id1_fetch -fmt fasta -gi 35 -maxplex bioseq<br /><br />&gt;emb|CAA35615.1||gi|35 Bovine mRNA for 3 beta hydroxy-5-ene steroid dehydrogenase/delta<br /> 5-delta4 isomerase (EC 1.1.1.145, EC 5.3.3.1), and translated products<br />MAGWSCLVTGGGGFLGQRIICLLVEEKDLQEIRVLDKVFRPEVREEFSKLQSKIKLTLLEGDILDEQCLK<br />GACQGTSVVIHTASVIDVRNAVPRETIMNVNVKGTQLLLEACVQASVPVFIHTSTIEVAGPNSYREIIQD<br />GREEEHHESAWSSPYPYSKKLAEKAVLGANGWALKNGGTLYTCALRPMYIYGEGSPFLSAYMHGALNNNG<br />ILTNHCKFSRVNPVYVGNVAWAHILALRALRDPKKVPNIQGQFYYISDDTPHQSYDDLNYTLSKEWGFCL<br />DSRMSLPISLQYWLAFLLEIVSFLLSPIYKYNPCFNRHLVTLSNSVFTFSYKKAQRDLGYEPLYTWEEAK<br />QKTKEWIGSLVKQHKETLKTKIH<br /><br />id1_fetch -lt ids -gi 35<br /><br />ID1server-back ::= ids {<br />    embl {<br />        accession "CAA35615",<br />        version 1<br />    },<br />    general {<br />        db "NCBI_EXT_ACC",<br />        tag str "FPAA037960"<br />    },<br />    gi 35<br />}<br /><br />id1_fetch -lt state -fasta 'emb|CAA35615' -fmt xml<br /><br />&lt;?xml version="1.0"?&gt;<br />&lt;!DOCTYPE ID1server-back PUBLIC "-//NCBI//NCBI ID1Access/EN" "NCBI_ID1Access.dtd"&gt;<br />&lt;ID1server-back&gt;<br />    &lt;ID1server-back_gistate&gt;40&lt;/ID1server-back_gistate&gt;<br />&lt;/ID1server-back&gt;<br /><br />id1_fetch -lt state -flat '5=CAA35615.1' -fmt asnb | od -t u1<br /><br />0000000 166 128 002 001 040 000 000<br />0000007<br /><br />id1_fetch -lt state -flat '5(,CAA35615)' -fmt fasta<br /><br />gi = 35, states: LIVE<br /><br />id1_fetch -lt history -flat '12:35' -fmt fasta<br /><br />GI  Loaded      DB    Retrieval No.<br />--  ------      --    -------------<br />35  03/08/1999  EMBL  274319<br /><br />id1_fetch -lt revisions -gi 35 -fmt fasta<br /><br />GI  Loaded      DB     Retrieval No.<br />--  ------      --     -------------<br />35  03/08/1999  EMBL   274319<br />35  06/06/1996  OLD02  84966<br />35  05/27/1995  OLDID  1524022<br />35  11/29/1994  OLDID  966346<br />35  08/31/1993  OLDID  426053<br />35  04/20/1993  OLDID  27<br /><br />id1_fetch -fmt quality -gi 13508865<br /><br />&gt;AL590146.2 Phrap Quality (Length: 121086, Min: 31, Max: 99)<br /> 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99<br /> 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99<br /> 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99<br /> 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99<br /> 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99<br />...<br /> 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99<br /> 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99<br /> 99 54 54 56 56 54 54 54 56 56 56 56 65 65 57 60 56 56 59 59<br /> 56 56 56 49 99 31 31 49 49 54 63 63 54 51 53 55 51 51 49 58<br /> 58 58 58 53 52 49 51 51 51 52 55 51 51 51 49 49 49 63 63 60<br /> 65 65 59 54 54 54 54 54 56 60 60 65 65 65 65 70 70 65 65 65<br /> 65 65 65 65 60 59 59 66 66 66 67 65 65 63 46 65 99 99 99 99<br /> 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99<br />...<br /> 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99<br /> 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99<br /> 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99<br /> 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99 99<br /> 99 99 99 99 99 99<br /></span></pre></div></div><div class="sec" id="ch_demo.examples4"><h2><span class="title">Examples from the Programming Manual</span></h2><ul><li><p><a href="ch_demo.html#ch_demo.examples_1_3">applic.cpp</a></p></li><li><p><a href="ch_demo.html#ch_demo.examples_1_4">smart.cpp</a></p></li><li><p><a href="ch_demo.html#ch_demo.prog_man_CARdemo_index.html">An Example of a Web-based CGI Application - Source Files</a></p><ul><li><p><a href="ch_demo.html#ch_demo.carcpp">car.cpp</a></p></li><li><p><a href="ch_demo.html#ch_demo.carhpp">car.hpp</a></p></li><li><p><a href="ch_demo.html#ch_demo.car_cgicpp">car_cgi.cpp</a></p></li><li><p><a href="ch_demo.html#ch_demo.Makefilecar_app">Makefile.car_app</a></p></li><li><p><a href="ch_demo.html#ch_demo.carhtml">car.html</a> </p></li></ul></li></ul><div class="sec" id="ch_demo.examples_1_3"><h3><span class="title">applic.cpp</span></h3><div class="box" id="idp60593408"><pre><span class="nctnt-pre ncbi-code"><br />// File name:   applic.cpp<br />// Description: Using the CNcbiApplication class with CNcbiDiag, CArgs<br />//              and CArgDescription classes<br /><br />#include &lt;corelib/ncbistd.hpp&gt;<br />#include &lt;corelib/ncbiutil.hpp&gt;<br />#include &lt;corelib/ncbiargs.hpp&gt;<br />#include &lt;corelib/ncbiapp.hpp&gt;<br />#include &lt;corelib/ncbienv.hpp&gt;<br /><br />USING_NCBI_SCOPE;<br /><br />class CTestApp : public CNcbiApplication {<br />public:<br />    virtual int Run();<br />};<br />int CTestApp::Run() {<br /><br />    auto_ptr&lt;CArgs&gt; args;<br /><br />    // create a CArgDescriptions object to constrain the input arguments;<br />    // Argument descriptions are added using:<br /><br />    // void AddKey(string&amp; name, string&amp; synopsis, string&amp; comment, EType, TFlags);<br />    // void AddOptionalKey(string&amp; name, string&amp; synopsis, string&amp; comment, EType,<br />    //                     string&amp; default, TFlags);<br />    // void AddFlag(string&amp; name, string&amp; comment);<br /><br />    {<br />       CArgDescriptions argDesc;<br /><br />       // Required arguments:<br />       argDesc.AddKey("n","integer","integer between 1 and 10",argDesc.eInteger);<br />       argDesc.AddKey("f","float","float between 0.0 and 1.0",argDesc.eDouble);<br />       argDesc.AddKey("i","inputFile","data file in",CArgDescriptions::eInputFile);<br /><br />       // optional arguments:<br />       argDesc.AddOptionalKey("l","logFile","log errors to &lt;logFile&gt;",<br />                              argDesc.eOutputFile);<br /><br />       // optional flags<br />       argDesc.AddFlag("it", "input text");<br />       argDesc.AddFlag("ib", "input binary");<br /><br />       try {<br />           args.reset(argDesc.CreateArgs(GetArguments()));<br />       }<br />       catch (exception&amp; e) {<br />           string a;<br />           argDesc.SetUsageContext(GetArguments()[0],<br />                                   "CArgDescriptions demo program");<br /><br />           cerr &lt;&lt; e.what() &lt;&lt; endl;<br />           cerr &lt;&lt; argDesc.PrintUsage(a);<br />           return (-1);<br />       }<br />    }<br /><br />    int   intIn   = (*args)["n"].AsInteger();<br />    float floatIn = (*args)["f"].AsDouble();<br />    string inFile = (*args)["i"].AsString();<br /><br />    // process optional args<br />    if ( args-&gt;Exest("l") ) {<br />        SetDiagStream(&amp;(*args)["l"].AsOutputFile());<br />    }<br /><br />    bool textIn = args-&gt;Exist("it");<br />    bool binIn = (*args)["ib"].AsBoolean();<br /><br />    if (! (textIn ^ binIn) ) {<br />       ERR_POST_X(1, Error &lt;&lt; "input type must be specified using -it or -ib");<br />    }<br /><br />    string InputType;<br />    if ( textIn ) {<br />      InputType = "text";<br />    } else if ( binIn ) {<br />      InputType = "binary";<br />    }<br /><br />    ERR_POST_X(2, Info &lt;&lt; "intIn = " &lt;&lt; intIn &lt;&lt; " floatIn = " &lt;&lt; floatIn<br />             &lt;&lt; " inFile = " &lt;&lt; inFile &lt;&lt; " input type = " &lt;&lt; InputType);<br /><br />    return 0;<br />}<br />int main(int argc, const char* argv[])<br />{<br />      CNcbiOfstream diag("moreApp.log");<br />      SetDiagStream(&amp;diag);<br /><br />      // Set the global severity level to Info<br />      SetDiagPostLevel(eDiag_Info);<br /><br />      CTestApp theTestApp;<br />      return theTestApp.AppMain(argc, argv);<br />}</span></pre></div></div><div class="sec" id="ch_demo.examples_1_4"><h3><span class="title">smart.cpp </span></h3><div class="box" id="idp60627024"><pre><span class="nctnt-pre ncbi-code"><br />// File name: smart.cpp<br />// Description: Memory management using auto_ptr versus CRef<br /><br />#include &lt;corelib/ncbiapp.hpp&gt;<br />#include &lt;corelib/ncbiobj.hpp&gt;<br /><br />USING_NCBI_SCOPE;<br /><br />class CTestApp : public CNcbiApplication {<br />public:<br />    virtual int Run(void);<br />};<br /><br />/////////////////////////////////////////////////////////////////<br />//<br />// 1. Install an auto_ptr to an int and make a copy - then try to<br />//    reference the value from the original auto_ptr.<br />//<br />// 2. Do the same thing using CRefs instead of auto_ptrs.<br />//<br />//////////////////////////////////////////////////////////////////<br /><br />int CTestApp::Run()<br />{<br />    auto_ptr&lt;int&gt; orig_ap;<br />    orig_ap.reset(new int(5));<br />    {<br />        auto_ptr&lt;int&gt; copy_ap = orig_ap;<br /><br />        if ( !orig_ap.get() ) {<br />            cout &lt;&lt; "orig_ap no longer exists - copy_ap = " &lt;&lt; *copy_ap &lt;&lt; endl;<br />        } else {<br />            cout &lt;&lt; "orig_ap = " &lt;&lt; *orig_ap &lt;&lt; ", copy_ap = "<br />                 &lt;&lt; *copy_ap &lt;&lt; endl;<br />        }<br />    }<br />    if ( orig_ap.get() ) {<br />        cout &lt;&lt; "orig_ap = " &lt;&lt; *orig_ap &lt;&lt; endl;<br />    }<br /><br />    CRef&lt; CObjectFor&lt;int&gt; &gt; orig(new  CObjectFor&lt;int&gt;);<br />    *orig = 5;<br />    {<br />        CRef&lt; CObjectFor&lt;int&gt; &gt; copy = orig;<br /><br />        if ( !orig ) {<br />            cout &lt;&lt; "orig no longer exists - copy = " &lt;&lt; (int&amp;) *copy &lt;&lt; endl;<br />        } else {<br />            cout &lt;&lt; "orig = " &lt;&lt; (int&amp;) *orig &lt;&lt; ", copy = "<br />                 &lt;&lt; (int&amp;) *copy &lt;&lt; endl;<br />        }<br />    }<br />    if ( orig ) {<br />        cout &lt;&lt; "orig = " &lt;&lt; (int&amp;) *orig &lt;&lt; endl;<br />    }<br />    return 0;<br />}<br /><br /><br />int main(int argc, const char* argv[])<br />{<br />    CTestApp theTestApp;<br />    return theTestApp.AppMain(argc, argv);<br />}<br /></span></pre></div></div><div class="sec" id="ch_demo.prog_man_CARdemo_index.html"><h3><span class="title">An Example of a Web-based CGI Application - Source Files</span></h3><div class="sec" id="ch_demo.carcpp"><h4><span class="title">car.cpp</span></h4><div class="box" id="idp60651616"><pre><span class="nctnt-pre ncbi-code"><br />// File name: car.cpp<br />// Description: Implement the CCarAttr class<br /><br />#include "car.hpp"<br /><br />BEGIN_NCBI_SCOPE<br /><br />/////////////////////////////////////////////////////////////////////////////<br />//  CCarAttr::<br /><br />set&lt;string&gt; CCarAttr::sm_Features;<br />set&lt;string&gt; CCarAttr::sm_Colors;<br /><br /><br />CCarAttr::CCarAttr(void)<br />{<br />    // make sure there is only one instance of this class<br />    if ( !sm_Features.empty() ) {<br />        _TROUBLE;<br />        return;<br />    }<br /><br />    // initialize static data<br />    sm_Features.insert("Air conditioning");<br />    sm_Features.insert("CD Player");<br />    sm_Features.insert("Four door");<br />    sm_Features.insert("Power windows");<br />    sm_Features.insert("Sunroof");<br /><br />    sm_Colors.insert("Black");<br />    sm_Colors.insert("Navy");<br />    sm_Colors.insert("Silver");<br />    sm_Colors.insert("Tan");<br />    sm_Colors.insert("White");<br />}<br /><br />// dummy instance of CCarAttr -- to provide initialization of<br />// CCarAttr::sm_Features and CCarAttr::sm_Colors<br />static CCarAttr s_InitCarAttr;<br /><br />END_NCBI_SCOPE</span></pre></div></div><div class="sec" id="ch_demo.carhpp"><h4><span class="title">car.hpp</span></h4><div class="box" id="idp60666432"><pre><span class="nctnt-pre ncbi-code"><br />// File name: car.hpp<br />// Description: Define the CCar and CCarAttr classes<br /><br />#ifndef CAR_HPP<br />#define CAR_HPP<br /><br />#include &lt;coreilib/ncbistd.hpp&gt;<br />#include &lt;set&gt;<br /><br />BEGIN_NCBI_SCOPE<br /><br />//////////////////////<br />//  CCar<br /><br />class CCar<br />{<br />public:<br />    CCar(unsigned base_price = 10000)  { m_Price = base_price; }<br /><br />    bool HasFeature(const string&amp;amp; feature_name) const<br />        { return m_Features.find(feature_name) != m_Features.end(); }<br />    void AddFeature(const string&amp; feature_name)<br />        { m_Features.insert(feature_name); }<br /><br />    void   SetColor(const string&amp; color_name)  { m_Color = color_name; }<br />    string GetColor(void) const                { return m_Color; }<br /><br />    const set&lt;string&gt;&amp; GetFeatures() const { return m_Features; }<br />    unsigned GetPrice(void) const<br />        { return m_Price + 1000 * m_Features.size(); }<br /><br />private:<br />    set&lt;string&gt;  m_Features;<br />    string       m_Color;<br />    unsigned     m_Price;<br />};<br /><br /><br /><br />//////////////////////<br />//  CCarAttr -- use a dummy all-static class to store available car attributes<br /><br />class CCarAttr {<br />public:<br />    CCarAttr(void);<br />    static const set&lt;string&gt;&amp; GetFeatures(void) { return sm_Features; }<br />    static const set&lt;string&gt;&amp; GetColors  (void) { return sm_Colors; }<br />private:<br />    static set&lt;string&gt; sm_Features;<br />    static set&lt;string&gt; sm_Colors;<br />};<br /><br /><br />END_NCBI_SCOPE<br /><br />#endif  /* CAR__HPP */</span></pre></div></div><div class="sec" id="ch_demo.car_cgicpp"><h4><span class="title">car_cgi.cpp</span></h4><div class="box" id="idp60686368"><pre><span class="nctnt-pre ncbi-code"><br />// File name: car_cgi.cpp<br />// Description: Implement the CCarCgi class and function main<br /><br />#include &lt;cgi/cgiapp.hpp&gt;<br />#include &lt;cgi/cgictx.hpp&gt;<br />#include &lt;html/html.hpp&gt;<br />#include &lt;html/page.hpp&gt;<br /><br />#include "car.hpp"<br /><br />USING_NCBI_SCOPE;<br /><br />/////////////////////////////////////////////////////////////////////////////<br />//  CCarCgi::  declaration<br /><br />class CCarCgi : public CCgiApplication<br />{<br />public:<br />    virtual int ProcessRequest(CCgiContext&amp; ctx);<br /><br />private:<br />    CCar* CreateCarByRequest(const CCgiContext&amp; ctx);<br /><br />    void PopulatePage(CHTMLPage&amp; page, const CCar&amp; car);<br /><br />    static CNCBINode* ComposeSummary(const CCar&amp; car);<br />    static CNCBINode* ComposeForm   (const CCar&amp; car);<br />    static CNCBINode* ComposePrice  (const CCar&amp; car);<br /><br />    static const char sm_ColorTag[];<br />    static const char sm_FeatureTag[];<br />};<br /><br /><br />/////////////////////////////////////////////////////////////////////////////<br />//  CCarCgi::  implementation<br /><br /><br />const char CCarCgi::sm_ColorTag[]   = "color";<br />const char CCarCgi::sm_FeatureTag[] = "feature";<br /><br /><br />int CCarCgi::ProcessRequest(CCgiContext&amp; ctx)<br />{<br />    // Create new "car" object with the attributes retrieved<br />    // from the CGI request parameters<br />    auto_ptr&lt;CCar&gt; car;<br />    try {<br />        car.reset( CreateCarByRequest(ctx) );<br />    } catch (exception&amp; e) {<br />        ERR_POST_X(1, "Failed to create car: " &lt;&lt; e.what());<br />        return 1;<br />    }<br /><br />    // Create an HTML page (using the template file "car.html")<br />    CRef&lt;CHTMLPage&gt; page;<br />    try {<br />        page = new CHTMLPage("Car", "car.html");<br />    } catch (exception&amp; e) {<br />        ERR_POST_X(2, "Failed to create the Car HTML page: " &lt;&lt; e.what());<br />        return 2;<br />    }<br /><br />    // Register all substitutions for the template parameters &lt;@XXX@&gt;<br />    // (fill them out using the "car" attributes)<br />    try {<br />        PopulatePage(*page, *car);<br />    } catch (exception&amp; e) {<br />        ERR_POST_X(3, "Failed to populate the Car HTML page: " &lt;&lt; e.what());<br />        return 3;<br />    }<br /><br />    // Compose and flush the resultant HTML page<br />    try {<br />        const CCgiResponse&amp; response = ctx.GetResponse();<br />        response.WriteHeader();<br />        page-&gt;Print(response.out(), CNCBINode::eHTML);<br />        response.Flush();<br />    } catch (exception&amp; e) {<br />        ERR_POST_X(4, "Failed to compose and send the Car HTML page: " &lt;&lt; e.what());<br />        return 4;<br />    }<br /><br />    return 0;<br />}<br /><br /><br />CCar* CCarCgi::CreateCarByRequest(const CCgiContext&amp; ctx)<br />{<br />    auto_ptr&lt;CCar&gt; car(new CCar);<br /><br />    // Get the list of CGI request name/value pairs<br />    const TCgiEntries&amp; entries = ctx.GetRequest().GetEntries();<br /><br />    TCgiEntriesCI it;<br /><br />    // load the car with selected features<br />    pair&lt;TCgiEntriesCI,TCgiEntriesCI&gt; feature_range =<br />        entries.equal_range(sm_FeatureTag);<br />    for (it = feature_range.first;  it != feature_range.second;  ++it) {<br />        car-&gt;AddFeature(it-&gt;second);<br />    }<br /><br />    // color<br />    if ((it = entries.find(sm_ColorTag)) != entries.end()) {<br />        car-&gt;SetColor(it-&gt;second);<br />    } else {<br />        car-&gt;SetColor(*CCarAttr::GetColors().begin());<br />    }<br /><br />    return car.release();<br />}<br /><br /><br />    /************ Create a form with the following structure:<br />      &lt;form&gt;<br />        &lt;table&gt;<br />          &lt;tr&gt;<br />            &lt;td&gt; (Features) &lt;/td&gt;<br />            &lt;td&gt; (Colors)   &lt;/td&gt;<br />            &lt;td&gt; (Submit)   &lt;/td&gt;<br />          &lt;/tr&gt;<br />        &lt;/table&gt;<br />      &lt;/form&gt;<br />    ********************/<br /><br />CNCBINode* CCarCgi::ComposeForm(const CCar&amp; car)<br />{<br />    set&lt;string&gt;::const_iterator it;<br /><br />    CRef&lt;CHTML_table&gt; Table = new CHTML_table();<br />    Table-&gt;SetCellSpacing(0)-&gt;SetCellPadding(4)<br />        -&gt;SetBgColor("#CCCCCC")-&gt;SetAttribute("border", "0");<br /><br />    CRef&lt;CHTMLNode&gt; Row = new CHTML_tr();<br /><br />    // features (check boxes)<br />    CRef&lt;CHTMLNode&gt; Features = new CHTML_td();<br />    Features-&gt;SetVAlign("top")-&gt;SetWidth("200");<br />    Features-&gt;AppendChild(new CHTMLText("Options: &lt;br&gt;"));<br /><br />    for (it = CCarAttr::GetFeatures().begin();<br />         it != CCarAttr::GetFeatures().end();  ++it) {<br />        Features-&gt;AppendChild<br />            (new CHTML_checkbox<br />             (sm_FeatureTag, *it, car.HasFeature(*it), *it));<br />        Features-&gt;AppendChild(new CHTML_br());<br />    }<br /><br />    // colors (radio buttons)<br />    CRef&lt;CHTMLNode&gt; Colors = new CHTML_td();<br />    Colors-&gt;SetVAlign("top")-&gt;SetWidth("128");<br />    Colors-&gt;AppendChild(new CHTMLText("Color: &lt;br&gt;"));<br /><br />    for (it = CCarAttr::GetColors().begin();<br />         it != CCarAttr::GetColors().end();  ++it) {<br />            Colors-&gt;AppendChild<br />                (new CHTML_radio<br />                 (sm_ColorTag, *it, !NStr::Compare(*it, car.GetColor()), *it));<br />            Colors-&gt;AppendChild(new CHTML_br());<br />    }<br /><br />    Row-&gt;AppendChild(&amp;*Features);<br />    Row-&gt;AppendChild(&amp;*Colors);<br />    Row-&gt;AppendChild<br />        ((new CHTML_td())-&gt;AppendChild<br />         (new CHTML_submit("submit", "submit")));<br />    Table-&gt;AppendChild(&amp;*Row);<br /><br />    // done<br />    return (new CHTML_form("car.cgi", CHTML_form::eGet))-&gt;AppendChild(&amp;*Table);<br />}<br /><br /><br />CNCBINode* CCarCgi::ComposeSummary(const CCar&amp; car)<br />{<br />    string summary = "You have ordered a " + car.GetColor() + " model";<br /><br />    if ( car.GetFeatures().empty() ) {<br />        summary += " with no additional features.&lt;br&gt;";<br />        return new CHTMLText(summary);<br />    }<br /><br />    summary += " with the following features:&lt;br&gt;";<br />    CRef&lt;CHTML_ol&gt; ol = new CHTML_ol();<br /><br />    for (set&lt;string&gt;::const_iterator i = car.GetFeatures().begin();<br />         i != car.GetFeatures().end();  ++i) {<br />        ol-&gt;AppendItem(*i);<br />    }<br />    return (new CHTMLText(summary))-&gt;AppendChild((CNodeRef&amp;)ol);<br />}<br /><br /><br />CNCBINode* CCarCgi::ComposePrice(const CCar&amp; car)<br />{<br />    return<br />        new CHTMLText("Total price:  $" + NStr::UIntToString(car.GetPrice()));<br />}<br /><br /><br />void CCarCgi::PopulatePage(CHTMLPage&amp; page, const CCar&amp; car)<br />{<br />    page.AddTagMap("FORM",     ComposeForm    (car));<br />    page.AddTagMap("SUMMARY",  ComposeSummary (car));<br />    page.AddTagMap("PRICE",    ComposePrice   (car));<br />}<br /><br /><br /><br />/////////////////////////////////////////////////////////////////////////////<br />//  MAIN<br /><br /><br />int main(int argc, char* argv[])<br />{<br />    SetDiagStream(&amp;NcbiCerr);<br />    return CCarCgi().AppMain(argc, argv);<br />}<br /></span></pre></div></div><div class="sec" id="ch_demo.Makefilecar_app"><h4><span class="title">Makefile.car_app</span></h4><div class="box" id="idp60757808"><pre><span class="nctnt-pre ncbi-code"><br /># Makefile:  /home/zimmerma/car/Makefile.car_app<br /># This file was originally generated by shell script "new_project"<br /><br />###  PATH TO A PRE-BUILT C++ TOOLKIT  ###<br />builddir = /netopt/ncbi_tools/c++/GCC-Debug/build<br /># builddir = $(NCBI)/c++/Release/build<br /><br /><br />###  DEFAULT COMPILATION FLAGS  -- DON'T EDIT OR MOVE THESE 4 LINES !!!  ###<br />include $(builddir)/Makefile.mk<br />srcdir = .<br />BINCOPY = @:<br />LOCAL_CPPFLAGS = -I.<br /><br /><br />#############################################################################<br />###  EDIT SETTINGS FOR THE DEFAULT (APPLICATION) TARGET HERE              ###<br />APP = car.cgi<br />SRC = car car_cgi<br /><br /># PRE_LIBS = $(NCBI_C_LIBPATH) .....<br />LIB        = xhtml xcgi xncbi<br /><br /># LIB      = xser xhtml xcgi xncbi xconnect<br /># LIBS     = $(NCBI_C_LIBPATH) -lncbi $(NETWORK_LIBS) $(ORIG_LIBS)<br /><br /># CPPFLAGS = $(ORIG_CPPFLAGS) $(NCBI_C_INCLUDE)<br /># CFLAGS   = $(ORIG_CFLAGS)<br /># CXXFLAGS = $(ORIG_CXXFLAGS)<br /># LDFLAGS  = $(ORIG_LDFLAGS)<br />#                                                                         ###<br />#############################################################################<br /><br />###  APPLICATION BUILD RULES  -- DON'T EDIT OR MOVE THIS LINE !!!  ###<br />include $(builddir)/Makefile.app<br /><br /><br />###  PUT YOUR OWN ADDITIONAL TARGETS (MAKE COMMANDS/RULES) BELOW HERE  ###</span></pre></div></div><div class="sec" id="ch_demo.carhtml"><h4><span class="title">car.html</span></h4><pre><span class="nctnt-pre ncbi-code">&lt;html&gt;<br />&lt;head&gt;<br />&lt;title&gt;Automobile Order Form&lt;/title&gt;<br />&lt;/head&gt;<br />&lt;body&gt;<br />&lt;h1&gt;Place your order here&lt;/h1&gt;<br />&lt;@FORM@&gt;<br />&lt;@SUMMARY@&gt;<br />&lt;@PRICE@&gt;<br />&lt;/body&gt;<br />&lt;/html&gt;</span></pre></div></div></div></div></div><div class="col four_col last"><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>In this page</h3></div></div><div class="portlet_content"><ul><li><a href="#ch_demo.id1_fetch.html">ID1_FETCH - the ID1 and Entrez2 client</a></li><li><a href="#ch_demo.examples4">Examples from the Programming Manual</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Download</h3></div></div><div class="portlet_content"><ul><li><a href="pdf/ch_demo.pdf">PDF version of this page</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Search</h3></div></div><div class="portlet_content"><form name="frmSymbolSearch" onsubmit="return SymbolSearch('toolkit')" action="/" id="frmSymbolSearch"><table><tbody><tr align="center" valign="middle"><td colspan="2" nowrap="nowrap"><input id="search_box" name="__symbol" style="vertical-align:                                          middle; line-height:normal " onkeypress="SymbolSearchKeyPress('toolkit',event)" type="text" />Â Â  
					<label for="search_box" class="offscreen_noflow">Search term</label><button type="submit" value="Search" style="vertical-align:                                          middle; line-height:normal;" onclick="javasrcipt:SymbolSearch('toolkit');">Search</button></td></tr><tr class="internal-only"><th style="text-align:left">Public</th><th style="text-align:left">Internal</th></tr><tr><td><input id="pToolkitAll" name="__symboloc" type="radio" checked="checked" /><label for="pToolkitAll" title="Search in the C/C++ source code and in this Book">All</label></td><td class="internal-only"><input id="iToolkitAll" name="__symboloc" type="radio" /><label for="iToolkitAll" title="Search in the C/C++ source code,  this Book, and Wiki">All</label></td></tr><tr><td><input id="pLXR" name="__symboloc" type="radio" /><label for="pLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td><td class="internal-only"><input id="iLXR" name="__symboloc" type="radio" /><label for="iLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td></tr><tr><td><input id="pLib" name="__symboloc" type="radio" /><label for="pLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td><td class="internal-only"><input id="iLib" name="__symboloc" type="radio" /><label for="iLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td></tr></tbody></table></form></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Source Browsers</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">LXR</a></li><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">Doxygen</a></li><li class="internal-only">LXR: <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">public</a> / <a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/lxr/ident">internal</a></li><li class="internal-only">Doxygen: <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml">public</a> / <a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/doxyhtml">internal</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>SVN Source Repository</h3></div></div><div class="portlet_content"><ul><li><a href="http://www.ncbi.nlm.nih.gov/bookshelf/br.fcgi?book=toolkit&amp;part=ch_getcode_svn#ch_getcode_svn.code_retrieval">Server</a></li><li><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/c++">Web
                            (dev)</a></li><li><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/internal/c++">Web
                            (extra)</a></li><li><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/production/components">Web (prod)</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Resources</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/wiki-private/CxxToolkit">Wiki</a></li><li class="internal-only"><a href="http://jira/browse/CXX">JIRA</a></li><li class="internal-only"><a href="http://intranet/ieb/ToolBox/STAT/test_stat/test_stat_ext.cgi"><span title="View results of testsuite runs on the automatic C++ Toolkit builds">Testsuite</span></a></li><li><a href="release_notes.html">Release Notes</a></li><li><a href="ch_getcode_svn.html#ch_getcode_svn.ftp_download">Download</a></li><li><a href="pdf/TOC.pdf">PDF of the book</a></li><li><a href="ch_style.html">Coding Style</a></li><li><a href="ch_faq.html#ch_faq.mailing_lists">Mailing Lists</a></li><li><a href="mailto:CPP-CORE@NCBI.NLM.NIH.GOV"><span title="Contact C++ Toolkit group">Help and Support</span></a></li></ul></div></div></div></div></div></div></body></html>
