<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><title>Release Notes (December 31, 2005) - The NCBI C++ Toolkit Book [Internet] - NCBI</title><meta name="description" content="A comprehensive manual on the NCBI C++ toolkit, including its design and development framework, a C++ library reference, software examples and demos, FAQs and release notes." /><meta name="robots" content="index,nofollow,noarchive" /><meta name="page_id" content="release_notes_12_31_2005" /><meta name="page_type" content="chapter" /><meta name="ncbipopper" content="triggerPosition:'bottom left',destPosition:'top left',openEvent:'click',closeEvent:'click'" /><script type="text/javascript" src="js/jig/1.7/js/jig.min.js"></script><script type="text/javascript" src="js/toolkitSymbolSearch.js"></script><link type="text/css" rel="stylesheet" href="css/ptbooks.css" /><link type="text/css" rel="stylesheet" href="css/toolkit.css" /></head><body><div class="grid no_max_width"><div class="col twelve_col nomargin shadow"><div class="content"><div class="col eight_col"><div class="nvgtn"><a href="#__nav_cntnt" class="jig-ncbipopper">Contents ▾</a><div style="display:none;width:700px;height:500px;overflow:auto;" id="__nav_cntnt"><a class="right" href="toc.html">Table of Contents Page</a><ul class="simple-list"><li class="half_rhythm"><a href="toolkit.fm.html">Book Information</a></li><li class="half_rhythm"><a href="part1.html">Part 1. Overview</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_intro.html">1. Introduction to the C++ Toolkit</a></li><li class="half_rhythm"><a href="ch_start.html">2. Getting Started</a></li></ul></li><li class="half_rhythm"><a href="part2.html">Part 2. Development Framework</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_getcode_svn.html">3. Retrieve the Source Code (FTP and Subversion)</a></li><li class="half_rhythm"><a href="ch_config.html">4. Configure, Build, and Use the Toolkit</a></li><li class="half_rhythm"><a href="ch_build.html">5. Working with Makefiles</a></li><li class="half_rhythm"><a href="ch_proj.html">6. Project Creation and Management</a></li><li class="half_rhythm"><a href="ch_style.html">7. Programming Policies and Guidelines</a></li></ul></li><li class="half_rhythm"><a href="part3.html">Part 3. C++ Toolkit Library Reference</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_core.html">8. Portability, Core Functionality and Application Framework</a></li><li class="half_rhythm"><a href="ch_conn.html">9. Networking and IPC</a></li><li class="half_rhythm"><a href="ch_dbapi.html">10. Database Access - SQL, Berkley DB</a></li><li class="half_rhythm"><a href="ch_cgi.html">11. CGI and Fast-CGI </a></li><li class="half_rhythm"><a href="ch_html.html">12. HTML</a></li><li class="half_rhythm"><a href="ch_ser.html">13. Data Serialization (ASN.1, XML)</a></li><li class="half_rhythm"><a href="ch_datamod.html">14. Biological Sequence Data Model</a></li><li class="half_rhythm"><a href="ch_objmgr.html">15. Biological Object Manager</a></li><li class="half_rhythm"><a href="ch_blast.html">16. BLAST API</a></li><li class="half_rhythm"><a href="ch_dataaccess.html">17. Access to NCBI data</a></li><li class="half_rhythm"><a href="ch_algoalign.html">18. Biological Sequence Alignment</a></li><li class="half_rhythm"><a href="ch_gui.html">19. GUI and Graphics</a></li><li class="half_rhythm"><a href="ch_boost.html">20. Using the Boost Unit Test Framework</a></li></ul></li><li class="half_rhythm"><a href="part4.html">Part 4. Wrappers for 3rd-Party Packages</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_xmlwrapp.html">21. XmlWrapp (XML parsing and handling, XSLT, XPath)</a></li></ul></li><li class="half_rhythm"><a href="part5.html">Part 5. Software</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_debug.html">22. Debugging, Exceptions, and Error Handling</a></li><li class="half_rhythm"><a href="ch_grid.html">23. Distributed Computing</a></li><li class="half_rhythm"><a href="ch_app.html">24. Applications</a></li><li class="half_rhythm"><a href="ch_demo.html">25. Examples and Demos</a></li><li class="half_rhythm"><a href="ch_res.html">26. C Toolkit Resources for C++ Toolkit Users</a></li></ul></li><li class="half_rhythm"><a href="part6.html">Part 6. Help and Support</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_browse.html">27. NCBI C++ Toolkit Source Browser</a></li><li class="half_rhythm"><a href="ch_devtools.html">28. Software Development Tools</a></li><li class="half_rhythm"><a href="ch_xmlauthor.html">29. XML Authoring using Word</a></li><li class="half_rhythm"><a href="ch_faq.html">30. FAQs, Useful Documentation Links, and Mailing Lists</a></li></ul></li><li class="half_rhythm"><a href="part7.html">Part 7. Library and Applications Configuration</a><ul class="simple-list"><li class="half_rhythm"><a href="ch_libconfig.html">31. Library Configuration</a></li></ul></li><li class="half_rhythm"><a href="part8.html">Release Notes</a><ul class="simple-list"><li class="half_rhythm"><a href="release_notes.html">Release Notes (Version 9, May 2012)</a></li><li class="half_rhythm"><a href="release_notes_7-05_2011.html">Release Notes (Version 7, May 2011)</a></li><li class="half_rhythm"><a href="release_notes_06_29_2010.html">Release Notes (June, 2010)</a></li><li class="half_rhythm"><a href="release_notes_05_15_2009.html">Release Notes (May, 2009)</a></li><li class="half_rhythm"><a href="release_notes_12_24_2008.html">Release Notes (December, 2008)</a></li><li class="half_rhythm"><a href="release_notes_12_31_2008.html">Release Notes (March, 2008)</a></li><li class="half_rhythm"><a href="release_notes_08_27_2007.html">Release Notes (August, 2007)</a></li><li class="half_rhythm"><a href="release_notes_03_12_2007.html">Release Notes (March, 2007)</a></li><li class="half_rhythm"><a href="release_notes_08_14_2006.html">Release Notes (August, 2006)</a></li><li class="half_rhythm"><a href="release_notes_04_30_2006.html">Release Notes (April 30, 2006)</a></li><li class="half_rhythm"><a href="release_notes_12_31_2005.html">Release Notes (December 31, 2005)</a></li><li class="half_rhythm"><a href="release_notes_10_03_2005.html">Release Notes (August, 2005)</a></li><li class="half_rhythm"><a href="release_notes_05_05_2005.html">Release Notes (April, 2005)</a></li><li class="half_rhythm"><a href="release_notes_03_09_2005.html">Release Notes (February, 2005)</a></li><li class="half_rhythm"><a href="release_notes_11_22_2004.html">Release Notes (November 22, 2004)</a></li><li class="half_rhythm"><a href="release_notes_10_2_2004.html">Release Notes (October 2, 2004)</a></li><li class="half_rhythm"><a href="release_notes_7_8_2004.html">Release Notes (July 8, 2004)</a></li><li class="half_rhythm"><a href="release_notes_april_16_2004.html">Release Notes (April 16, 2004)</a></li><li class="half_rhythm"><a href="release_notes_12_08_2003.html">Release Notes (December 8, 2003)</a></li><li class="half_rhythm"><a href="release_notes_08_01_2003.html">Release Notes (August 1, 2003)</a></li></ul></li><li class="half_rhythm"><a href="app1.appendix1.html">Appendix - Books and Styles</a></li></ul></div><div class="pagination"><a class="pagelink active prev" title="Previous page" href="release_notes_04_30_2006.html">&lt; Prev</a><a class="pagelink active next" title="Next page" href="release_notes_10_03_2005.html">Next &gt;</a></div></div><div class="meta_content"><h1><span class="title">Release Notes (December 31, 2005)</span></h1><p class="small">Created: December 31, 2005; Last Update: February 1, 2006.</p></div><div class="body_content"><ul><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.Download_Locati">Download Location</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.Source_Archive_">Source Archive Contents</a></p><ul><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.Source_Code_Arc">Source Code Archives</a></p></li></ul></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.New_Development">New Development</a></p><ul><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.CORELIB__Portab">CORELIB — Portability and Application Framework</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.CONNECT__Data_s">CONNECT — Data streaming, Networking, and Dispatching</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.UTIL__Miscellan">UTIL — Miscellaneous Low-Level APIs</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.SERIAL__Data_Se">SERIAL — Data Serialization (ASN.1, XML)</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.CGI__CGI_and_Fa">CGI — CGI and Fast-CGI Application Framework</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.HTML__HTML_Gene">HTML — HTML Generation Library</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.BerkeleyDB_API_">BerkeleyDB API (bdb) — Much Enriched C++ API Based On BerkeleyDB</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.DBAPI__SQL_Data">DBAPI — Generic SQL Database Connectivity</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.PYTHON_database">PYTHON database module based on DBAPI</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.ALGO_ALIGN_SPLI">ALGO/ALIGN — Generic Alignment Algorithms</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.ALNMGR__Biosequ">ALNMGR — Bio-sequence Alignment Manager</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.BLAST">BLAST</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.BIOOBJECTS__Bio">BIO-OBJECTS — Bio-Object Specific Utility Functions (Not Involving OM++)</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.BIOTOOLS">BIO-TOOLS</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.LDS__Local_data">LDS - Local data storage</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.OM__Object_Mana">OM++ — Object Manager — For Retrieving and Processing Bio-Objects</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.OM_LOADERS_READ">OM++ LOADERS/READERS — Data Retrieval Libraries for OM++</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.OM_DEMO_program">OM++ DEMO program (objmgr_demo)</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.OM_DEMO_program">BUILD FRAMEWORK (UNIX)</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.PTB__Project_Tr">PTB — Project Tree Builder for MSVC++ .NET</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.APPLICATIONS">APPLICATIONS</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.GRID_DISTRIBUTE">GRID (DISTRIBUTED COMPUTING) FRAMEWORK</a></p></li></ul></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.Documentation">Documentation</a></p><ul><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.Document_Locati">Document Location</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.Document_Conten">Document Content</a></p></li></ul></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.Building_on_the">Building on the MacOS</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.Platforms_OSs__">Platforms (OS's, compilers used inside NCBI)</a></p><ul><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.T2">Unix</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.T3">MS Windows</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.T4">Mac OS X</a></p></li></ul></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.Caveats_and_Hin">Caveats and Hints</a></p><ul><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.MacOS_10X___Cod">MacOS 10.X / CodeWarrior 9.2</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.MacOS_102_GCC_3">MacOS 10.2/GCC 3.3</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.GCC_295">GCC 2.95</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.GCC_304">GCC 3.0.4</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.GCC_33">GCC 3.3</a></p></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.GCC_34x__40x">GCC 3.4.x, 4.0.x</a></p></li></ul></li><li><p><a href="release_notes_12_31_2005.html#release_notes_12_31_2005.Last_Updated">Last Updated</a></p></li></ul><div class="sec" id="release_notes_12_31_2005.Download_Locati"><h2><span class="title">Download Location</span></h2><p>
			<a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2005/Dec_31_2005">ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/2005/Dec_31_2005/</a>
		</p></div><div class="sec" id="release_notes_12_31_2005.Source_Archive_"><h2><span class="title">Source Archive Contents</span></h2><div class="sec" id="release_notes_12_31_2005.Source_Code_Arc"><h3><span class="title">Source Code Archives</span></h3><ul><li><p>
						<a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/CURRENT/ncbi_cxx--Dec_31_2005.tar.gz"><span class="nctnt ncbi-path">ncbi_cxx--Dec_31_2005.tar.gz</span></a> — for UNIX'es (see the list of UNIX flavors below) and MacOSX/GCC</p></li><li><p>
						<a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/CURRENT/ncbi_cxx--Dec_31_2005.gtar.gz"><span class="nctnt ncbi-path">ncbi_cxx--Dec_31_2005.gtar.gz</span></a> — for UNIX'es (see the list of UNIX flavors below) and MacOSX/GCC</p></li><li><p>
						<a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/CURRENT/ncbi_cxx--Dec_31_2005.exe"><span class="nctnt ncbi-path">ncbi_cxx--Dec_31_2005.exe</span></a>  — for MS-Windows / MSVC++ 7.1 (self-extracting)</p></li><li><p>
						<a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/CURRENT/ncbi_cxx--Dec_31_2005.zip"><span class="nctnt ncbi-path">ncbi_cxx--Dec_31_2005.zip</span></a> — for MS-Windows / MSVC++ 7.1</p></li><li><p>
						<a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/CURRENT/ncbi_cxx_mac_cw--Dec_31_2005.tgz"><span class="nctnt ncbi-path">ncbi_cxx_mac_cw--Dec_31_2005.tgz</span></a> — for MacOSX 10.4 / CodeWarrior DevStudio for MacOS 9.5</p></li><li><p>
						<a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/CURRENT/ncbi_cxx_mac_xcode--Dec_31_2005.gtar.gz"><span class="nctnt ncbi-path">ncbi_cxx_mac_xcode--Dec_31_2005.gtar.gz</span></a> — for MacOSX 10.4 / Xcode 1.5-2.2.1</p></li></ul><p>
				<span class="nctnt ncbi-monospace">The sources correspond to the NCBI </span>
production tree 
				<span class="nctnt ncbi-monospace"> sources from patch "CATCHUP_DEC_2005", which in turn roughly corresponds to the </span>
				development tree
				<span class="nctnt ncbi-monospace"> sources from the very beginning of December, 2005.</span>
			</p><p>
				<span class="nctnt ncbi-monospace">There are also two sub-directories, containing easily buildable source distributives of the NCBI C Toolkit (for MS Windows and UNIX) and selected 3rd-party packages (for MS Windows only). These are the versions that the NCBI C++ Toolkit should build with. For build instructions, see README files there:</span></p><ul><li><p><a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/CURRENT/NCBI_C_Toolkit/README"><span class="nctnt ncbi-path">NCBI_C_Toolkit </span></a></p></li><li><p><a href="ftp://ftp.ncbi.nih.gov/toolbox/ncbi_tools++/CURRENT/ThirdParty/README"><span class="nctnt ncbi-path">ThirdParty</span></a></p></li></ul></div></div><div class="sec" id="release_notes_12_31_2005.New_Development"><h2><span class="title">New Development</span></h2><div class="sec" id="release_notes_12_31_2005.CORELIB__Portab"><h3><span class="title">CORELIB — Portability and Application Framework</span></h3><div class="sec" id="release_notes_12_31_2005.A_File_system"><h4><span class="title">A. File system</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CDirEntry</span> — added <span class="nctnt ncbi-monospace">EGetEntriesFlags</span> type and 'flag' versions of <span class="nctnt ncbi-func">GetEntries[Ptr]()</span> methods. Marked all '<span class="nctnt ncbi-monospace">enum</span>' versions as obsolete.</p></dd><dt>2</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-func">CDirEntry::GetObjectType()</span> implemented for all <span class="nctnt ncbi-class">CDirEntry</span> based classes.</p></dd><dt>3</dt><dd><p class="no_top_margin">Added class <span class="nctnt ncbi-class">CFileUtil</span>, now with <span class="nctnt ncbi-func">GetDiskSpace()</span> and <span class="nctnt ncbi-func">GetFreeDiskSpace()</span> methods.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.B_Strings"><h4><span class="title">B. Strings</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-func">NStr::JavaScriptEncode()</span> — new method.</p></dd><dt>2</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-func">NStr::StringToXxx()</span> — the obsolete versions (ones using '<span class="nctnt ncbi-monospace">enum</span>') are removed for good.</p></dd><dt>3</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-func">NStr::XxxToString()</span> — added new radix base parameter.</p></dd><dt>4</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CStringUTF8</span> class — enhanced to also support <span class="nctnt ncbi-monospace">ISO8859-1</span> (Latin1) and Windows <span class="nctnt ncbi-monospace">CP-1252</span> character encodings, and to guess the encoding of a text.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.C_Time"><h4><span class="title">C. Time</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CTime</span> — now using bit fields and fixed-size types for class members, to save memory.</p></dd><dt>2</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CStopWatch</span> —added new constructor <span class="nctnt ncbi-func">CStopWatch(EStart)</span>. <span class="nctnt ncbi-func">CStopWatch(bool)</span> declared as deprecated.</p></dd><dt>3</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-func">CFastLocalTime::GetLocalTimezone()</span> — new method, to obtain the difference between UTC and current local time.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.D_Application_p"><h4><span class="title">D. Application parametrization</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CParam&lt;&gt;</span> — new class to handle application run-time parameters.</p></dd><dt>2</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-class">CTreeNode</span> — redesigned to support both simple values and key-value pairs, added search methods, removed <span class="nctnt ncbi-class">CPairTreeNode</span>.</p></dd><dt>3</dt><dd><p class="no_top_margin">Configuration file syntax now permits the use of <span class="nctnt ncbi-monospace">CR, LF</span> and <span class="nctnt ncbi-monospace">TAB</span> separators in <span class="nctnt ncbi-monospace">.Include, .SubNode</span> etc. values.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.E_Miscellaneous"><h4><span class="title">E. Miscellaneous</span></h4><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Functions to convert <span class="nctnt ncbi-type">"streampos"</span> to/from <span class="nctnt ncbi-type">"Int8"</span>, where possible.</p></dd><dt>2</dt><dd><p class="no_top_margin">
							<span class="nctnt ncbi-macro">ArraySize</span> — template (sometimes a macro) to get a number of elements in a static or stack array.</p></dd></dl></div></div><div class="sec" id="release_notes_12_31_2005.CONNECT__Data_s"><h3><span class="title">CONNECT — Data streaming, Networking, and Dispatching</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">HTTP connectors have got "flushable" feature (controlled by constructor's <span class="nctnt ncbi-cmd">fHCC_Flushable</span> flag):  "flushable" connectors allow forced <span class="nctnt ncbi-func">CONN_Flush()</span> to actually flush the accumulated contents down to server (previous implementation defined this as no operation, until an explicit READ or WAIT-ON-READ conditions).  This flag has also been added to SERVICE connectors (to access the cases when HTTP is the underlying implementation).</p></dd><dt>2</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-monospace">CRC32</span> (pure C version) added; declared in <span class="nctnt ncbi-path">&lt;connect/ncbi_connutil.h&gt;</span>.</p></dd><dt>3</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">CSocket::ReadLine()</span> — new method that mimics behavior of its C counterpart, and is aimed to speed up reading by reducing the number of reallocations when filling up a string.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.UTIL__Miscellan"><h3><span class="title">UTIL — Miscellaneous Low-Level APIs</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Implemented ZIP-style <span class="nctnt ncbi-monospace">CRC32</span>.</p></dd><dt>2</dt><dd><p class="no_top_margin">Ensure that functions to compute <span class="nctnt ncbi-func">erf()</span> and <span class="nctnt ncbi-func">erfc()</span> are available, using system-supplied implementations where possible.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.SERIAL__Data_Se"><h3><span class="title">SERIAL — Data Serialization (ASN.1, XML)</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Enhanced character encoding support in the XML serialization. </p></dd><dt>2</dt><dd><p class="no_top_margin">Corrected stream position calculations to support large (greater than 2GB) files on the capable platforms.</p></dd><dt>3</dt><dd><p class="no_top_margin">Reviewed and classified fail flags in serial object streams to allow for a more detailed reporting of an input stream state after the read operations. In particular, it's now easier to tell whether EOF occured in the middle of an object or between the objects.</p></dd><dt>4</dt><dd><p class="no_top_margin">Added support of <span class="nctnt ncbi-monospace">ASN.1</span>
						<span class="nctnt ncbi-type">BIT STRING</span> data type.</p></dd><dt>5</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">CClassTypeInfoBase::GetRegistered{Module,Class}Names</span> — new methods, to get the list of registered module and class names.</p></dd><dt>6</dt><dd><p class="no_top_margin">Implemented serialization of set and map with custom comparator.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.CGI__CGI_and_Fa"><h3><span class="title">CGI — CGI and Fast-CGI Application Framework</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CCgiUserAgent</span> — new class, to parse user agent strings.</p></dd><dt>2</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CCgiArgs_Parser</span>, <span class="nctnt ncbi-class">CCgiArgs</span> — new classes to handle (parse, hold and compose) CGI arguments.</p></dd><dt>3</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CUrl</span> — new class, to handle URLs.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.HTML__HTML_Gene"><h3><span class="title">HTML — HTML Generation Library</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CHTMLHelper</span> — added <span class="nctnt ncbi-func">HTMLJavaScriptEncode()</span> method.</p></dd><dt>2</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CHTMLOpenElement</span> — added parsing mapping tags <span class="nctnt ncbi-var">&lt;@...@&gt;</span> for attributes.</p></dd><dt>3</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">CHTML_Table::GetCurrent[Row|Col]</span> return 0 instead of -1 if current row/column in the table is not defined. By default, the current row/col in the table always is <span class="nctnt ncbi-monospace">(0,0)</span>.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.BerkeleyDB_API_"><h3><span class="title">BerkeleyDB API (bdb) — Much Enriched C++ API Based On BerkeleyDB</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Implemeneted the reading of BLOBs in cursors.</p></dd><dt>2</dt><dd><p class="no_top_margin">BDB Environment - added methods to customize locking. Added option to place transaction logs in memory.</p></dd><dt>3</dt><dd><p class="no_top_margin">Added a bit vector storage file.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.DBAPI__SQL_Data"><h3><span class="title">DBAPI — SQL Database Connectivity</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">DBAPI driver library names have changed from <span class="nctnt ncbi-lib">'dbapi_driver_*'</span> to <span class="nctnt ncbi-lib">'ncbi_xdbapi_'</span>.</p></dd><dt>2</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-var">CDB_DateTime</span>, <span class="nctnt ncbi-var">CDB_SmallDateTime</span> — now '<span class="nctnt ncbi-monospace">null</span>' if initializated with empty <span class="nctnt ncbi-type">CTime</span>.</p></dd><dt>3</dt><dd><p class="no_top_margin">Handle <span class="nctnt ncbi-type">SYBUNIQUE</span> data type as <span class="nctnt ncbi-type">eDB_VarBinary</span> in case of the FTDS driver.</p></dd><dt>4</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">I_DriverContext</span> - added methods <span class="nctnt ncbi-func">Set/GetApplicationName</span>, <span class="nctnt ncbi-func">Set/GetHostName</span>.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.PYTHON_database"><h3><span class="title">PYTHON database module based on DBAPI</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Now use TDS protocol version 12.5 when connecting to Sybase servers.</p></dd><dt>2</dt><dd><p class="no_top_margin">Fixed the reading of <span class="nctnt ncbi-type">LOB (Text/Image)</span> data types.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.ALGO_ALIGN_SPLI"><h3><span class="title">ALGO/ALIGN/SPLIGN — Spliced Alignment Algorithms</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">Splign</span> code — now uses Object Manager data loaders  and LDS to load and look up sequence data. Handling of sequence data has been optimized which led to a dramatic improvement in the overall batch performance.</p></dd><dt>2</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">CSplign::SetMaxGenomicExtent()</span>— new method, to adjust the maximum span of genomic sequence beyond the hit boundaries where SPLIGN should look for terminal exons.</p></dd><dt>3</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">Compartmentization</span> code has been moved to the <span class="nctnt ncbi-lib">'xalgoalignutil'</span> library. See <span class="nctnt ncbi-class">CCompartmentFinder</span> interface for details.</p></dd><dt>4</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-app">SPLIGN</span> application — now supports traditional pairwise text output.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.ALNMGR__Biosequ"><h3><span class="title">ALNMGR — Bio-sequence Alignment Manager</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CSeq-align:</span>
					</p><ul><li><p>
								<span class="nctnt ncbi-class">CreateDensegFromStdseg</span> now copies the Seq-align's scores too.</p></li></ul></dd><dt>2</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CAlnMixMerger</span> interface:</p><ul><li><p>Derive from <span class="nctnt ncbi-class">CTaskProgressReporter</span>
							</p></li><li><p>Delegated truncation to <span class="nctnt ncbi-class">CDiagRangeCollection</span>.</p></li><li><p>Added progress feedback.</p></li></ul></dd><dt>3</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CAlnMix</span> interface:</p><ul><li><p>Derive from <span class="nctnt ncbi-class">CTaskProgressReporter</span>
							</p></li><li><p>Added progress feedback.</p></li></ul></dd><dt>4</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CAlnMrgApp</span>:</p><ul><li><p>Added optional progress feedback.</p></li></ul></dd><dt>5</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CDiagRangeCollection</span>:</p><ul><li><p>Implemented initial version of a diagonal alignment range collection.</p></li></ul></dd></dl></div><div class="sec" id="release_notes_12_31_2005.BLAST"><h3><span class="title">BLAST</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added search classes to perform preliminary and traceback stages of the BLAST search separately.</p></dd><dt>2</dt><dd><p class="no_top_margin">Added location-transparent, uniform search interface with support for limited functionality (<span class="nctnt ncbi-path">algo/blast/api/uniform_search.hpp</span>).</p></dd><dt>3</dt><dd><p class="no_top_margin">Renamed headers for PSSM engine (<span class="nctnt ncbi-path">algo/blast/api/blast_psi.hpp</span> -&gt; <span class="nctnt ncbi-path">algo/blast/api/pssm_engine.hpp</span>)</p></dd><dt>4</dt><dd><p class="no_top_margin">Added query sequence retrieval interface to allow sequence data retrieval with and without the NCBI C++ Object Manager.</p></dd><dt>5</dt><dd><p class="no_top_margin">Introduced usage of precomputed statistical parameters for blastn:</p><ul><li><p>Added support for blastn reward/penalty values of 1/-5, 3/-4, and 3/-2.</p></li><li><p>Added adjustment of odd blastn scores when match reward = 2</p></li></ul></dd><dt>6</dt><dd><p class="no_top_margin">Simplified and made organization of filtering locations data (<span class="nctnt ncbi-class">BlastMaskLoc::seqloc_array</span>) more consistent throughout the code.</p></dd><dt>7</dt><dd><p class="no_top_margin">Gapped alignment changes in CORE BLAST:</p><ul><li><p>Remove ability to decline alignments</p></li><li><p>Streamlined memory usage in core dynamic programming routines.</p></li><li><p>Capped at 1000 the number of diagonals examined by the greedy aligner.</p></li></ul></dd><dt>8</dt><dd><p class="no_top_margin">Changed convention for unset gap parameters — from zero to negative number.</p></dd><dt>9</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CPsiBl2Seq</span> — new class, to perform PSSM to protein sequence comparisons.</p></dd><dt>10</dt><dd><p class="no_top_margin">New library dependencies: </p><ul><li><p>
								<span class="nctnt ncbi-lib">xblast</span> now depends on composition_adjustment</p></li><li><p>
								<span class="nctnt ncbi-lib">xblastformat</span> now depends on xblast</p></li></ul></dd><dt>11</dt><dd><p class="no_top_margin">Created a new library in <span class="nctnt ncbi-path">algo/blast/composition_adjustment</span>. This library contains routines for adjusting the scoring system of <span class="nctnt ncbi-monospace">blastp</span> and <span class="nctnt ncbi-monospace">tblastn</span> searches to reflect the composition of the sequences being aligned. The <span class="nctnt ncbi-lib">composition_adjustment</span> library is used in the <span class="nctnt ncbi-app">blastpgp</span> executable compiled from the C toolkit. The library is included because it will be used in the future to provide composition adjustment to BLAST executables compiled from the C++ Toolkit. One goal is to implement as much of this functionality as possible in a library that is shared between the C and C++ Toolkits.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.BIOOBJECTS__Bio"><h3><span class="title">BIO-OBJECTS — Bio-Object Specific Utility Functions (Not Involving OM++)</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CSeq_id</span> — allow (but do not require) FASTA-style dbSNP IDs to have "extra" vertical bars per historical practice.</p></dd><dt>2</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CSeq_id</span> — support new prefix "DV" in <span class="nctnt ncbi-func">IdentifyAccession()</span> method.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.BIOTOOLS"><h3><span class="title">BIO-TOOLS</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CFastaOstream</span> — improve output when instantiating gaps, and streamline internally.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.LDS__Local_data"><h3><span class="title">LDS - Local data storage</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Library <span class="nctnt ncbi-lib">lds_admin</span> merged with <span class="nctnt ncbi-lib">lds.lib</span>
					</p></dd><dt>2</dt><dd><p class="no_top_margin">Improved indexing of large files. Use 64-bit file offsets.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.OM__Object_Mana"><h3><span class="title">OM++ — Object Manager — For Retrieving and Processing Bio-Objects</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Implemented new feature editing API.</p></dd><dt>2</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">Added CBioseq_Handle::IsProtein()</span> &amp;<span class="nctnt ncbi-func"> IsNucleotide()</span>.</p></dd><dt>3</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CSeq_loc_Mapper</span> — adjust segment length when mapping between nucleotide and protein.</p></dd><dt>4</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-class">CAlign_CI</span>I — now reports (as warning) and skips alignments of zero length.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.OM_LOADERS_READ"><h3><span class="title">OM++ LOADERS/READERS — Data Retrieval Libraries for OM++</span></h3><p>GenBank data loader:</p><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Changed default reader to ID2.</p></dd><dt>2</dt><dd><p class="no_top_margin">HaveCache(), PurgeCache()— new methods, for cache control.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.OM_DEMO_program"><h3><span class="title">OM++ DEMO program (objmgr_demo)</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Added options to use LDS and BLAST data loaders.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.BUILD_FRAMEWORK"><h3><span class="title">BUILD FRAMEWORK (UNIX)</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Support running the test suite under auxiliary checkers such as 'valgrind'.</p></dd><dt>2</dt><dd><p class="no_top_margin">Add checks for <span class="nctnt ncbi-lib">wxWidgets 2.6</span> and <span class="nctnt ncbi-lib">Freetype</span> (version 2), and make small improvements to checks for other third-party libraries.</p></dd><dt>3</dt><dd><p class="no_top_margin">Add an <span class="nctnt ncbi-macro">NCBI_DEPRECATED</span> macro that can be used to mark deprecated functions.</p></dd><dt>4</dt><dd><p class="no_top_margin">Allow projects to specify "negative requirements", which must be unsatisfied (but known to '<span class="nctnt ncbi-app">configure</span>') for them to be buildable.</p></dd><dt>5</dt><dd><p class="no_top_margin">Relink applications and libraries when their makefiles change (but recompile them only by explicit request, or if actual sources change).</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.PTB__Project_Tr"><h3><span class="title">PTB — Project Tree Builder for MSVC++ .NET</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Pay more attention to project requirements when analyzing project tree — "unrecognized" requirements now result in the exclusion of the project.</p></dd><dt>2</dt><dd><p class="no_top_margin">Added support of negation of requirements in source makefile.</p></dd><dt>3</dt><dd><p class="no_top_margin">Implemented processing of local (for a specific makefile) macros.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.APPLICATIONS"><h3><span class="title">APPLICATIONS</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-app">WindowMasker</span>:</p></dd></dl><p>
				<span class="nctnt ncbi-app">WindowMasker</span> is a program that identifies and masks out highly repetitive DNA sequences and DNA sequences with low complexity in a genome using only the sequence of the genome itself.</p><p>
				<span class="nctnt ncbi-app">WindowMasker</span> is described in: Morgulis A, Gertz EM, Schaffer AA, Agarwala R; WindowMasker: Window based masker for sequence genomes. Bioinformatics, to appear. Advance access: Nov. 15, 2005. (Please cite this paper in any publication that uses WindowMasker.)</p><dl class="labeled-list"><dt>2</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-app">DustMasker</span>:</p></dd></dl><p>
				<span class="nctnt ncbi-app">DustMasker</span> is a program that identifies and masks out low complexity parts of a genome using a new and improved DUST algorithm. The main advantages of the new algorithm are symmetry with respect to taking reverse complements, context insensitivity, and much better performance.</p><p>The new DUST algorithm is described in: Morgulis A, Gertz EM, Schaffer AA, Agarwala R; A Fast and Symmetric DUST Implementation to Mask Low-Complexity DNA Sequences. Journal of Computational Biology, to appear. (Please cite this paper in any publication that uses <span class="nctnt ncbi-app">DustMasker</span>.)</p></div><div class="sec" id="release_notes_12_31_2005.GRID_DISTRIBUTE"><h3><span class="title">GRID (DISTRIBUTED COMPUTING) FRAMEWORK</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-app">NetSchedule</span> - number of performance optimizations and bug fixes.</p></dd><dt>2</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-app">NetCache</span>
					</p><ul><li><p>Implemented new transmission protocol with error checking.</p></li><li><p>Improved error processing. Fixed bugs in communication protocol. </p></li><li><p>Implemented no-wait-if-locked mode in GetBlob. </p></li><li><p>Implemented BLOB locking detection. Improved overflow file management.</p></li></ul></dd><dt>3</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">CWorkerNodeJobContext::PutProgressMessage()</span> — added an optional parameter to allow sending progress messages regardless of the rate control.</p></dd><dt>4</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-func">CRemoteCgiApp::PutProgressMessage()</span> — added an optional parameter to allow sending progress messages regardless of the rate control.</p></dd><dt>5</dt><dd><p class="no_top_margin">
						<span class="nctnt ncbi-monospace">CGridCgiApplication</span> — new optional parameter "expect_complete" in the configuration file, to control for how long the CGI should wait before showing the "wait" HTML page. If during this time the result fromWorker Node is received, then the "real" result HTML page is showed.</p></dd></dl></div></div><div class="sec" id="release_notes_12_31_2005.Documentation"><h2><span class="title">Documentation</span></h2><div class="sec" id="release_notes_12_31_2005.Document_Locati"><h3><span class="title">Document Location</span></h3><p>The documentation is available online at <!--ro--><a href="toc.html">http://www.ncbi.nlm.nih.gov/books/bv.fcgi?rid=toolkit.TOC&amp;depth=2</a> as a book titled "The NCBI C++ Toolkit". This is an online searchable book.</p><p>The C++ Toolkit book also provides PDF version of the chapters; although these are not up to date in this release. The PDF version can be accessed by a link that appears on each page.</p><p>The older HTML documentation has been deprecated and is no longer being updated, and "The NCBI C++ Toolkit" online book at the previously listed URLs is the official documentation.</p></div><div class="sec" id="release_notes_12_31_2005.Document_Conten"><h3><span class="title">Document Content</span></h3><p>Documentation has been grouped into chapters and sections that provide a more logical coherence and flow. New sections and paragraphs continue to be added to update and clarify the older documentation or provide new documentation. The chapter titled "Introduction to the C++ Toolkit" gives an overview of the C++ Toolkit. This chapter contains links to other chapters containing more details on a specific topic and is a good starting point for the new comer.</p><p>The DOXYGEN source browser is used to complement the traditional docs with structured DOXYGEN-generated "Library Reference" ones based on the in-source comments. You can access the DOXYGEN source browser for the NCBI code from:</p><p>
				<a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml">http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml/</a>
			</p><p>The above link is also available under the "Browsers" that appears on each page.</p><p>You can also access the CVS code repository via a Web interface. These links also appear in the sidebar box on each page.</p><p>A C/C++ Symbol Search query appears on each page of the online Toolkit documentation. You can use this to perform a symbol search on the public or in-house versions of LXR, Doxygen and Library. The Library search runs a CGI script that lists the library name where the symbol (such as function) is defined in.</p></div></div><div class="sec" id="release_notes_12_31_2005.Building_on_the"><h2><span class="title">Building on the MacOS</span></h2><p>We now build all the libraries and most of the applications including the Genome Workbench (<span class="nctnt ncbi-app">gbench</span>), some test and a few demo applications. We also build the FLTK library's GUI editor, <span class="nctnt ncbi-app">fluid</span>.</p><p>All apps are built as application bundles except <span class="nctnt ncbi-app">gbench_plugin_scan</span> and <span class="nctnt ncbi-app">datatool</span> which are built as command line apps. Any of the applications can be built as command line apps by tweaking the build scripts or the CodeWarrior projects.</p><div class="sec" id="release_notes_12_31_2005.GCC"><h3><span class="title">GCC</span></h3><p>Uses a regular <span class="nctnt ncbi-monospace">Unix</span> build pattern: run <span class="nctnt ncbi-app">configure</span>, then <span class="nctnt ncbi-app">make</span>.</p></div><div class="sec" id="release_notes_12_31_2005.Xcode"><h3><span class="title">Xcode</span></h3><p>When building the toolkit with <span class="nctnt ncbi-app">Xcode</span>, the latest version of <span class="nctnt ncbi-app">Xcode</span> (at least 1.5) is required for the trouble-free build. Build procedure is as follows: open, build and run a project file in <span class="nctnt ncbi-path">compilers/xCode</span>. This is a GUI tool to generate a new NCBI C++ Toolkit <span class="nctnt ncbi-app">Xcode</span> project. You'll have an option to specify third-party installation directories and choose which packages (libs, applications and tests) to include into the final project. The option to automatically download and install all the third-party libraries is also available.</p><p>
				<span class="nctnt ncbi-app">Xcode</span> build fully supports all the latest Apple innovations: distributed builds, optional CPU specific optimization, pre-compiled headers, fix &amp; continue, code cleanup and Zero Link (with few exceptions).</p><p>
				<span class="nctnt ncbi-app">Xcode</span> build has a Shell Script build phase for each Target dependent on generated ASN files. These shell scripts use <span class="nctnt ncbi-app">datatool</span> to regenerate source files each time the ASN specification files do change.</p><p>
				<span class="nctnt ncbi-app">Xcode</span> builds all libraries as a Mach-O dynamically linked shared ones (<span class="nctnt ncbi-path">.dylib</span>) and all Genome Workbench plugins as Mach-O bundles (also <span class="nctnt ncbi-path">.dylib</span> extension). Note, that <span class="nctnt ncbi-app">Xcode</span> will place Genome Workbench plugins inside Genome Workbench application bundle (<span class="nctnt ncbi-path">Genome Workbench.app/Contents/MacOS/plugins</span>).</p></div><div class="sec" id="release_notes_12_31_2005.CodeWarrior"><h3><span class="title">CodeWarrior</span></h3><p>To build the toolkit with <span class="nctnt ncbi-app">CodeWarrior</span> use an AppleScript editor to open and run the script files <span class="nctnt ncbi-path">makeLibs.met</span> and <span class="nctnt ncbi-path">makeApps.met</span>. You must use a script editor capable of opening a script file larger than 32K, such as Apple's Script Editor v2.0, or Smile. Script Editor v1.9 will not work since makeLibs.met just got too big. The command-line tool <span class="nctnt ncbi-app">osascript</span> also works.</p><p>On running the scripts you will be prompted as to which targets you want to build. Or if you always build the same targets they can be specified by including an empty file or folder in the compilers:mac_prj folder with the name 'Build' followed by the keywords of the targets you want built. The keywords are: Debug and Final. For example, to build only the Debug targets use: "Build Debug", to build both debug and release (final) versions: "Build".</p><p>If you install the C++ Toolkit under a different name than "<span class="nctnt ncbi-path">ncbi_cxx</span>" or in a different location than your home directory, you can edit the script's properties, <span class="nctnt ncbi-var">pRootFolderName</span> and <span class="nctnt ncbi-var">pRootFolderPath</span>, to override these defaults. Note: these paths, and those mentioned below, must be entered in a Mac format (e.g. <span class="nctnt ncbi-path">disk:Users:username</span>
				<i>:</i>) not in Unix format (e.g. <span class="nctnt ncbi-path">/Users/username/</span>). The disk name (and its following colon) may be omitted.</p><p>Certain third party libraries (see <a href="release_notes_12_31_2005.html#release_notes_12_31_2005.T1">Table 1</a>) are required to build some parts of the C++ toolkit. The scripts will try and find them if they are in your home directory, or you can specify where they were installed using properties at the beginning of the script.</p><div class="table" id="release_notes_12_31_2005.T1"><h4><span class="title">Table 1. Third Party Libraries</span></h4><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Library </th><th align="left" valign="top" rowspan="1" colspan="1">Property </th><th align="left" valign="top" rowspan="1" colspan="1">Example</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-lib">FLTK 1.1.6 </span>(w/ NCBI patches #5) </td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-var">pFLTKRootFolder </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">"</span>
								<span class="nctnt ncbi-monospace">home:mhome:fltk-1.1.6-ncbi5"</span>
							</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-lib">BerkeleyDB 4.3.21</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-var">pBdbRootFolder </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">"Users:myhome:mylibs:db-4.3.21"</span><br />
								<span class="nctnt ncbi-monospace">(or "... db-4.3.21")</span>
							</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-lib">SQLite 2.8.13 </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-var">gSqliteFolder </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1"></td></tr></tbody></table></div><p>You do not have to build the <span class="nctnt ncbi-lib">FLTK</span> or <span class="nctnt ncbi-lib">BDB</span> libraries separately. This is done by the scripts and <span class="nctnt ncbi-app">CodeWarrior</span> along with the Toolkit libraries. Just unpack the source bundles in your home directory or where ever you have specified in the appropriate properties. The root folders for FLTK and BDB do not have to have any particular names. If there is more than one version the scripts will grab whichever is last alphabetically (<span class="nctnt ncbi-path">e.g. fltk-1.1.4r2</span> will get used instead of <span class="nctnt ncbi-path">fltk-1.1.3</span>).</p><p>The scripts normally halt on any <span class="nctnt ncbi-app">CodeWarrior</span> compilation errors. If you want them to continue and save errors, set the next script property, <span class="nctnt ncbi-var">pSaveContinueOnErrors</span>, to true. Compilation errors for a project will be saved in a file in the same folder as the project being built, with a name in the following format: <span class="nctnt ncbi-path">projectName-targetNumber.errs</span> (e.g. <span class="nctnt ncbi-path">xncbi-2.errs</span>).</p><p>The Genome Workbench's configuration file(s) is stored in the user's <span class="nctnt ncbi-path">Library:Application Support:gbench</span> folder.</p><p>CFM builds are not supported. OS 8 or 9 are not supported. We know 10.4 works. We think 10.1 still works, 10.2 might work, and 10.3 most probably works.</p></div></div><div class="sec" id="release_notes_12_31_2005.Platforms_OSs__"><h2><span class="title">Platforms (OS's, compilers used inside NCBI)</span></h2><p>This release was successfully tested on at least the following platforms — but may also work on other platforms. Since the previous release, some platforms were dropped from this list, just because we do not use them here anymore, and some were added (these new platforms are highlighted using bold font). Also, it can happen that some projects would not work (or even compile) in the absence of 3rd-party packages, or with older or newer versions of such packages — in these cases, just skipping such projects (e.g. using flag "-k" for <span class="nctnt ncbi-app">make</span> on UNIX), can get you through.</p><div class="table" id="release_notes_12_31_2005.T2"><h4><span class="title">Table 2. Unix OS's and Supported Compilers</span></h4><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Architecture </th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">Linux-2.4.23 (w/ LIBC 2.3.2) </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">INTEL </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								GCC 3.4.0<br />ICC 8.0<br />(GCC 3.0.4, 2.95.3 - nominal support)</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">Linux-2.6.11.10 (w/ LIBC 2.3.3) </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">INTEL-64 </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								GCC 4.0.1<br />ICC 9.0</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">Solaris-8</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">SPARC </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								C++ 5.3 (WorkShop 6u2) patch 111685-<b>23</b> (64-, 32-bit)<br />(GCC 3.4.3 - nominal support)</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">Solaris-10 </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">SPARC </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								Sun C++ 5.5 (Studio8) patch 113817-15,<br /><b>GCC 4.0.1</b>
							</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">Solaris-9 </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">INTEL </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">C++ 5.3 (WorkShop 6u2) patch 11168<b>6</b>-13</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">IRIX64-6.5 </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">SGI-Mips </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">MIPSpro 7.3.1.3m (64-bit, 32-bit)</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">FreeBSD-4.10 </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">INTEL </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">GCC 3.4.2</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">Tru64 (OSF1) V5.1 </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">ALPHA </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">GCC 3.3.2</td></tr></tbody></table></div><div class="table" id="release_notes_12_31_2005.T3"><h4><span class="title">Table 3. MS Windows and Supported Compilers</span></h4><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">MS Windows </span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">
								MSVC++ 7.1. See documentation for building the Toolkit with MS Visual C++ .NET<br />.NOTE: We also have 3rd-party packages archive for this platform, easily built with MSVC++ .NET (7.1).</td></tr></tbody></table></div><div class="table" id="release_notes_12_31_2005.T4"><h4><span class="title">Table 4. Mac OS X, and Supported Compilers</span></h4><table><thead><tr><th align="left" valign="top" rowspan="1" colspan="1">Operating System </th><th align="left" valign="top" rowspan="1" colspan="1">Compilers</th></tr></thead><tbody><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">MacOS 10.</span>
								<span class="nctnt ncbi-monospace">4</span>
								<span class="nctnt ncbi-monospace"> (darwin7.9.0)</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">GCC 3.3</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">MacOS 10.</span>
								<span class="nctnt ncbi-monospace">4</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">CodeWarrior 9.<b>5</b>
							</td></tr><tr><td align="left" valign="top" rowspan="1" colspan="1">
								<span class="nctnt ncbi-monospace">MacOS 10.</span>
								<span class="nctnt ncbi-monospace">4</span>
							</td><td align="left" valign="top" rowspan="1" colspan="1">Xcode 1.5 - 2.2.<b>1</b>
							</td></tr></tbody></table></div></div><div class="sec" id="release_notes_12_31_2005.Caveats_and_Hin"><h2><span class="title">Caveats and Hints</span></h2><div class="sec" id="release_notes_12_31_2005.MacOS_10X___Cod"><h3><span class="title">MacOS 10.X / CodeWarrior 9.2</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Not all of the test or demo applications are built.</p></dd><dt>2</dt><dd><p class="no_top_margin">The source code for the latest release of FLTK (1.1.x), BerkeleyDB (4.3.x) and SQLite (2.x) should be present. See the installation instructions for details.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.MacOS_102_GCC_3"><h3><span class="title">MacOS 10.2/GCC 3.3</span></h3><p>At least the GCC 3.3 update for Dec. 2002 Developers Tools required from Apple.</p></div><div class="sec" id="release_notes_12_31_2005.GCC_295"><h3><span class="title">GCC 2.95</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Poor MT-safety record.</p></dd><dt>2</dt><dd><p class="no_top_margin">Relatively incomplete/incorrect (comparing to modern compilers) STL implementation.</p></dd><dt>3</dt><dd><p class="no_top_margin">It is going to be deprecated in NCBI as soon as we have any significant trouble with its maintenance.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.GCC_304"><h3><span class="title">GCC 3.0.4</span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Destructor of constructed class member is not called when exception is thrown from a method called from class constructor body (fixed in GCC 3.3).</p></dd><dt>2</dt><dd><p class="no_top_margin">STL stream uses locale in thread unsafe way which may result to segmentation fault when run in multithread mode (fixed in GCC 3.3).</p></dd><dt>3</dt><dd><p class="no_top_margin">Long-file support for C++ streams is disabled/broken (first broken in 3.0, fixed in 3.4).</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.GCC_33"><h3><span class="title">GCC 3.3</span></h3><p>Other than the feature described below, GCC 3.3.2 had been very good to us; it had a lot of very ugly bugs finally fixed.</p><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">Painfully slow linking in debug mode on Linux with GCC-3.3 compiler. — Starting with binutils 2.12 linker tries to merge constant/debug strings marked for merging in object files. But it seems it does this job very inefficiently - I've seen messages about it in internet.  GCC starting with version 3.2 marks section of string constants ready for merging, and also has an option to disable this flag in object files (-fno-merge-constants). Adding this flag to compilation stage allows to avoid slow linking. GCC 3.3 also sets merge flag for debug sections and unfortunately there is no option to disable this flag. As a result, linking of debug executables significatly slower than with gcc 3.0.4. The slowdown rate depends on size of debug strings section and it's non-linear, so bigger projects will suffer more of this bug (N^2). Binutils 2.15 fixes this. The link time still 2 times slower than without symbol merge, but the resultant executable is about two times smaller in size, and no compiler patching is necessary. We are still testing it in-house. We had to patch GCC 3.3 in-house with the fix described at http://lists.boost.org/MailArchives/boost/msg53004.php.</p></dd><dt>2</dt><dd><p class="no_top_margin">Long-file support still broken.</p></dd></dl></div><div class="sec" id="release_notes_12_31_2005.GCC_34x__40x"><h3><span class="title">GCC 3.4<b>.x, 4.0.x</b></span></h3><dl class="labeled-list"><dt>1</dt><dd><p class="no_top_margin">The "Painfully slow linking..." (see GCC3.3, [1] above) was an issue, and we had to patch it in-house to speed up, a la GCC 3.3 — until we finally upgraded to binutils 2.15.</p></dd><dt>2</dt><dd><p class="no_top_margin">At least on Linux, ifstream::readsome() does not always work for large files, as it calls an ioctl that doesn't work properly for large files (we didn't test whether 4.0.x fixed this).</p></dd><dt>3</dt><dd><p class="no_top_margin">At least on Linux, GCC 3.4.[0,1] optimizer (very rarely) generates incorrect code when comparing enumerated values in else-ifs. (Fixed in 3.4.2)</p></dd><dt>4</dt><dd><p class="no_top_margin">GCC 3.4.3, 3.4.4 and 4.0 have a bug in the C++ stream library that affects some parts of our code, notably CGI framework. (Fixed in 4.0.1).</p></dd></dl></div></div><div class="sec" id="release_notes_12_31_2005.Last_Updated"><h2><span class="title">Last Updated</span></h2><p>This section last updated on January 25, 2006.</p></div></div></div><div class="col four_col last"><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>In this page</h3></div></div><div class="portlet_content"><ul><li><a href="#release_notes_12_31_2005.Download_Locati">Download Location</a></li><li><a href="#release_notes_12_31_2005.Source_Archive_">Source Archive Contents</a></li><li><a href="#release_notes_12_31_2005.New_Development">New Development</a></li><li><a href="#release_notes_12_31_2005.Documentation">Documentation</a></li><li><a href="#release_notes_12_31_2005.Building_on_the">Building on the MacOS</a></li><li><a href="#release_notes_12_31_2005.Platforms_OSs__">Platforms (OS's, compilers used inside NCBI)</a></li><li><a href="#release_notes_12_31_2005.Caveats_and_Hin">Caveats and Hints</a></li><li><a href="#release_notes_12_31_2005.Last_Updated">Last Updated</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Download</h3></div></div><div class="portlet_content"><ul><li><a href="pdf/release_notes_12_31_2005.pdf">PDF version of this page</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Search</h3></div></div><div class="portlet_content"><form name="frmSymbolSearch" onsubmit="return SymbolSearch('toolkit')" action="/" id="frmSymbolSearch"><table><tbody><tr align="center" valign="middle"><td colspan="2" nowrap="nowrap"><input id="search_box" name="__symbol" style="vertical-align:                                          middle; line-height:normal " onkeypress="SymbolSearchKeyPress('toolkit',event)" type="text" />   
					<label for="search_box" class="offscreen_noflow">Search term</label><button type="submit" value="Search" style="vertical-align:                                          middle; line-height:normal;" onclick="javasrcipt:SymbolSearch('toolkit');">Search</button></td></tr><tr class="internal-only"><th style="text-align:left">Public</th><th style="text-align:left">Internal</th></tr><tr><td><input id="pToolkitAll" name="__symboloc" type="radio" checked="checked" /><label for="pToolkitAll" title="Search in the C/C++ source code and in this Book">All</label></td><td class="internal-only"><input id="iToolkitAll" name="__symboloc" type="radio" /><label for="iToolkitAll" title="Search in the C/C++ source code,  this Book, and Wiki">All</label></td></tr><tr><td><input id="pLXR" name="__symboloc" type="radio" /><label for="pLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td><td class="internal-only"><input id="iLXR" name="__symboloc" type="radio" /><label for="iLXR" title="Search the symbol in the source code using LINUX Cross-Reference browser">LXR</label></td></tr><tr><td><input id="pLib" name="__symboloc" type="radio" /><label for="pLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td><td class="internal-only"><input id="iLib" name="__symboloc" type="radio" /><label for="iLib" title="Locate the Toolkit library(es) where the symbol is defined">Library</label></td></tr></tbody></table></form></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Source Browsers</h3></div></div><div class="portlet_content"><ul><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">LXR</a></li><li class="external-only"><a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">Doxygen</a></li><li class="internal-only">LXR: <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/lxr/ident">public</a> / <a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/lxr/ident">internal</a></li><li class="internal-only">Doxygen: <a href="http://www.ncbi.nlm.nih.gov/IEB/ToolBox/CPP_DOC/doxyhtml">public</a> / <a href="http://intranet.ncbi.nlm.nih.gov/ieb/ToolBox/CPP_DOC/doxyhtml">internal</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>SVN Source Repository</h3></div></div><div class="portlet_content"><ul><li><a href="http://www.ncbi.nlm.nih.gov/bookshelf/br.fcgi?book=toolkit&amp;part=ch_getcode_svn#ch_getcode_svn.code_retrieval">Server</a></li><li><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/c++">Web
                            (dev)</a></li><li><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/trunk/internal/c++">Web
                            (extra)</a></li><li><a href="https://svn.ncbi.nlm.nih.gov/viewvc/toolkit/production/components">Web (prod)</a></li></ul></div></div><div class="portlet"><div class="portlet_head"><div class="portlet_title"><h3>Resources</h3></div></div><div class="portlet_content"><ul><li class="internal-only"><a href="http://intranet.ncbi.nlm.nih.gov/wiki-private/CxxToolkit">Wiki</a></li><li class="internal-only"><a href="http://jira/browse/CXX">JIRA</a></li><li class="internal-only"><a href="http://intranet/ieb/ToolBox/STAT/test_stat/test_stat_ext.cgi"><span title="View results of testsuite runs on the automatic C++ Toolkit builds">Testsuite</span></a></li><li><a href="release_notes.html">Release Notes</a></li><li><a href="ch_getcode_svn.html#ch_getcode_svn.ftp_download">Download</a></li><li><a href="pdf/TOC.pdf">PDF of the book</a></li><li><a href="ch_style.html">Coding Style</a></li><li><a href="ch_faq.html#ch_faq.mailing_lists">Mailing Lists</a></li><li><a href="mailto:CPP-CORE@NCBI.NLM.NIH.GOV"><span title="Contact C++ Toolkit group">Help and Support</span></a></li></ul></div></div></div></div></div></div></body></html>
